// @siemens/ix/components/ix-action-card.js
import { p as proxyCustomElement, H, h, d as Host } from './p-D-__gyet.js';
import { g as getFallbackLabelFromIconName, a as a11yBoolean } from './p-Bb7pDeaQ.js';
import { d as defineCustomElement$4 } from './p-hwPzPVf8.js';
import { d as defineCustomElement$3 } from './p-B45tl2du.js';
import { d as defineCustomElement$2 } from './p-BBwXlKhE.js';

const actionCardCss = ":host{display:block;position:relative;min-width:13.375rem;width:13.375rem;height:7.5rem;min-height:7.5rem;cursor:pointer;margin:0.25rem 0}:host *,:host *::after,:host *::before{box-sizing:border-box}:host ::-webkit-scrollbar-button{display:none}@-moz-document url-prefix(){:host *{scrollbar-color:var(--theme-scrollbar-thumb--background) var(--theme-scrollbar-track--background);scrollbar-width:thin}}:host{}:host ::-webkit-scrollbar{width:0.5rem;height:0.5rem}:host{}:host ::-webkit-scrollbar-track{border-radius:5px;background:var(--theme-scrollbar-track--background)}:host ::-webkit-scrollbar-track:hover{background:var(--theme-scrollbar-track--background--hover)}:host{}:host ::-webkit-scrollbar-thumb{border-radius:5px;background:var(--theme-scrollbar-thumb--background)}:host{}:host ::-webkit-scrollbar-thumb:hover{background:var(--theme-scrollbar-thumb--background--hover)}:host ::-webkit-scrollbar-corner{display:none}:host ix-card{width:100%;height:100%}";

const IxActionCard$1 = /*@__PURE__*/ proxyCustomElement(class IxActionCard extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
        /**
         * Card variant
         */
        this.variant = 'outline';
        /**
         * Card icon
         */
        this.icon = undefined;
        /**
         * Card selection
         */
        this.selected = false;
    }
    getSubheadingTextColor() {
        return this.variant === 'outline' || this.variant === 'filled'
            ? 'soft'
            : undefined;
    }
    render() {
        const ariaLabelledBy = !this.ariaLabelCard && this.heading
            ? 'ix-action-card-heading'
            : undefined;
        return (h(Host, { key: '167db61afb0880052edd6e28ec9e907dbe98a186' }, h("ix-card", { key: 'e8014c544dbadee66135b8def9b868488a07773b', selected: this.selected, variant: this.variant, class: 'pointer', "aria-label": this.ariaLabelCard, "aria-labelledby": ariaLabelledBy, role: ariaLabelledBy ? 'group' : undefined }, h("ix-card-content", { key: 'bcb177b0fbe2a709d3ab9d8479077eb64ef2496a' }, this.icon ? (h("ix-icon", { class: 'icon', name: this.icon, size: "32", "aria-label": this.ariaLabelIcon || getFallbackLabelFromIconName(this.icon) })) : null, h("div", { key: '65aa3882ad626df59598626bf7a67e1431fcf6e4' }, this.heading ? (h("ix-typography", { id: "ix-action-card-heading", "aria-hidden": a11yBoolean(!ariaLabelledBy), format: "h4" }, this.heading)) : null, this.subheading ? (h("ix-typography", { format: "h5", "text-color": this.getSubheadingTextColor() }, this.subheading)) : null, h("slot", { key: '9ffacb3ea7d9642a2adbad41d9e6fc36a8e5118b' }))))));
    }
    static get style() { return actionCardCss; }
}, [257, "ix-action-card", {
        "variant": [1],
        "icon": [1],
        "ariaLabelIcon": [1, "aria-label-icon"],
        "heading": [1],
        "subheading": [1],
        "selected": [4],
        "ariaLabelCard": [1, "aria-label-card"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-action-card", "ix-card", "ix-card-content", "ix-typography"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-action-card":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, IxActionCard$1);
            }
            break;
        case "ix-card":
            if (!customElements.get(tagName)) {
                defineCustomElement$4();
            }
            break;
        case "ix-card-content":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "ix-typography":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
    } });
}

const IxActionCard = IxActionCard$1;
const defineCustomElement = defineCustomElement$1;

export { IxActionCard, defineCustomElement };
