// @siemens/ix/components/ix-custom-field.js
import { p as proxyCustomElement, H, h, d as Host } from './p-D-__gyet.js';
import { H as HookValidationLifecycle } from './p-Dq_H2flK.js';
import { d as defineCustomElement$5 } from './p-B1p0kpQJ.js';
import { d as defineCustomElement$4 } from './p-BIg1TSrD.js';
import { d as defineCustomElement$3 } from './p-DXFN8JXp.js';
import { d as defineCustomElement$2 } from './p-BBwXlKhE.js';

const customFieldCss = ":host{display:block;width:-moz-fit-content;width:fit-content;height:-moz-fit-content;height:fit-content}";

/*
 * SPDX-FileCopyrightText: 2024 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var __decorate = (undefined && undefined.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
        r = Reflect.decorate(decorators, target, key, desc);
    else
        for (var i = decorators.length - 1; i >= 0; i--)
            if (d = decorators[i])
                r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
const CustomField = /*@__PURE__*/ proxyCustomElement(class CustomField extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
        /**
         * A value is required or must be checked for the form to be submittable
         */
        this.required = false;
        this.isInvalid = false;
        this.isValid = false;
        this.isInfo = false;
        this.isWarning = false;
    }
    updateValidationState({ isInvalid, isValid, isInfo, isWarning, }) {
        this.isInvalid = isInvalid;
        this.isValid = isValid;
        this.isInfo = isInfo;
        this.isWarning = isWarning;
    }
    render() {
        return (h(Host, { key: '732dae3e5ba3f37b5eac745ce48338cae5397d9e' }, h("ix-field-wrapper", { key: '5df555252c2de619f3dd5daa082632695a88e3f6', label: this.label, helperText: this.helperText, infoText: this.infoText, warningText: this.warningText, invalidText: this.invalidText, validText: this.validText, showTextAsTooltip: this.showTextAsTooltip, isInvalid: this.isInvalid, isValid: this.isValid, isInfo: this.isInfo, isWarning: this.isWarning, required: this.required }, h("slot", { key: '81d169fce11d76b04876eb3483909bf163c44370' }))));
    }
    get hostElement() { return this; }
    static get style() { return customFieldCss; }
}, [257, "ix-custom-field", {
        "required": [4],
        "label": [1],
        "helperText": [1, "helper-text"],
        "infoText": [1, "info-text"],
        "warningText": [1, "warning-text"],
        "invalidText": [1, "invalid-text"],
        "validText": [1, "valid-text"],
        "showTextAsTooltip": [4, "show-text-as-tooltip"],
        "isInvalid": [32],
        "isValid": [32],
        "isInfo": [32],
        "isWarning": [32]
    }]);
__decorate([
    HookValidationLifecycle({
        includeChildren: true,
    })
], CustomField.prototype, "updateValidationState", null);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-custom-field", "ix-field-label", "ix-field-wrapper", "ix-tooltip", "ix-typography"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-custom-field":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, CustomField);
            }
            break;
        case "ix-field-label":
            if (!customElements.get(tagName)) {
                defineCustomElement$5();
            }
            break;
        case "ix-field-wrapper":
            if (!customElements.get(tagName)) {
                defineCustomElement$4();
            }
            break;
        case "ix-tooltip":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "ix-typography":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
    } });
}

const IxCustomField = CustomField;
const defineCustomElement = defineCustomElement$1;

export { IxCustomField, defineCustomElement };
