// @siemens/ix/components/ix-drawer.js
import { p as proxyCustomElement, H, e as createEvent, h, d as Host } from './p-D-__gyet.js';
import { v as iconClose } from './p-BvsadYCu.js';
import { a as animate } from './p-DhE1t8Qh.js';
import { A as Animation } from './p-C5MWUgDN.js';
import { d as defineCustomElement$3 } from './p-BEDohapR.js';
import { d as defineCustomElement$2 } from './p-1pd_0dhF.js';

const drawerCss = ":host{top:0;right:0;display:block;position:absolute;justify-content:flex-start;align-items:center;opacity:0;max-height:100vh;min-height:1.5rem;background-color:var(--theme-color-1);border-radius:0.25rem;box-shadow:var(--theme-box-shadow-level-3)}:host *,:host *::after,:host *::before{box-sizing:border-box}:host ::-webkit-scrollbar-button{display:none}@-moz-document url-prefix(){:host *{scrollbar-color:var(--theme-scrollbar-thumb--background) var(--theme-scrollbar-track--background);scrollbar-width:thin}}:host{}:host ::-webkit-scrollbar{width:0.5rem;height:0.5rem}:host{}:host ::-webkit-scrollbar-track{border-radius:5px;background:var(--theme-scrollbar-track--background)}:host ::-webkit-scrollbar-track:hover{background:var(--theme-scrollbar-track--background--hover)}:host{}:host ::-webkit-scrollbar-thumb{border-radius:5px;background:var(--theme-scrollbar-thumb--background)}:host{}:host ::-webkit-scrollbar-thumb:hover{background:var(--theme-scrollbar-thumb--background--hover)}:host ::-webkit-scrollbar-corner{display:none}:host .drawer-container{z-index:100}:host .header{display:flex;position:relative;align-items:center;justify-content:flex-end;padding:0.5rem;width:100%;order:1}:host .header .header-content{flex-grow:1;margin-right:1rem}:host .content{position:relative;flex:1;order:2;height:100%;width:100%;overflow-y:auto}:host(.display-none){display:none}";

const Drawer = /*@__PURE__*/ proxyCustomElement(class Drawer extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
        this.open = createEvent(this, "open", 7);
        this.drawerClose = createEvent(this, "drawerClose", 7);
        /**
         * Show or hide the drawer
         */
        this.show = false;
        /**
         * Fired in case of an outside click during drawer showed state
         */
        // eslint-disable-next-line @stencil-community/ban-default-true
        this.closeOnClickOutside = true;
        /**
         * Render the drawer with maximum height
         */
        this.fullHeight = false;
        /**
         * Min width interpreted as REM
         */
        this.minWidth = 16;
        /**
         * Max width interpreted as REM
         */
        this.maxWidth = 28;
        /**
         * Width interpreted as REM if not set to 'auto'
         */
        this.width = this.minWidth;
        this.toggle = false;
        this.callback = this.clickedOutside.bind(this);
        this.showContent = true;
    }
    onShowChanged(newValue, oldValue) {
        if (newValue === oldValue) {
            return;
        }
        this.toggleDrawer(newValue);
    }
    /**
     * Toggle or define show state of drawer
     * @param show Overwrite toggle state with boolean
     */
    async toggleDrawer(show) {
        show = show !== null && show !== void 0 ? show : !this.show;
        if (show) {
            const { defaultPrevented } = this.open.emit();
            if (defaultPrevented) {
                return;
            }
            this.show = true;
            if (!this.toggle && this.divElement) {
                this.slideInRight(this.divElement);
                setTimeout(() => {
                    window.addEventListener('mousedown', this.callback);
                }, Animation.mediumTime);
            }
        }
        else {
            const { defaultPrevented } = this.drawerClose.emit();
            if (defaultPrevented) {
                return;
            }
            this.show = false;
            if (this.toggle && this.divElement) {
                this.slideOutRight(this.divElement);
            }
            window.removeEventListener('mousedown', this.callback);
        }
        this.toggle = this.show;
        return Promise.resolve();
    }
    onCloseClicked() {
        this.toggleDrawer(false);
    }
    clickedOutside(evt) {
        if (!this.closeOnClickOutside) {
            return;
        }
        const target = evt.target;
        const closestElement = target.closest('#div-container');
        const btn = target.closest('#drawer-btn');
        if (evt.target.type !== 'button' &&
            closestElement !== this.divElement &&
            target !== btn) {
            this.show = false;
        }
    }
    getConstrainedWidth(width) {
        return Math.min(Math.max(width, this.minWidth), this.maxWidth);
    }
    slideOutRight(el) {
        const initialWidth = `${this.getConstrainedWidth(this.width === 'auto' ? this.minWidth : this.width)}rem`;
        animate(el, {
            duration: Animation.mediumTime,
            width: [initialWidth, 0],
            opacity: [1, 0],
            easing: 'easeInSine',
            onComplete: () => {
                el.classList.add('display-none');
            },
        });
    }
    slideInRight(el) {
        const targetWidth = `${this.getConstrainedWidth(this.width === 'auto' ? this.minWidth : this.width)}rem`;
        animate(el, {
            duration: Animation.mediumTime,
            width: [0, targetWidth],
            opacity: [0, 1],
            easing: 'easeOutSine',
            onBegin: () => {
                el.classList.remove('display-none');
            },
            onComplete: () => {
                this.showContent = true;
            },
        });
    }
    componentDidLoad() {
        this.toggleDrawer(this.show);
    }
    render() {
        return (h(Host, { key: 'f40c42d1e7b7e8420c4a8f804a46b722580dd132', class: {
                'drawer-container': true,
                'display-none': true,
            }, style: {
                width: '0',
                'max-width': `${this.maxWidth}rem`,
                height: this.fullHeight ? '100%' : 'auto',
                overflow: 'hidden',
            }, ref: (el) => (this.divElement = el), "data-testid": "container", id: "div-container" }, h("div", { key: '810a3738ce5350281d758808ba93981914760877', style: {
                width: this.width === 'auto'
                    ? 'auto'
                    : `${this.getConstrainedWidth(this.width)}rem`,
            } }, h("div", { key: '5601266a5a02c122d1392234fe1c1a755e440bad', class: "header" }, h("div", { key: 'a43e56c621a03939233e10bb907e6f2bd7a7e176', class: "header-content" }, h("slot", { key: 'ab7c9e45ca78c512c1a242f53a616a16a98187fe', name: "header" })), h("ix-icon-button", { key: 'eaaa6f03ffdf2f5e2d9100ab3f45c4a1f46ef63d', class: "close-button", style: {
                display: this.showContent ? 'block' : 'none',
            }, icon: iconClose, iconColor: "color-soft-text", size: "24", variant: "tertiary", onClick: () => this.onCloseClicked(), "data-testid": "close-button", "aria-label": this.ariaLabelCloseButton })), h("div", { key: '170d74d584fb7b641adaa45800144a27f4354d92', class: "content", style: {
                display: this.showContent ? 'block' : 'none',
            } }, h("slot", { key: 'de087859ac50be70b998ab696698df91f6816f21' })))));
    }
    static get watchers() { return {
        "show": ["onShowChanged"]
    }; }
    static get style() { return drawerCss; }
}, [257, "ix-drawer", {
        "show": [1028],
        "closeOnClickOutside": [4, "close-on-click-outside"],
        "fullHeight": [4, "full-height"],
        "minWidth": [2, "min-width"],
        "maxWidth": [2, "max-width"],
        "width": [8],
        "ariaLabelCloseButton": [1, "aria-label-close-button"],
        "showContent": [32],
        "toggleDrawer": [64]
    }, undefined, {
        "show": ["onShowChanged"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-drawer", "ix-icon-button", "ix-spinner"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-drawer":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, Drawer);
            }
            break;
        case "ix-icon-button":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "ix-spinner":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
    } });
}

const IxDrawer = Drawer;
const defineCustomElement = defineCustomElement$1;

export { IxDrawer, defineCustomElement };
