// @siemens/ix/components/ix-flip-tile-content.js
import { p as proxyCustomElement, H, h, d as Host } from './p-D-__gyet.js';

const flipTileContentCss = ":host{display:block}";

const FlipTileContent = /*@__PURE__*/ proxyCustomElement(class FlipTileContent extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
        /**
         * Controls the visibility of the content
         *
         * @internal
         */
        this.contentVisible = false;
    }
    render() {
        return h(Host, { key: '264e3539d6c5abcd8716968a4128da99ca191375' }, this.contentVisible ? h("slot", null) : null);
    }
    static get style() { return flipTileContentCss; }
}, [257, "ix-flip-tile-content", {
        "contentVisible": [4, "content-visible"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-flip-tile-content"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-flip-tile-content":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, FlipTileContent);
            }
            break;
    } });
}

const IxFlipTileContent = FlipTileContent;
const defineCustomElement = defineCustomElement$1;

export { IxFlipTileContent, defineCustomElement };
