// @siemens/ix/components/ix-layout-auto.js
import { p as proxyCustomElement, H, h, d as Host } from './p-D-__gyet.js';

const layoutAutoCss = ":host{display:block;position:relative;margin:0 0.75rem;--ix-layout-grid-gap:1.5rem}:host .container{display:flex;align-items:stretch;flex-wrap:wrap;gap:var(--ix-layout-grid-gap)}:host ::slotted(*){flex-grow:0;flex-shrink:0}";

const LayoutForm = /*@__PURE__*/ proxyCustomElement(class LayoutForm extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
        /**
         * Defines the layout of the form.
         */
        this.layout = [
            { minWidth: '0', columns: 1 },
            { minWidth: '48em', columns: 2 },
        ];
        this.mediaQueryList = [];
        this.observer = new MutationObserver(() => this.calculateGridTemplateColumns());
        this.resizeObserver = new ResizeObserver(() => {
            this.calculateGridTemplateColumns();
        });
    }
    connectedCallback() {
        this.observer.observe(this.hostElement, {
            childList: true,
            subtree: true,
            attributes: true,
            attributeFilter: ['data-colspan'],
        });
        this.resizeObserver.observe(this.hostElement);
        this.calculateGridTemplateColumns();
    }
    componentWillLoad() {
        this.calculateGridTemplateColumns();
    }
    disconnectedCallback() {
        if (this.observer) {
            this.observer.disconnect();
        }
        if (this.resizeObserver) {
            this.resizeObserver.disconnect();
        }
    }
    updateMediaQueryList() {
        this.mediaQueryList = [];
        this.layout.forEach((layout) => {
            const mediaQuery = window.matchMedia(`(min-width: ${layout.minWidth})`);
            this.mediaQueryList.push({
                mediaQuery,
                layout,
            });
        });
    }
    parseNumber(number) {
        if (!number) {
            return 1;
        }
        const result = parseInt(number);
        if (isNaN(result)) {
            return 1;
        }
        return result;
    }
    calculateGridTemplateColumns() {
        this.updateMediaQueryList();
        let layoutColumns = 1;
        let columnSpacing = 'var(--ix-layout-grid-gap)';
        this.mediaQueryList.forEach((mediaQuery) => {
            if (mediaQuery.mediaQuery.matches) {
                layoutColumns = mediaQuery.layout.columns;
            }
        });
        Array.from(this.hostElement.children).forEach((child) => {
            let colspan = this.parseNumber(child.getAttribute('data-colspan'));
            colspan = Math.min(colspan, layoutColumns);
            const childRatio = colspan / layoutColumns;
            child.style.width = `calc(${childRatio * 99.9}% - ${1 - childRatio} * ${columnSpacing})`;
        });
    }
    render() {
        return (h(Host, { key: '48511195af19191b832713a828ac0945b06a4898' }, h("div", { key: '4a51469ffe227db5c361734fc3122a98f7e0fbcc', class: "container" }, h("slot", { key: '6ab367e83612df056e40392bced22c5f80457486' }))));
    }
    get hostElement() { return this; }
    static get watchers() { return {
        "layout": ["updateMediaQueryList"]
    }; }
    static get style() { return layoutAutoCss; }
}, [257, "ix-layout-auto", {
        "layout": [16]
    }, undefined, {
        "layout": ["updateMediaQueryList"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-layout-auto"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-layout-auto":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, LayoutForm);
            }
            break;
    } });
}

const IxLayoutAuto = LayoutForm;
const defineCustomElement = defineCustomElement$1;

export { IxLayoutAuto, defineCustomElement };
