// @siemens/ix/components/ix-menu-about-item.js
import { p as proxyCustomElement, H, e as createEvent, h, d as Host } from './p-D-__gyet.js';

const menuAboutItemCss = ":host{display:block}";

const MenuAboutItem = /*@__PURE__*/ proxyCustomElement(class MenuAboutItem extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
        this.labelChange = createEvent(this, "labelChange", 7);
    }
    watchLabel(newValue, oldValue) {
        this.labelChange.emit({
            name: 'ix-menu-about-item',
            oldLabel: oldValue,
            newLabel: newValue,
        });
    }
    render() {
        return (h(Host, { key: 'c150379d52f7dd0744ec480c35909937abef9c90' }, h("slot", { key: '37edb415a550ba73903bd6142162c42fe7be9d34' })));
    }
    static get watchers() { return {
        "label": ["watchLabel"]
    }; }
    static get style() { return menuAboutItemCss; }
}, [257, "ix-menu-about-item", {
        "label": [513]
    }, undefined, {
        "label": ["watchLabel"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-menu-about-item"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-menu-about-item":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, MenuAboutItem);
            }
            break;
    } });
}

const IxMenuAboutItem = MenuAboutItem;
const defineCustomElement = defineCustomElement$1;

export { IxMenuAboutItem, defineCustomElement };
