// @siemens/ix/components/ix-modal-loading.js
import { p as proxyCustomElement, H, h, d as Host } from './p-D-__gyet.js';
import { d as defineCustomElement$2 } from './p-1pd_0dhF.js';

const modalLoadingCss = ":host{display:flex;justify-content:flex-start;align-items:center;gap:0.5rem;overflow:hidden}:host .loading-text{display:block;position:relative;overflow:hidden;white-space:nowrap;text-overflow:ellipsis}";

const ModalLoading = /*@__PURE__*/ proxyCustomElement(class ModalLoading extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
    }
    render() {
        return (h(Host, { key: '404214a461e55c9ee78aad8dc277dab5be8a9b42' }, h("ix-spinner", { key: '96278c04d8c7798b411f6807e5c7a1ea26b03d45', variant: "primary" }), h("span", { key: 'cad3eec78defa832ffc679f86941ffc301bc5b5c', class: 'loading-text' }, h("slot", { key: '96cf92b74f62260ee97b7bdae73557a2ee5517ec' }))));
    }
    static get style() { return modalLoadingCss; }
}, [257, "ix-modal-loading"]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-modal-loading", "ix-spinner"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-modal-loading":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModalLoading);
            }
            break;
        case "ix-spinner":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
    } });
}

const IxModalLoading = ModalLoading;
const defineCustomElement = defineCustomElement$1;

export { IxModalLoading, defineCustomElement };
