// @siemens/ix/components/ix-message-bar.js
import { p as proxyCustomElement, H, e as createEvent, h, d as Host } from './p-D-__gyet.js';
import { a as animate } from './p-DhE1t8Qh.js';
import { J as iconNotification, o as iconInfo, r as iconSuccess, p as iconWarning, K as iconWarningRhomb, q as iconError, v as iconClose } from './p-BvsadYCu.js';
import { d as defineCustomElement$3 } from './p-BEDohapR.js';
import { d as defineCustomElement$2 } from './p-1pd_0dhF.js';

const messageBarCss = ":host{margin:0.5rem 0.5rem 0rem 0.5rem}:host *,:host *::after,:host *::before{box-sizing:border-box}:host ::-webkit-scrollbar-button{display:none}@-moz-document url-prefix(){:host *{scrollbar-color:var(--theme-scrollbar-thumb--background) var(--theme-scrollbar-track--background);scrollbar-width:thin}}:host{}:host ::-webkit-scrollbar{width:0.5rem;height:0.5rem}:host{}:host ::-webkit-scrollbar-track{border-radius:5px;background:var(--theme-scrollbar-track--background)}:host ::-webkit-scrollbar-track:hover{background:var(--theme-scrollbar-track--background--hover)}:host{}:host ::-webkit-scrollbar-thumb{border-radius:5px;background:var(--theme-scrollbar-thumb--background)}:host{}:host ::-webkit-scrollbar-thumb:hover{background:var(--theme-scrollbar-thumb--background--hover)}:host ::-webkit-scrollbar-corner{display:none}:host .message-container{display:flex;flex-direction:row;align-items:flex-start;flex-wrap:nowrap;justify-content:space-between;min-height:3.375rem;padding:calc(0.75rem - var(--theme-message-bar--border-thickness)) 0.75rem calc(0.75rem - var(--theme-message-bar--border-thickness)) 1rem;border-radius:var(--theme-message-bar--border-radius);background-color:var(--theme-messagebar--background)}:host .alarm{border:solid var(--theme-message-bar--border-thickness) var(--theme-color-alarm)}:host .danger{border:solid var(--theme-message-bar--border-thickness) var(--theme-color-alarm)}:host .critical{border:solid var(--theme-message-bar--border-thickness) var(--theme-color-critical)}:host .warning{border:solid var(--theme-message-bar--border-thickness) var(--theme-color-warning)}:host .success{border:solid var(--theme-message-bar--border-thickness) var(--theme-color-success)}:host .info{border:solid var(--theme-message-bar--border-thickness) var(--theme-color-info)}:host .neutral{border:solid var(--theme-message-bar--border-thickness) var(--theme-color-neutral)}:host .primary{border:solid var(--theme-message-bar--border-thickness) var(--theme-color-primary)}:host .message-content{flex-grow:1;align-self:center;min-height:1.25rem;padding:0 1rem;font-weight:normal;white-space:normal;overflow:hidden;text-overflow:ellipsis;white-space:nowrap}:host ix-icon{margin-top:0.25rem}:host .message-bar-hidden{display:none}";

const MessageBar = /*@__PURE__*/ proxyCustomElement(class MessageBar extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
        this.closedChange = createEvent(this, "closedChange", 7);
        this.closeAnimationCompleted = createEvent(this, "closeAnimationCompleted", 7);
        /**
         * Specifies the type of the alert.
         */
        this.type = 'info';
        /**
         * If true, close button is disabled and alert cannot be dismissed by the user
         */
        this.persistent = false;
    }
    componentWillRender() {
        const config = MessageBar.messageTypeConfigs[this.type];
        if (config) {
            this.icon = config.icon;
            this.color = config.color;
        }
    }
    closeAlert(el) {
        const { defaultPrevented } = this.closedChange.emit();
        if (!defaultPrevented) {
            animate(el, {
                duration: MessageBar.duration,
                opacity: [1, 0],
                easing: 'easeOutSine',
                onComplete: () => {
                    el.classList.add('message-bar-hidden');
                    this.closeAnimationCompleted.emit();
                },
            });
        }
    }
    render() {
        return (h(Host, { key: '538fcc08ad07fe808f424e067a361f4a56f4c0ec' }, h("div", { key: '4a75052efe1b22c247dcfcd298aabbd1098ff7b6', class: { 'message-container': true, [this.type]: true }, role: "alert", ref: (el) => (this.divElement = el) }, h("ix-icon", { key: 'd348829692e38a1023e044834aefa45516b4d916', color: this.color, name: this.icon }), h("div", { key: '0ac0a509efe1dd246f7732660e4cf16416c8655d', class: "message-content" }, h("slot", { key: '6143bfa9cf39d32d149c2893b7712358337833c9' })), !this.persistent && (h("ix-icon-button", { key: 'b118c91956af0fd00d02cfefdb7da36d46f2d360', icon: iconClose, iconColor: "color-soft-text", size: "24", variant: "tertiary", onClick: () => {
                if (this.divElement) {
                    this.closeAlert(this.divElement);
                }
            }, "data-testid": "close-btn" })))));
    }
    static get style() { return messageBarCss; }
}, [257, "ix-message-bar", {
        "type": [1],
        "persistent": [4],
        "icon": [32],
        "color": [32]
    }]);
MessageBar.duration = 300;
MessageBar.messageTypeConfigs = {
    //TODO(IX-3400): Replace icon colors with proper CSS variables when available
    alarm: { icon: iconError, color: 'color-alarm' },
    critical: { icon: iconWarningRhomb, color: 'color-critical' },
    warning: { icon: iconWarning, color: 'color-warning-text' },
    success: { icon: iconSuccess, color: 'color-success' },
    info: { icon: iconInfo, color: 'color-info' },
    neutral: { icon: iconNotification, color: 'color-neutral' },
    primary: { icon: iconNotification, color: 'color-primary' },
};
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-message-bar", "ix-icon-button", "ix-spinner"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-message-bar":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, MessageBar);
            }
            break;
        case "ix-icon-button":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "ix-spinner":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
    } });
}

const IxMessageBar = MessageBar;
const defineCustomElement = defineCustomElement$1;

export { IxMessageBar, defineCustomElement };
