// @siemens/ix/components/ix-menu-settings-item.js
import { p as proxyCustomElement, H, e as createEvent, h, d as Host } from './p-D-__gyet.js';

const menuSettingsItemCss = ":host{display:block}";

const MenuSettingsItem = /*@__PURE__*/ proxyCustomElement(class MenuSettingsItem extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
        this.labelChange = createEvent(this, "labelChange", 7);
    }
    watchLabel(newValue, oldValue) {
        this.labelChange.emit({
            name: 'ix-menu-settings-item',
            oldLabel: oldValue,
            newLabel: newValue,
        });
    }
    render() {
        return (h(Host, { key: '69dac244a2c3b427237e8dceec64fad9c16600e9' }, h("slot", { key: 'ccfc87d1f061f0d4195f943dec3782b655626aae' })));
    }
    static get watchers() { return {
        "label": ["watchLabel"]
    }; }
    static get style() { return menuSettingsItemCss; }
}, [257, "ix-menu-settings-item", {
        "label": [513]
    }, undefined, {
        "label": ["watchLabel"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-menu-settings-item"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-menu-settings-item":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, MenuSettingsItem);
            }
            break;
    } });
}

const IxMenuSettingsItem = MenuSettingsItem;
const defineCustomElement = defineCustomElement$1;

export { IxMenuSettingsItem, defineCustomElement };
