// @siemens/ix/components/ix-time-input.js
import { p as proxyCustomElement, H, e as createEvent, h, d as Host } from './p-D-__gyet.js';
import { Q as iconClock } from './p-BvsadYCu.js';
import { D as DateTime } from './p-bDicGJUN.js';
import { h as handleSubmitOnEnterKeydown, a as addDisposableChangesAndVisibilityObservers, b as adjustPaddingForStartAndEnd, S as SlotEnd, c as SlotStart } from './p-DUU5I7wZ.js';
import { H as HookValidationLifecycle, c as createClassMutationObserver, g as getValidationText } from './p-Dq_H2flK.js';
import { m as makeRef } from './p-bcj7UEIC.js';
import { h as handleIconClick, o as openDropdown, c as closeDropdown, a as createValidityState } from './p-BvFOMxgl.js';
import { d as defineCustomElement$b } from './p-C6_1qmP4.js';
import { d as defineCustomElement$a } from './p-COWPG_SD.js';
import { d as defineCustomElement$9 } from './p-BI7B-PUA.js';
import { d as defineCustomElement$8 } from './p-B1p0kpQJ.js';
import { d as defineCustomElement$7 } from './p-BIg1TSrD.js';
import { d as defineCustomElement$6 } from './p-BEDohapR.js';
import { d as defineCustomElement$5 } from './p-1pd_0dhF.js';
import { d as defineCustomElement$4 } from './p-BmMUyJRQ.js';
import { d as defineCustomElement$3 } from './p-DXFN8JXp.js';
import { d as defineCustomElement$2 } from './p-BBwXlKhE.js';

const timeInputCss = "input{min-height:2rem;width:auto;padding:0.25rem 0.5rem;background-color:var(--theme-input--background);color:var(--theme-input--color);-webkit-appearance:textfield;-moz-appearance:textfield;appearance:textfield;text-overflow:ellipsis;border:var(--theme-input--border-thickness, 1px) solid var(--theme-input--border-color);border-radius:var(--theme-input--border-radius);box-shadow:var(--theme-input--box-shadow);font-feature-settings:\"clig\" off, \"liga\" off;font-family:Siemens Sans, Siemens Sans, Arial, Helvetica, sans-serif;font-style:normal;font-size:var(--theme-ms-0);line-height:var(--theme-line-height-md);font-weight:var(--theme-font-weight-normal);letter-spacing:var(--theme-letter-spacing-xl);text-decoration:none;-webkit-font-smoothing:antialiased;-moz-osx-font-smooting:grayscale}input[type=number]{text-align:right}input[type=number]::-webkit-inner-spin-button{margin-right:-2px;margin-left:2px;display:none}input:-webkit-autofill{-webkit-box-shadow:0 0 0 1000px var(--theme-color-component-info) inset !important;-webkit-text-fill-color:var(--theme-input--color--autofill) !important;background-color:var(--theme-input--background--autofill) !important;border:var(--theme-input--border-thickness, 1px) solid var(--theme-input--border-color--autofill) !important;color:var(--theme-input--color--autofill) !important}input:-webkit-autofill,input:autofill{-webkit-box-shadow:0 0 0 1000px var(--theme-color-component-info) inset !important;-webkit-text-fill-color:var(--theme-input--color--autofill) !important;background-color:var(--theme-input--background--autofill) !important;border:var(--theme-input--border-thickness, 1px) solid var(--theme-input--border-color--autofill) !important;color:var(--theme-input--color--autofill) !important}input::-moz-placeholder{color:var(--theme-input-hint--color)}input::placeholder{color:var(--theme-input-hint--color)}input.hover:not(.readonly,.read-only,.disabled,[readonly],[disabled],:-moz-read-only),input:hover:not(.readonly,.read-only,.disabled,[readonly],[disabled],:-moz-read-only){border-color:var(--theme-input--border-color--hover) !important;background-color:var(--theme-input--background--hover)}input.hover:not(.readonly,.read-only,.disabled,[readonly],[disabled],:read-only),input:hover:not(.readonly,.read-only,.disabled,[readonly],[disabled],:read-only){border-color:var(--theme-input--border-color--hover) !important;background-color:var(--theme-input--background--hover)}input.focus:not(.readonly,.read-only,.disabled,[readonly],[disabled],:-moz-read-only),input:focus:not(.readonly,.read-only,.disabled,[readonly],[disabled],:-moz-read-only){outline:1px solid var(--theme-color-focus-bdr);outline-offset:var(--theme-input--focus--outline-offset);border-color:var(--theme-input--border-color--focus) !important}input.focus:not(.readonly,.read-only,.disabled,[readonly],[disabled],:read-only),input:focus:not(.readonly,.read-only,.disabled,[readonly],[disabled],:read-only){outline:1px solid var(--theme-color-focus-bdr);outline-offset:var(--theme-input--focus--outline-offset);border-color:var(--theme-input--border-color--focus) !important}input:-moz-read-only{box-shadow:none;background-color:transparent;outline:none;border:none;border-radius:0;border-bottom:var(--theme-input--border-thickness, 1px) solid var(--theme-input--border-color-bottom--readonly)}input.read-only,input:read-only{box-shadow:none;background-color:transparent;outline:none;border:none;border-radius:0;border-bottom:var(--theme-input--border-thickness, 1px) solid var(--theme-input--border-color-bottom--readonly)}input.read-only::-moz-placeholder,input:read-only::-moz-placeholder{color:transparent}input:-moz-read-only::placeholder{color:transparent}input.read-only::placeholder,input:read-only::placeholder{color:transparent}input:disabled,input.disabled{box-shadow:none;background-color:transparent;outline:none;border:none;border-radius:0;color:var(--theme-input--color--disabled);border-bottom:var(--theme-input--border-thickness, 1px) solid var(--theme-input--border-color-bottom--disabled)}input:disabled::-moz-placeholder,input.disabled::-moz-placeholder{color:transparent}input:disabled::placeholder,input.disabled::placeholder{color:transparent}textarea{min-height:2rem;width:auto;padding:0.25rem 0.5rem;background-color:var(--theme-input--background);color:var(--theme-input--color);-webkit-appearance:textfield;-moz-appearance:textfield;appearance:textfield;text-overflow:ellipsis;border:var(--theme-input--border-thickness, 1px) solid var(--theme-input--border-color);border-radius:var(--theme-input--border-radius);box-shadow:var(--theme-input--box-shadow);font-feature-settings:\"clig\" off, \"liga\" off;font-family:Siemens Sans, Siemens Sans, Arial, Helvetica, sans-serif;font-style:normal;font-size:var(--theme-ms-0);line-height:var(--theme-line-height-md);font-weight:var(--theme-font-weight-normal);letter-spacing:var(--theme-letter-spacing-xl);text-decoration:none;-webkit-font-smoothing:antialiased;-moz-osx-font-smooting:grayscale}textarea[type=number]{text-align:right}textarea[type=number]::-webkit-inner-spin-button{margin-right:-2px;margin-left:2px;display:none}textarea:-webkit-autofill{-webkit-box-shadow:0 0 0 1000px var(--theme-color-component-info) inset !important;-webkit-text-fill-color:var(--theme-input--color--autofill) !important;background-color:var(--theme-input--background--autofill) !important;border:var(--theme-input--border-thickness, 1px) solid var(--theme-input--border-color--autofill) !important;color:var(--theme-input--color--autofill) !important}textarea:-webkit-autofill,textarea:autofill{-webkit-box-shadow:0 0 0 1000px var(--theme-color-component-info) inset !important;-webkit-text-fill-color:var(--theme-input--color--autofill) !important;background-color:var(--theme-input--background--autofill) !important;border:var(--theme-input--border-thickness, 1px) solid var(--theme-input--border-color--autofill) !important;color:var(--theme-input--color--autofill) !important}textarea::-moz-placeholder{color:var(--theme-input-hint--color)}textarea::placeholder{color:var(--theme-input-hint--color)}textarea.hover:not(.readonly,.read-only,.disabled,[readonly],[disabled],:-moz-read-only),textarea:hover:not(.readonly,.read-only,.disabled,[readonly],[disabled],:-moz-read-only){border-color:var(--theme-input--border-color--hover) !important;background-color:var(--theme-input--background--hover)}textarea.hover:not(.readonly,.read-only,.disabled,[readonly],[disabled],:read-only),textarea:hover:not(.readonly,.read-only,.disabled,[readonly],[disabled],:read-only){border-color:var(--theme-input--border-color--hover) !important;background-color:var(--theme-input--background--hover)}textarea.focus:not(.readonly,.read-only,.disabled,[readonly],[disabled],:-moz-read-only),textarea:focus:not(.readonly,.read-only,.disabled,[readonly],[disabled],:-moz-read-only){outline:1px solid var(--theme-color-focus-bdr);outline-offset:var(--theme-input--focus--outline-offset);border-color:var(--theme-input--border-color--focus) !important}textarea.focus:not(.readonly,.read-only,.disabled,[readonly],[disabled],:read-only),textarea:focus:not(.readonly,.read-only,.disabled,[readonly],[disabled],:read-only){outline:1px solid var(--theme-color-focus-bdr);outline-offset:var(--theme-input--focus--outline-offset);border-color:var(--theme-input--border-color--focus) !important}textarea:-moz-read-only{box-shadow:none;background-color:transparent;outline:none;border:none;border-radius:0;border-bottom:var(--theme-input--border-thickness, 1px) solid var(--theme-input--border-color-bottom--readonly)}textarea.read-only,textarea:read-only{box-shadow:none;background-color:transparent;outline:none;border:none;border-radius:0;border-bottom:var(--theme-input--border-thickness, 1px) solid var(--theme-input--border-color-bottom--readonly)}textarea.read-only::-moz-placeholder,textarea:read-only::-moz-placeholder{color:transparent}textarea:-moz-read-only::placeholder{color:transparent}textarea.read-only::placeholder,textarea:read-only::placeholder{color:transparent}textarea:disabled,textarea.disabled{box-shadow:none;background-color:transparent;outline:none;border:none;border-radius:0;color:var(--theme-input--color--disabled);border-bottom:var(--theme-input--border-thickness, 1px) solid var(--theme-input--border-color-bottom--disabled)}textarea:disabled::-moz-placeholder,textarea.disabled::-moz-placeholder{color:transparent}textarea:disabled::placeholder,textarea.disabled::placeholder{color:transparent}textarea{min-height:2rem;height:3.25rem;padding:calc(0.375rem - var(--theme-input--border-thickness)) calc(0.5rem - var(--theme-input--border-thickness))}textarea.ix-info:not(.disabled):not(:disabled):not([disabled]),input.ix-info:not(.disabled):not(:disabled):not([disabled]){border-color:var(--theme-input--border-color--info)}textarea.ix-info:not(.disabled):not(:disabled):not([disabled]):hover,input.ix-info:not(.disabled):not(:disabled):not([disabled]):hover{border-color:var(--theme-input--border-color--info--hover) !important}textarea.ix-info:not(.disabled):not(:disabled):not([disabled]):active,input.ix-info:not(.disabled):not(:disabled):not([disabled]):active{border-color:var(--theme-input--border-color--info--active) !important}textarea.ix-warning:not(.disabled):not(:disabled):not([disabled]),input.ix-warning:not(.disabled):not(:disabled):not([disabled]){background-color:var(--theme-input--background--warning);border-color:var(--theme-input--border-color--warning--active) !important}textarea.ix-warning:not(.disabled):not(:disabled):not([disabled]):hover,input.ix-warning:not(.disabled):not(:disabled):not([disabled]):hover{background-color:var(--theme-input--background--warning--hover);border-color:var(--theme-input--border-color--warning--hover) !important}textarea.ix-warning:not(.disabled):not(:disabled):not([disabled]):active,input.ix-warning:not(.disabled):not(:disabled):not([disabled]):active{border-color:var(--theme-input--border-color--warning--active) !important}textarea[class*=ix-invalid]:not(.disabled):not(:disabled):not([disabled]),input[class*=ix-invalid]:not(.disabled):not(:disabled):not([disabled]){background-color:var(--theme-input--background--invalid);border-color:var(--theme-input--border-color--invalid) !important}textarea[class*=ix-invalid]:not(.disabled):not(:disabled):not([disabled]):hover,input[class*=ix-invalid]:not(.disabled):not(:disabled):not([disabled]):hover{background-color:var(--theme-input--background--invalid--hover);border-color:var(--theme-input--border-color--invalid--hover) !important}textarea[class*=ix-invalid]:not(.disabled):not(:disabled):not([disabled]):active,input[class*=ix-invalid]:not(.disabled):not(:disabled):not([disabled]):active{border-color:var(--theme-input--border-color--invalid--active) !important}:host{display:inline-block;position:relative;width:auto}:host *,:host *::after,:host *::before{box-sizing:border-box}:host ::-webkit-scrollbar-button{display:none}@-moz-document url-prefix(){:host *{scrollbar-color:var(--theme-scrollbar-thumb--background) var(--theme-scrollbar-track--background);scrollbar-width:thin}}:host{}:host ::-webkit-scrollbar{width:0.5rem;height:0.5rem}:host{}:host ::-webkit-scrollbar-track{border-radius:5px;background:var(--theme-scrollbar-track--background)}:host ::-webkit-scrollbar-track:hover{background:var(--theme-scrollbar-track--background--hover)}:host{}:host ::-webkit-scrollbar-thumb{border-radius:5px;background:var(--theme-scrollbar-thumb--background)}:host{}:host ::-webkit-scrollbar-thumb:hover{background:var(--theme-scrollbar-thumb--background--hover)}:host ::-webkit-scrollbar-corner{display:none}:host .input-wrapper{display:flex;position:relative;align-items:center;width:100%;height:100%}:host input{width:100%;height:100%}:host .start-container,:host .end-container{display:flex;position:absolute;align-items:center;justify-content:center;z-index:1}:host .start-container{left:0}:host .end-container{right:0}:host .start-container ::slotted(*){margin-left:0.5rem}:host .start-container ::slotted(ix-icon.size-24),:host .start-container ::slotted(ix-icon-button.btn-icon-16){margin-left:0.25rem}:host .start-container ::slotted(ix-icon-button.btn-icon-32){margin-left:0}:host .end-container ::slotted(*){margin-right:0.5rem}:host .end-container ::slotted(ix-icon.size-24),:host .end-container ::slotted(ix-icon-button.btn-icon-16){margin-right:0.25rem}:host .end-container ::slotted(ix-icon-button.btn-icon-32){margin-right:0}:host .bottom-text{margin-top:0.25rem;margin-bottom:0.25rem}:host(.disabled){pointer-events:none}:host(.disabled) input,:host(.disabled) textarea{pointer-events:none}:host(.ix-info:not(.disabled):not(:disabled):not([disabled])) input{border-color:var(--theme-input--border-color--info)}:host(.ix-info:not(.disabled):not(:disabled):not([disabled])) input:hover{border-color:var(--theme-input--border-color--info--hover) !important}:host(.ix-info:not(.disabled):not(:disabled):not([disabled])) input:active{border-color:var(--theme-input--border-color--info--active) !important}:host(.ix-warning:not(.disabled):not(:disabled):not([disabled])) input{background-color:var(--theme-input--background--warning);border-color:var(--theme-input--border-color--warning--active) !important}:host(.ix-warning:not(.disabled):not(:disabled):not([disabled])) input:hover{background-color:var(--theme-input--background--warning--hover);border-color:var(--theme-input--border-color--warning--active) !important}:host(.ix-warning:not(.disabled):not(:disabled):not([disabled])) input:active{border-color:var(--theme-input--border-color--warning--active) !important}:host([class*=ix-invalid]:not(.disabled):not(:disabled):not([disabled])) input,:host(.ix-invalid--required:not(.disabled):not(:disabled):not([disabled])) input{background-color:var(--theme-input--background--invalid);border-color:var(--theme-input--border-color--invalid) !important}:host([class*=ix-invalid]:not(.disabled):not(:disabled):not([disabled])) input:hover,:host(.ix-invalid--required:not(.disabled):not(:disabled):not([disabled])) input:hover{background-color:var(--theme-input--background--invalid--hover);border-color:var(--theme-input--border-color--invalid--hover) !important}:host([class*=ix-invalid]:not(.disabled):not(:disabled):not([disabled])) input:active,:host(.ix-invalid--required:not(.disabled):not(:disabled):not([disabled])) input:active{border-color:var(--theme-input--border-color--invalid--active) !important}:host{display:inline-block;position:relative}:host *,:host *::after,:host *::before{box-sizing:border-box}:host ::-webkit-scrollbar-button{display:none}@-moz-document url-prefix(){:host *{scrollbar-color:var(--theme-scrollbar-thumb--background) var(--theme-scrollbar-track--background);scrollbar-width:thin}}:host{}:host ::-webkit-scrollbar{width:0.5rem;height:0.5rem}:host{}:host ::-webkit-scrollbar-track{border-radius:5px;background:var(--theme-scrollbar-track--background)}:host ::-webkit-scrollbar-track:hover{background:var(--theme-scrollbar-track--background--hover)}:host{}:host ::-webkit-scrollbar-thumb{border-radius:5px;background:var(--theme-scrollbar-thumb--background)}:host{}:host ::-webkit-scrollbar-thumb:hover{background:var(--theme-scrollbar-thumb--background--hover)}:host ::-webkit-scrollbar-corner{display:none}:host input{width:100%;height:100%}:host .time-icon-hidden{display:none}:host(.readonly) input{pointer-events:none}.input-wrapper{position:relative}";

/*
 * SPDX-FileCopyrightText: 2025 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var __decorate = (undefined && undefined.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
        r = Reflect.decorate(decorators, target, key, desc);
    else
        for (var i = decorators.length - 1; i >= 0; i--)
            if (d = decorators[i])
                r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
const TimeInput = /*@__PURE__*/ proxyCustomElement(class TimeInput extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
        this.valueChange = createEvent(this, "valueChange", 6);
        this.validityStateChange = createEvent(this, "validityStateChange", 7);
        this.ixFocus = createEvent(this, "ixFocus", 7);
        this.ixBlur = createEvent(this, "ixBlur", 7);
        this.formInternals = this.attachInternals();
        /**
         * Value of the input element
         */
        this.value = '';
        /**
         * Format of time string
         * See {@link https://moment.github.io/luxon/#/formatting?id=table-of-tokens} for all available tokens.
         */
        this.format = 'TT';
        /**
         * Readonly attribute
         */
        this.readonly = false;
        /**
         * Disabled attribute
         */
        this.disabled = false;
        /**
         * I18n string for the error message when the time is not parsable
         */
        this.i18nErrorTimeUnparsable = 'Time is not valid';
        /**
         * Interval for hour selection
         */
        this.hourInterval = 1;
        /**
         * Interval for minute selection
         */
        this.minuteInterval = 1;
        /**
         * Interval for second selection
         */
        this.secondInterval = 1;
        /**
         * Interval for millisecond selection
         */
        this.millisecondInterval = 100;
        /**
         * Text of time-picker time select button
         */
        this.i18nSelectTime = 'Confirm';
        /**
         * Text for time-picker top label
         */
        this.i18nTime = 'Time';
        /**
         * Text for time-picker hour column header
         */
        this.i18nHourColumnHeader = 'hr';
        /**
         * Text for time-picker minute column header
         */
        // eslint-disable-next-line @stencil-community/decorators-style
        this.i18nMinuteColumnHeader = 'min';
        /**
         * Text for time-picker second column header
         */
        // eslint-disable-next-line @stencil-community/decorators-style
        this.i18nSecondColumnHeader = 'sec';
        /**
         * Text for time-picker millisecond column header
         */
        // eslint-disable-next-line @stencil-community/decorators-style
        this.i18nMillisecondColumnHeader = 'ms';
        /**
         * If false, pressing Enter will submit the form (if inside a form).
         * Set to true to suppress submit on Enter.
         */
        this.suppressSubmitOnEnter = false;
        /**
         * Hides the header of the picker.
         *
         * @since 4.0.0
         */
        this.hideHeader = false;
        /**
         * Text alignment within the time input. 'start' aligns the text to the start of the input, 'end' aligns the text to the end of the input.
         */
        this.textAlignment = 'start';
        this.show = false;
        this.time = null;
        this.isInputInvalid = false;
        this.isInvalid = false;
        this.isValid = false;
        this.isInfo = false;
        this.isWarning = false;
        this.focus = false;
        this.slotStartRef = makeRef();
        this.slotEndRef = makeRef();
        this.timePickerRef = makeRef();
        this.inputElementRef = makeRef();
        this.dropdownElementRef = makeRef();
        this.touched = false;
    }
    watchValuePropHandler(newValue) {
        this.onInput(newValue);
    }
    handleInputKeyDown(event) {
        handleSubmitOnEnterKeydown(event, this.suppressSubmitOnEnter, this.formInternals.form);
    }
    updateFormInternalValue(value) {
        this.formInternals.setFormValue(value);
        this.value = value;
    }
    connectedCallback() {
        this.classObserver = createClassMutationObserver(this.hostElement, () => this.checkClassList());
        this.disposableChangesAndVisibilityObservers =
            addDisposableChangesAndVisibilityObservers(this.hostElement, this.updatePaddings.bind(this));
    }
    componentWillLoad() {
        if (!this.value) {
            const now = DateTime.now();
            if (now.isValid) {
                this.value = now.toFormat(this.format);
            }
        }
        this.onInput(this.value);
        if (this.isInputInvalid) {
            this.time = null;
        }
        else {
            this.watchValue();
        }
        this.checkClassList();
        this.updateFormInternalValue(this.value);
    }
    updatePaddings() {
        adjustPaddingForStartAndEnd(this.slotStartRef.current, this.slotEndRef.current, this.inputElementRef.current);
    }
    disconnectedCallback() {
        var _a, _b;
        (_a = this.classObserver) === null || _a === void 0 ? void 0 : _a.destroy();
        (_b = this.disposableChangesAndVisibilityObservers) === null || _b === void 0 ? void 0 : _b.call(this);
    }
    watchValue() {
        this.time = this.value;
    }
    /** @internal */
    hasValidValue() {
        return Promise.resolve(!!this.value);
    }
    /** @internal */
    getAssociatedFormElement() {
        return Promise.resolve(this.formInternals.form);
    }
    async onInput(value) {
        this.value = value;
        if (!value) {
            this.isInputInvalid = false;
            this.updateFormInternalValue(value);
            this.valueChange.emit(value);
            return;
        }
        if (!this.format) {
            return;
        }
        const time = DateTime.fromFormat(value, this.format);
        if (time.isValid) {
            this.isInputInvalid = false;
        }
        else {
            this.isInputInvalid = true;
            this.invalidReason = time.invalidReason;
        }
        this.updateFormInternalValue(value);
        this.valueChange.emit(value);
    }
    onTimeIconClick(event) {
        handleIconClick(event, this.show, () => this.openDropdown(), this.inputElementRef);
    }
    async openDropdown() {
        // keep picker in sync with input
        this.time = this.value;
        return openDropdown(this.dropdownElementRef);
    }
    async closeDropdown() {
        return closeDropdown(this.dropdownElementRef);
    }
    checkClassList() {
        this.isInvalid = this.hostElement.classList.contains('ix-invalid');
    }
    renderInput() {
        return (h("div", { class: "input-wrapper" }, h(SlotStart, { slotStartRef: this.slotStartRef, onSlotChange: () => this.updatePaddings() }), h("input", { autoComplete: "off", class: {
                'is-invalid': this.isInputInvalid,
            }, style: {
                textAlign: this.textAlignment,
            }, disabled: this.disabled, readOnly: this.readonly, required: this.required, ref: this.inputElementRef, type: "text", value: this.value, placeholder: this.placeholder, name: this.name, onInput: (event) => {
                const target = event.target;
                this.onInput(target.value);
            }, onClick: (event) => {
                if (this.show) {
                    event.stopPropagation();
                    event.preventDefault();
                }
            }, onFocus: async () => {
                this.openDropdown();
                this.ixFocus.emit();
            }, onBlur: () => {
                this.ixBlur.emit();
                this.touched = true;
            }, onKeyDown: (event) => this.handleInputKeyDown(event) }), h(SlotEnd, { slotEndRef: this.slotEndRef, onSlotChange: () => this.updatePaddings() }, h("ix-icon-button", { "data-testid": "open-time-picker", class: { 'time-icon-hidden': this.disabled || this.readonly }, variant: "subtle-tertiary", icon: iconClock, onClick: (event) => this.onTimeIconClick(event), "aria-label": "Toggle time picker", "aria-expanded": this.show }))));
    }
    hookValidationLifecycle({ isInfo, isInvalid, isInvalidByRequired, isValid, isWarning, }) {
        this.isInvalid = isInvalid || isInvalidByRequired || this.isInputInvalid;
        this.isInfo = isInfo;
        this.isValid = isValid;
        this.isWarning = isWarning;
    }
    async onInputValidationChange() {
        const state = await this.getValidityState();
        this.validityStateChange.emit({
            patternMismatch: state.patternMismatch,
            invalidReason: this.invalidReason,
        });
    }
    /** @internal */
    getValidityState() {
        return Promise.resolve(createValidityState(this.isInputInvalid, !!this.required, this.value));
    }
    /**
     * Get the native input element
     */
    getNativeInputElement() {
        return this.inputElementRef.waitForCurrent();
    }
    /**
     * Focuses the input field
     */
    async focusInput() {
        return (await this.getNativeInputElement()).focus();
    }
    /**
     * Returns whether the text field has been touched.
     * @internal
     */
    isTouched() {
        return Promise.resolve(this.touched);
    }
    render() {
        var _a;
        const invalidText = getValidationText(this.isInputInvalid, this.invalidText, this.i18nErrorTimeUnparsable);
        return (h(Host, { key: 'aff9b793fae69ba2eb713ae3b1bd2222b548fb25', class: {
                disabled: this.disabled,
                readonly: this.readonly,
            } }, h("ix-field-wrapper", { key: '5e927fa4c2eeb94dba26b1f4ad1879a6a80ed823', label: this.label, helperText: this.helperText, isInvalid: this.isInvalid, invalidText: invalidText, infoText: this.infoText, isInfo: this.isInfo, isWarning: this.isWarning, warningText: this.warningText, isValid: this.isValid, validText: this.validText, showTextAsTooltip: this.showTextAsTooltip, required: this.required, controlRef: this.inputElementRef }, this.renderInput()), h("ix-dropdown", { key: '0b48b3c8574b0aafe42db74c990f1d269037bd82', "data-testid": "time-dropdown", trigger: this.inputElementRef.waitForCurrent(), ref: this.dropdownElementRef, closeBehavior: "outside", suppressOverflowBehavior: true, show: this.show, onShowChanged: (event) => {
                this.show = event.detail;
            } }, h("ix-time-picker", { key: 'b3b5536c702fc45c9fc5f323282149c2c2ccb542', ref: this.timePickerRef, format: this.format, time: (_a = this.time) !== null && _a !== void 0 ? _a : '', hourInterval: this.hourInterval, minuteInterval: this.minuteInterval, secondInterval: this.secondInterval, millisecondInterval: this.millisecondInterval, embedded: true, hideHeader: this.hideHeader, i18nConfirmTime: this.i18nSelectTime, i18nHeader: this.i18nTime, i18nHourColumnHeader: this.i18nHourColumnHeader, i18nSecondColumnHeader: this.i18nSecondColumnHeader, i18nMinuteColumnHeader: this.i18nMinuteColumnHeader, i18nMillisecondColumnHeader: this.i18nMillisecondColumnHeader, onTimeSelect: (event) => {
                this.onInput(event.detail);
                this.show = false;
            } }))));
    }
    static get formAssociated() { return true; }
    get hostElement() { return this; }
    static get watchers() { return {
        "value": ["watchValuePropHandler", "watchValue"],
        "isInputInvalid": ["onInputValidationChange"]
    }; }
    static get style() { return timeInputCss; }
}, [321, "ix-time-input", {
        "name": [513],
        "placeholder": [513],
        "value": [1537],
        "format": [1],
        "required": [4],
        "helperText": [1, "helper-text"],
        "label": [1],
        "invalidText": [513, "invalid-text"],
        "readonly": [4],
        "disabled": [4],
        "infoText": [1, "info-text"],
        "warningText": [1, "warning-text"],
        "validText": [1, "valid-text"],
        "showTextAsTooltip": [4, "show-text-as-tooltip"],
        "i18nErrorTimeUnparsable": [1, "i18n-error-time-unparsable"],
        "hourInterval": [2, "hour-interval"],
        "minuteInterval": [2, "minute-interval"],
        "secondInterval": [2, "second-interval"],
        "millisecondInterval": [2, "millisecond-interval"],
        "i18nSelectTime": [1, "i18n-select-time"],
        "i18nTime": [1, "i18n-time"],
        "i18nHourColumnHeader": [1, "i18n-hour-column-header"],
        "i18nMinuteColumnHeader": [1, "i18n-minute-column-header"],
        "i18nSecondColumnHeader": [1, "i18n-second-column-header"],
        "i18nMillisecondColumnHeader": [1, "i18n-millisecond-column-header"],
        "suppressSubmitOnEnter": [516, "suppress-submit-on-enter"],
        "hideHeader": [4, "hide-header"],
        "textAlignment": [1, "text-alignment"],
        "show": [32],
        "time": [32],
        "isInputInvalid": [32],
        "isInvalid": [32],
        "isValid": [32],
        "isInfo": [32],
        "isWarning": [32],
        "focus": [32],
        "hasValidValue": [64],
        "getAssociatedFormElement": [64],
        "getValidityState": [64],
        "getNativeInputElement": [64],
        "focusInput": [64],
        "isTouched": [64]
    }, undefined, {
        "value": ["watchValuePropHandler", "watchValue"],
        "isInputInvalid": ["onInputValidationChange"]
    }]);
__decorate([
    HookValidationLifecycle()
], TimeInput.prototype, "hookValidationLifecycle", null);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-time-input", "ix-button", "ix-date-time-card", "ix-dropdown", "ix-field-label", "ix-field-wrapper", "ix-icon-button", "ix-spinner", "ix-time-picker", "ix-tooltip", "ix-typography"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-time-input":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, TimeInput);
            }
            break;
        case "ix-button":
            if (!customElements.get(tagName)) {
                defineCustomElement$b();
            }
            break;
        case "ix-date-time-card":
            if (!customElements.get(tagName)) {
                defineCustomElement$a();
            }
            break;
        case "ix-dropdown":
            if (!customElements.get(tagName)) {
                defineCustomElement$9();
            }
            break;
        case "ix-field-label":
            if (!customElements.get(tagName)) {
                defineCustomElement$8();
            }
            break;
        case "ix-field-wrapper":
            if (!customElements.get(tagName)) {
                defineCustomElement$7();
            }
            break;
        case "ix-icon-button":
            if (!customElements.get(tagName)) {
                defineCustomElement$6();
            }
            break;
        case "ix-spinner":
            if (!customElements.get(tagName)) {
                defineCustomElement$5();
            }
            break;
        case "ix-time-picker":
            if (!customElements.get(tagName)) {
                defineCustomElement$4();
            }
            break;
        case "ix-tooltip":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "ix-typography":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
    } });
}

const IxTimeInput = TimeInput;
const defineCustomElement = defineCustomElement$1;

export { IxTimeInput, defineCustomElement };
