// @siemens/ix/components/p-B006bhbX.js
import { p as proxyCustomElement, H, h, d as Host } from './p-D-__gyet.js';

const rowCss = ":host{display:flex;flex-wrap:wrap}:host(:not(:first-of-type)){margin-block-start:var(--ix-layout-grid-row-margin, 0)}";

const Row = /*@__PURE__*/ proxyCustomElement(class Row extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
    }
    render() {
        return (h(Host, { key: 'ff07b78a26aaca76105d0cf296807512620aa6e6' }, h("slot", { key: '46cf7e7f85f4a44831b6482df9a9b667519082f0' })));
    }
    static get style() { return rowCss; }
}, [257, "ix-row"]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-row"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-row":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, Row);
            }
            break;
    } });
}

export { Row as R, defineCustomElement as d };
