// @siemens/ix/components/p-BIg1TSrD.js
import { p as proxyCustomElement, H, h, d as Host } from './p-D-__gyet.js';
import { m as makeRef } from './p-bcj7UEIC.js';
import { h as hasAnyText, H as HelperText } from './p-CrqgL2m1.js';
import { d as defineCustomElement$3 } from './p-B1p0kpQJ.js';
import { d as defineCustomElement$2 } from './p-DXFN8JXp.js';
import { d as defineCustomElement$1 } from './p-BBwXlKhE.js';

const fieldWrapperCss = ":host{display:flex;position:relative;flex-direction:column}:host .slot-wrapper{display:flex;position:relative;align-items:center;justify-content:flex-start;gap:0.25rem}:host .field-bottom,:host .field-top{display:flex;flex-direction:row;position:relative;justify-content:space-between;gap:1rem;width:-moz-min-content;width:min-content;min-width:100%}:host .field-bottom .bottom-right{margin-left:auto;margin-right:0px}:host .bottom-text{display:flex;position:relative;align-items:flex-start;justify-content:flex-start;gap:0.25rem;margin-right:0.25rem}:host .text-icon{margin:0.125rem}:host .text-icon.invalid{color:var(--theme-helper-icon--color--invalid)}:host .text-icon.info{color:var(--theme-helper-icon--color--info)}:host .text-icon.warning{color:var(--theme-helper-icon--color--warning)}:host .text-icon.valid{color:var(--theme-helper-icon--color--valid)}:host .bottom-text{margin-top:0.25rem;margin-bottom:0.25rem}";

const FieldWrapper = /*@__PURE__*/ proxyCustomElement(class FieldWrapper extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
        /**
         * Is the field component invalid
         */
        this.isInvalid = false;
        /**
         * Is the field component valid
         */
        this.isValid = false;
        /**
         * Is the field component info
         */
        this.isInfo = false;
        /**
         * Is the field component warning
         */
        this.isWarning = false;
        /**
         * Show helper, error, info, warning text as tooltip
         */
        this.showTextAsTooltip = false;
        /**
         * Show label as required
         */
        this.required = false;
        this.slotRef = makeRef();
    }
    render() {
        const textOptions = {
            invalidText: this.invalidText,
            isInvalid: this.isInvalid,
            isValid: this.isValid,
            validText: this.validText,
            isWarning: this.isWarning,
            warningText: this.warningText,
            isInfo: this.isInfo,
            infoText: this.infoText,
            helperText: this.helperText,
        };
        return (h(Host, { key: 'a27a9cefcce30e8dc30ca6a3fd72164227e86fdf' }, this.label && (h("div", { key: 'ef4a8585d24c3b0400e85b1a21bd9344e01bcb47', class: "field-top" }, h("ix-field-label", { key: '93c9f7cfbbde681956a7676721f1dd0821907e89', required: this.required, htmlFor: this.htmlForLabel, controlRef: this.controlRef, isInvalid: this.isInvalid }, this.label))), h("div", { key: '3a26cb08a91fc7c0f10de83a5b722b5af66f059f', class: {
                'slot-wrapper': true,
            }, ref: this.slotRef }, h("slot", { key: 'fe73a444a23928cee7f7899cc3bb6aef76eefa68' })), h("div", { key: 'bc5fa34c5a05e49427e63a87d70db0735a0a0b80', class: 'field-bottom' }, !this.showTextAsTooltip && h(HelperText, Object.assign({ key: '67abdfcf2448d8df422cd0ee39c162f423e5c564' }, textOptions)), h("div", { key: 'fecb02a2e534980491ea776cdc952b1aeea41286', class: "bottom-right" }, h("slot", { key: '9a958f810edf477be93394d887724c9763c57f17', name: "bottom-right" }))), this.showTextAsTooltip === true && hasAnyText(textOptions) && (h("ix-tooltip", { key: 'd3d8810548d267811e5f298fbb41f657abbae27d', for: this.slotRef.waitForCurrent(), showDelay: 500, placement: "bottom" }, h(HelperText, Object.assign({ key: '44c241959d5db7e91c00a08a843a51dfc9307256' }, textOptions))))));
    }
    get hostElement() { return this; }
    static get style() { return fieldWrapperCss; }
}, [257, "ix-field-wrapper", {
        "helperText": [1, "helper-text"],
        "label": [1],
        "invalidText": [1, "invalid-text"],
        "validText": [1, "valid-text"],
        "infoText": [1, "info-text"],
        "warningText": [1, "warning-text"],
        "isInvalid": [4, "is-invalid"],
        "isValid": [4, "is-valid"],
        "isInfo": [4, "is-info"],
        "isWarning": [4, "is-warning"],
        "showTextAsTooltip": [4, "show-text-as-tooltip"],
        "required": [4],
        "htmlForLabel": [1, "html-for-label"],
        "controlRef": [16, "control-ref"]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-field-wrapper", "ix-field-label", "ix-tooltip", "ix-typography"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-field-wrapper":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, FieldWrapper);
            }
            break;
        case "ix-field-label":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "ix-tooltip":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
        case "ix-typography":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}

export { FieldWrapper as F, defineCustomElement as d };
