// @siemens/ix/components/p-BiqdnsgA.js
import { p as proxyCustomElement, H, e as createEvent, i as forceUpdate, h, d as Host } from './p-D-__gyet.js';
import { v as iconClose } from './p-BvsadYCu.js';
import { d as defineCustomElement$4 } from './p-BEDohapR.js';
import { d as defineCustomElement$3 } from './p-1pd_0dhF.js';
import { d as defineCustomElement$2 } from './p-7E7AMiQV.js';
import { d as defineCustomElement$1 } from './p-10zMtvCF.js';

const menuSettingsCss = ".text-xs{font-family:Siemens Sans, sans-serif;font-size:0.625rem;font-weight:400;line-height:1.4em;color:var(--theme-color-std-text);-webkit-font-smoothing:antialiased;-moz-osx-font-smooting:grayscale}.text-s{font-family:Siemens Sans, sans-serif;font-size:0.75rem;font-weight:400;line-height:1.5em;color:var(--theme-color-std-text);-webkit-font-smoothing:antialiased;-moz-osx-font-smooting:grayscale}.text-caption{font-family:Siemens Sans, sans-serif;font-size:0.75rem;font-weight:700;line-height:1.5em;color:var(--theme-color-std-text);-webkit-font-smoothing:antialiased;-moz-osx-font-smooting:grayscale}.text-caption-single{font-family:Siemens Sans, sans-serif;font-size:0.75rem;font-weight:700;line-height:1em;color:var(--theme-color-std-text);-webkit-font-smoothing:antialiased;-moz-osx-font-smooting:grayscale}.text-default{font-family:Siemens Sans, sans-serif;font-size:0.875rem;font-weight:400;line-height:1.429em;color:var(--theme-color-std-text);-webkit-font-smoothing:antialiased;-moz-osx-font-smooting:grayscale}.text-default-single{font-family:Siemens Sans, sans-serif;font-size:0.875rem;font-weight:400;line-height:1.143em;color:var(--theme-color-std-text);-webkit-font-smoothing:antialiased;-moz-osx-font-smooting:grayscale}.text-default-title{font-family:Siemens Sans, sans-serif;font-size:0.875rem;font-weight:700;line-height:1.429em;color:var(--theme-color-std-text);-webkit-font-smoothing:antialiased;-moz-osx-font-smooting:grayscale}.text-default-title-single{font-family:Siemens Sans, sans-serif;font-size:0.875rem;font-weight:700;line-height:1.143em;color:var(--theme-color-std-text);-webkit-font-smoothing:antialiased;-moz-osx-font-smooting:grayscale}.text-l{font-family:Siemens Sans, sans-serif;font-size:1rem;font-weight:400;line-height:1.5em;color:var(--theme-color-std-text);-webkit-font-smoothing:antialiased;-moz-osx-font-smooting:grayscale}.text-l-single{font-family:Siemens Sans, sans-serif;font-size:1rem;font-weight:400;line-height:1.25em;color:var(--theme-color-std-text);-webkit-font-smoothing:antialiased;-moz-osx-font-smooting:grayscale}.text-l-title{font-family:Siemens Sans, sans-serif;font-size:1rem;font-weight:700;line-height:1.5em;color:var(--theme-color-std-text);-webkit-font-smoothing:antialiased;-moz-osx-font-smooting:grayscale}.text-l-title-single{font-family:Siemens Sans, sans-serif;font-size:1rem;font-weight:700;line-height:1.25em;color:var(--theme-color-std-text);-webkit-font-smoothing:antialiased;-moz-osx-font-smooting:grayscale}.text-h2{font-family:Siemens Sans, sans-serif;font-size:1.375rem;font-weight:700;line-height:1.455em;color:var(--theme-color-std-text);-webkit-font-smoothing:antialiased;-moz-osx-font-smooting:grayscale}.text-xl{font-family:Siemens Sans, sans-serif;font-size:1.375rem;font-weight:400;line-height:1.091em;color:var(--theme-color-std-text);-webkit-font-smoothing:antialiased;-moz-osx-font-smooting:grayscale}a{color:var(--theme-color-primary)}:host{display:block;background-color:var(--theme-nav-overlay--background);padding:0.75rem 1rem 1rem 2rem;flex-grow:1;position:absolute;width:100%;height:100%}:host *,:host *::after,:host *::before{box-sizing:border-box}:host ::-webkit-scrollbar-button{display:none}@-moz-document url-prefix(){:host *{scrollbar-color:var(--theme-scrollbar-thumb--background) var(--theme-scrollbar-track--background);scrollbar-width:thin}}:host{}:host ::-webkit-scrollbar{width:0.5rem;height:0.5rem}:host{}:host ::-webkit-scrollbar-track{border-radius:5px;background:var(--theme-scrollbar-track--background)}:host ::-webkit-scrollbar-track:hover{background:var(--theme-scrollbar-track--background--hover)}:host{}:host ::-webkit-scrollbar-thumb{border-radius:5px;background:var(--theme-scrollbar-thumb--background)}:host{}:host ::-webkit-scrollbar-thumb:hover{background:var(--theme-scrollbar-thumb--background--hover)}:host ::-webkit-scrollbar-corner{display:none}:host .settings-header{display:flex;justify-content:space-between;flex-direction:row;align-items:center;height:2rem;margin-bottom:0.5rem}:host .settings-header h2{color:var(--theme-nav-overlay-header--color);margin-bottom:1rem}:host ix-tabs{margin-bottom:1.5rem}:host .settings-tabs{margin-bottom:1.5rem}";

const MenuSettings = /*@__PURE__*/ proxyCustomElement(class MenuSettings extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
        this.tabChange = createEvent(this, "tabChange", 7);
        this.close = createEvent(this, "close", 7);
        /**
         * Label of first tab
         */
        this.label = 'Settings';
        /** @internal */
        this.show = false;
    }
    updateTab(newLabel, oldLabel) {
        if (newLabel !== oldLabel) {
            syncTabDisplay(this, newLabel);
        }
    }
    componentWillLoad() {
        initialize(this);
    }
    componentDidLoad() {
        forceUpdate(this.el);
    }
    render() {
        return h(MenuTabs, { key: 'bff35655d1e57ea807a1d8a6bd0879f8804e3ef4', context: this });
    }
    get el() { return this; }
    static get watchers() { return {
        "activeTabLabel": ["updateTab"]
    }; }
    static get style() { return menuSettingsCss; }
}, [257, "ix-menu-settings", {
        "activeTabLabel": [1025, "active-tab-label"],
        "label": [1],
        "show": [4],
        "items": [32]
    }, undefined, {
        "activeTabLabel": ["updateTab"]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-menu-settings", "ix-icon-button", "ix-spinner", "ix-tab-item", "ix-tabs"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-menu-settings":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, MenuSettings);
            }
            break;
        case "ix-icon-button":
            if (!customElements.get(tagName)) {
                defineCustomElement$4();
            }
            break;
        case "ix-spinner":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "ix-tab-item":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
        case "ix-tabs":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}

/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
function getItems(context) {
    return Array.from(context.el.querySelectorAll(context instanceof MenuSettings
        ? 'ix-menu-settings-item'
        : 'ix-menu-about-item'));
}
function setTab(context, label) {
    if (context.activeTabLabel === label) {
        return;
    }
    const { defaultPrevented } = context.tabChange.emit(label);
    if (defaultPrevented) {
        return;
    }
    context.activeTabLabel = label;
}
function syncTabDisplay(context, label) {
    context.items.forEach((i) => {
        i.style.display = 'none';
        if (i.label === label) {
            i.style.display = 'block';
        }
    });
}
function initialize(context) {
    context.items = getItems(context);
    if (context.items.length) {
        const selectedLabel = context.activeTabLabel || context.items[0].label;
        if (selectedLabel) {
            setTab(context, selectedLabel);
        }
    }
    context.items.forEach((item) => {
        item.addEventListener('labelChange', (e) => {
            context.items = getItems(context);
            if (e.detail.oldLabel === context.activeTabLabel) {
                context.activeTabLabel = e.detail.newLabel;
            }
        });
    });
}

/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
const getTabItems = (context) => {
    return context.items.map(({ label }) => {
        return (h("ix-tab-item", { selected: label === context.activeTabLabel, onTabClick: (e) => e.preventDefault(), onClick: () => {
                if (label) {
                    setTab(context, label);
                }
            } }, label));
    });
};
const MenuTabs = ({ context }) => {
    const selectedIndex = context.items.findIndex((item) => item.label === context.activeTabLabel);
    return (h(Host, { slot: context instanceof MenuSettings ? 'ix-menu-settings' : 'ix-menu-about', class: {
            show: context.show,
        } }, h("div", { class: context instanceof MenuSettings ? 'settings-header' : 'about-header' }, h("h2", { class: "text-h2" }, context.label), h("ix-icon-button", { variant: "tertiary", size: "24", icon: iconClose, iconColor: "color-soft-text", onClick: (e) => context.close.emit({
            name: context instanceof MenuSettings
                ? 'ix-menu-settings'
                : 'ix-menu-about',
            nativeEvent: e,
        }) })), h("ix-tabs", { selected: selectedIndex !== -1 ? selectedIndex : 0 }, getTabItems(context)), h("slot", null)));
};

export { MenuTabs as M, MenuSettings as a, defineCustomElement as d, initialize as i, syncTabDisplay as s };
