// @siemens/ix/components/p-CvD2yWeJ.js
import { p as proxyCustomElement, H, h, d as Host } from './p-D-__gyet.js';

const layoutGridCss = ":host{--ix-layout-grid-gutter:24px;display:block;flex:1 1 0%;width:100%;padding-left:calc(var(--ix-layout-grid-gutter) * 0.5);padding-right:calc(var(--ix-layout-grid-gutter) * 0.5)}:host(.no-margin){padding-left:0;padding-right:0}";

const LayoutGrid = /*@__PURE__*/ proxyCustomElement(class LayoutGrid extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
        /**
         * The grid will not have any horizontal padding
         */
        this.noMargin = false;
        /**
         * Grid gap
         */
        this.gap = '24';
        /**
         * Overwrite the default number of columns. Choose between 2 and 12 columns.
         */
        this.columns = 12;
    }
    render() {
        return (h(Host, { key: '040fa69fa09d20fdbfc6ae40f85cdf70f74b3eee', class: {
                'no-margin': this.noMargin,
            }, style: {
                '--ix-layout-grid-columns': `${this.columns}`,
                '--ix-layout-grid-gutter': `${this.gap}px`,
            } }, h("slot", { key: '50cae64ab0cc69aa2ffe55bfce8f069ea22b240b' })));
    }
    static get style() { return layoutGridCss; }
}, [257, "ix-layout-grid", {
        "noMargin": [4, "no-margin"],
        "gap": [1],
        "columns": [2]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-layout-grid"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-layout-grid":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, LayoutGrid);
            }
            break;
    } });
}

export { LayoutGrid as L, defineCustomElement as d };
