// @siemens/ix/components/p-Db0RazS2.js
import { p as proxyCustomElement, H, h, d as Host } from './p-D-__gyet.js';

const dividerCss = ":host{display:block;position:relative;width:100%;border-bottom:0.0625rem solid var(--theme-color-x-weak-bdr);margin:0.25rem 0px}";

const Divider = /*@__PURE__*/ proxyCustomElement(class Divider extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
    }
    render() {
        return h(Host, { key: 'ca814e8d386105be46954dfba6ae6f9a0c8ba4a9' });
    }
    static get style() { return dividerCss; }
}, [257, "ix-divider"]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-divider"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-divider":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, Divider);
            }
            break;
    } });
}

export { Divider as D, defineCustomElement as d };
