// @siemens/ix/components/p-N4h8LKe5.js
import { p as proxyCustomElement, H, h, d as Host } from './p-D-__gyet.js';

const cardTitleCss = ":host{display:flex;position:relative;flex-direction:row;align-items:center;margin-top:0.25rem;margin-bottom:0.25rem;gap:1rem;width:100%}:host .title-actions{display:flex;position:relative;flex-direction:row;gap:0.25rem;margin-left:auto;margin-right:0}";

const CardHeader = /*@__PURE__*/ proxyCustomElement(class CardHeader extends H {
    constructor(registerHost) {
        super();
        if (registerHost !== false) {
            this.__registerHost();
        }
        this.__attachShadow();
    }
    render() {
        return (h(Host, { key: 'a25925e3092733c276c366875802163184f79a2b' }, h("slot", { key: 'c379e97f4966141ac879c9df014f793955d1d2f3' }), h("div", { key: '5ea530ca1110748ed53b7b0724d5ec0920194628', class: "title-actions" }, h("slot", { key: '10f91158d202b321c21d07dda30ff7e59999aaa9', name: "title-actions" }))));
    }
    static get style() { return cardTitleCss; }
}, [257, "ix-card-title"]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["ix-card-title"];
    components.forEach(tagName => { switch (tagName) {
        case "ix-card-title":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, CardHeader);
            }
            break;
    } });
}

export { CardHeader as C, defineCustomElement as d };
