// @siemens/ix/dist/cjs/ix-application-sidebar.cjs.entry.js
'use strict';

var index = require('./index-C8Xo8L1k.js');
var anime_esm = require('./anime.esm-AXTM6yn_.js');
var animation = require('./animation-DtTuOXbI.js');

const applicationSidebarCss = ":host{display:block;position:relative;width:22rem;height:100%;padding:0}:host *,:host *::after,:host *::before{box-sizing:border-box}:host ::-webkit-scrollbar-button{display:none}@-moz-document url-prefix(){:host *{scrollbar-color:var(--theme-scrollbar-thumb--background) var(--theme-scrollbar-track--background);scrollbar-width:thin}}:host{}:host ::-webkit-scrollbar{width:0.5rem;height:0.5rem}:host{}:host ::-webkit-scrollbar-track{border-radius:5px;background:var(--theme-scrollbar-track--background)}:host ::-webkit-scrollbar-track:hover{background:var(--theme-scrollbar-track--background--hover)}:host{}:host ::-webkit-scrollbar-thumb{border-radius:5px;background:var(--theme-scrollbar-thumb--background)}:host{}:host ::-webkit-scrollbar-thumb:hover{background:var(--theme-scrollbar-thumb--background--hover)}:host ::-webkit-scrollbar-corner{display:none}:host(.visible){border-right:2px solid var(--theme-color-soft-bdr);padding:1.5rem 0.5rem 0}";

const ApplicationSidebar = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.visible = true;
    }
    listenToggleEvent() {
        const visibility = !this.visible;
        const animation$1 = {
            width: visibility ? ['0', '22rem'] : ['22rem', '0'],
            opacity: visibility ? [0, 1] : [1, 0],
            easing: 'easeInSine',
            duration: animation.Animation.defaultTime,
        };
        const changeVisibility = () => {
            this.visible = visibility;
        };
        if (visibility) {
            animation$1.begin = changeVisibility.bind(this);
        }
        else {
            animation$1.complete = changeVisibility.bind(this);
        }
        anime_esm.animate(this.hostElement, animation$1);
    }
    render() {
        return (index.h(index.Host, { key: '9930b3444817c8ad6c3da375484018d294897472', slot: "application-sidebar", class: {
                visible: this.visible,
            } }, this.visible ? index.h("slot", null) : null));
    }
    get hostElement() { return index.getElement(this); }
};
ApplicationSidebar.style = applicationSidebarCss;

exports.ix_application_sidebar = ApplicationSidebar;
