// @siemens/ix/dist/cjs/ix-checkbox-group.cjs.entry.js
'use strict';

var index = require('./index-C8Xo8L1k.js');
var validation = require('./validation-DTmSqO2d.js');
var makeRef = require('./make-ref-BaH6_YJk.js');

const checkboxGroupCss = ":host{display:inline-block;position:relative}:host .checkbox-container{display:flex;flex-direction:column;gap:1rem;margin:0.375rem 0;flex-wrap:wrap}:host .checkbox-container.row-layout{flex-direction:row}";

var __decorate = (undefined && undefined.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
        r = Reflect.decorate(decorators, target, key, desc);
    else
        for (var i = decorators.length - 1; i >= 0; i--)
            if (d = decorators[i])
                r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
const CheckboxGroup = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /**
         * Alignment of the checkboxes in the group
         */
        this.direction = 'column';
        /**
         * Show helper, info, warning, error and valid text as tooltip
         */
        this.showTextAsTooltip = false;
        /**
         * @internal
         */
        this.required = false;
        this.isInvalid = false;
        this.isInfo = false;
        this.isValid = false;
        this.isWarning = false;
        this.touched = false;
        this.groupRef = makeRef.makeRef();
        this.observer = new MutationObserver(() => {
            this.checkForRequiredCheckbox();
        });
    }
    get checkboxElements() {
        return Array.from(this.hostElement.querySelectorAll('ix-checkbox'));
    }
    checkForRequiredCheckbox() {
        this.required = this.checkboxElements.some((checkbox) => checkbox.required);
    }
    connectedCallback() {
        this.observer.observe(this.hostElement, {
            childList: true,
            subtree: true,
            attributes: true,
            attributeFilter: ['checked', 'required'],
        });
    }
    componentWillLoad() {
        this.checkForRequiredCheckbox();
    }
    disconnectedCallback() {
        if (this.observer) {
            this.observer.disconnect();
        }
    }
    onClassFieldUpdate({ isInvalid, isInvalidByRequired, isInfo, isValid, isWarning, }) {
        this.isInvalid = isInvalid || isInvalidByRequired;
        this.isInfo = isInfo;
        this.isValid = isValid;
        this.isWarning = isWarning;
    }
    /**
     * @internal
     */
    isTouched() {
        return Promise.resolve(this.touched);
    }
    /**
     * @internal
     */
    hasValidValue() {
        return Promise.resolve(this.checkboxElements.some((checkbox) => checkbox.checked));
    }
    render() {
        return (index.h(index.Host, { key: 'ac1e91fd23038317e1a39b612f67e91d53ba946c', ref: this.groupRef, onIxBlur: () => (this.touched = true) }, index.h("ix-field-wrapper", { key: '1278bca33da0339052756f01e8ca2de2641feffb', label: this.label, helperText: this.helperText, invalidText: this.invalidText, infoText: this.infoText, validText: this.validText, warningText: this.warningText, showTextAsTooltip: this.showTextAsTooltip, isInvalid: this.isInvalid, isInfo: this.isInfo, isValid: this.isValid, isWarning: this.isWarning, controlRef: this.groupRef }, index.h("div", { key: '7ba4ca7e379f49c6234c681a682cb08ec3d44b87', class: {
                'checkbox-container': true,
                'row-layout': this.direction === 'row',
            } }, index.h("slot", { key: 'a843f1bfd2942b4f0e9d2060646b50e97e55bca8' })))));
    }
    get hostElement() { return index.getElement(this); }
};
__decorate([
    validation.HookValidationLifecycle({
        includeChildren: true,
    })
], CheckboxGroup.prototype, "onClassFieldUpdate", null);
CheckboxGroup.style = checkboxGroupCss;

exports.ix_checkbox_group = CheckboxGroup;
