// @siemens/ix/dist/cjs/ix-content.cjs.entry.js
'use strict';

var index = require('./index-C8Xo8L1k.js');
var shadowDom = require('./shadow-dom-DT01VwcB.js');

const contentCss = ":host{display:flex;flex-direction:column;position:relative;padding:1.5rem 2rem 0.25rem 2rem;width:100%;height:100%;overflow:hidden}:host *,:host *::after,:host *::before{box-sizing:border-box}:host ::-webkit-scrollbar-button{display:none}@-moz-document url-prefix(){:host *{scrollbar-color:var(--theme-scrollbar-thumb--background) var(--theme-scrollbar-track--background);scrollbar-width:thin}}:host{}:host ::-webkit-scrollbar{width:0.5rem;height:0.5rem}:host{}:host ::-webkit-scrollbar-track{border-radius:5px;background:var(--theme-scrollbar-track--background)}:host ::-webkit-scrollbar-track:hover{background:var(--theme-scrollbar-track--background--hover)}:host{}:host ::-webkit-scrollbar-thumb{border-radius:5px;background:var(--theme-scrollbar-thumb--background)}:host{}:host ::-webkit-scrollbar-thumb:hover{background:var(--theme-scrollbar-thumb--background--hover)}:host ::-webkit-scrollbar-corner{display:none}:host .content{flex-grow:1;overflow:auto;position:relative}:host .content-header.slotted{margin-bottom:1rem}";

const Content = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.isContentHeaderSlotted = false;
    }
    get contentHeaderSlot() {
        return this.hostElement.shadowRoot.querySelector('.content-header slot');
    }
    render() {
        return (index.h(index.Host, { key: '2ca4af3b088beab36b86cb3431f1d1b59fcf8756' }, index.h("div", { key: '04f73ba733dc7fbf57ecf6fdc0afc1fec8249b8e', class: {
                'content-header': true,
                slotted: this.isContentHeaderSlotted,
            } }, index.h("slot", { key: '962e83643e1ebcd86384cae1e7a3632b6f47554e', name: "header", onSlotchange: () => {
                this.isContentHeaderSlotted = shadowDom.hasSlottedElements(this.contentHeaderSlot);
            } })), index.h("div", { key: '4315ac1e71743ba30c7b8f592997c662eed392b4', class: "content" }, index.h("slot", { key: 'c1f31f8e283bcc55011b9f2d5746d9b7c10988da' }))));
    }
    get hostElement() { return index.getElement(this); }
};
Content.style = contentCss;

exports.ix_content = Content;
