// @siemens/ix/dist/cjs/ix-link-button.cjs.entry.js
'use strict';

var index = require('./index-C8Xo8L1k.js');
var index$1 = require('./index-Dyz3z1FD.js');

const linkButtonCss = ":host{display:inline-flex;height:2rem;font-size:0.875rem;font-weight:400;min-width:2rem}:host .link-button{display:inline-flex;position:relative;width:100%;padding:0 0.25rem 0 0;align-items:center;justify-content:center;background-color:transparent;color:var(--theme-color-primary);cursor:pointer;text-decoration:none}:host .link-button .link{display:block;position:relative;width:100%;white-space:nowrap;text-overflow:ellipsis;overflow:hidden;text-decoration:underline;text-underline-offset:0.2rem}:host .link-button:not(.disabled):not(:disabled){cursor:pointer}:host .link-button:not(.disabled):not(:disabled):hover,:host .link-button:not(.disabled):not(:disabled).hover{color:var(--theme-color-dynamic--hover)}:host .link-button:not(.disabled):not(:disabled){cursor:pointer}:host .link-button:not(.disabled):not(:disabled):active,:host .link-button:not(.disabled):not(:disabled).active{color:var(--theme-color-dynamic--active)}:host .link-button.disabled{cursor:default;color:var(--theme-color-weak-text)}:host .link-button a{all:unset}:host :focus-visible{outline:1px solid var(--theme-color-focus-bdr)}";

const LinkButton = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /**
         * Disable the link button
         */
        this.disabled = false;
        /**
         * Specifies where to open the link
         *
         * https://www.w3schools.com/html/html_links.asp
         */
        this.target = '_self';
    }
    render() {
        return (index.h(index.Host, { key: '83b41ef8a3bef06a033a9a20c30a4ae1e83066e8' }, index.h("a", { key: '955dc6e492eec929e592947ec18009d744d1f973', title: this.url, tabindex: "0", class: {
                'link-button': true,
                disabled: this.disabled,
            }, href: this.disabled ? undefined : this.url, target: this.target }, index.h("ix-icon", { key: 'b0589b7cedf757244950b1595916f8c838ae603d', class: "icon", name: index$1.iconChevronRightSmall, size: "16", "aria-hidden": "true" }), index.h("div", { key: 'c3ceb43628b0a48d60bbcd6bb54a4ad493ae3de8', class: {
                link: true,
                disabled: this.disabled,
            } }, index.h("slot", { key: '2bd6071e8aad57fb13226d8452b0649367c6b750' })))));
    }
};
LinkButton.style = linkButtonCss;

exports.ix_link_button = LinkButton;
