// @siemens/ix/dist/cjs/picker-input.util-DJNAVluZ.js
'use strict';

var dropdownController = require('./dropdown-controller-BLOjtc6w.js');

/*
 * SPDX-FileCopyrightText: 2025 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
async function openDropdown(dropdownElementRef) {
    const dropdownElement = await dropdownElementRef.waitForCurrent();
    const id = dropdownElement.getAttribute('data-ix-dropdown');
    dropdownController.dropdownController.dismissAll();
    if (!id) {
        return;
    }
    const dropdown = dropdownController.dropdownController.getDropdownById(id);
    if (!dropdown) {
        return;
    }
    dropdownController.dropdownController.present(dropdown);
}
async function closeDropdown(dropdownElementRef) {
    const dropdownElement = await dropdownElementRef.waitForCurrent();
    const id = dropdownElement.getAttribute('data-ix-dropdown');
    if (!id) {
        return;
    }
    const dropdown = dropdownController.dropdownController.getDropdownById(id);
    if (!dropdown) {
        return;
    }
    dropdownController.dropdownController.dismiss(dropdown);
}
function handleIconClick(event, show, openDropdownFn, inputElementRef) {
    if (!show) {
        event.stopPropagation();
        event.preventDefault();
        openDropdownFn();
    }
    if (inputElementRef.current) {
        inputElementRef.current.focus();
    }
}
function createValidityState(isInputInvalid, required, value) {
    return {
        badInput: false,
        customError: false,
        patternMismatch: isInputInvalid,
        rangeOverflow: false,
        rangeUnderflow: false,
        stepMismatch: false,
        tooLong: false,
        tooShort: false,
        typeMismatch: false,
        valid: !isInputInvalid,
        valueMissing: !!required && !value,
    };
}

exports.closeDropdown = closeDropdown;
exports.createValidityState = createValidityState;
exports.handleIconClick = handleIconClick;
exports.openDropdown = openDropdown;
