// @siemens/ix/dist/collection/components/slider/slider.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
        r = Reflect.decorate(decorators, target, key, desc);
    else
        for (var i = decorators.length - 1; i >= 0; i--)
            if (d = decorators[i])
                r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { h, Host, } from "@stencil/core";
import { a11yHostAttributes } from "../utils/a11y";
import { OnListener } from "../utils/listener";
import { makeRef } from "../utils/make-ref";
function between(min, value, max) {
    if (value < min) {
        return min;
    }
    else if (value > max) {
        return max;
    }
    else {
        return value;
    }
}
/**
 * @slot label-start - Element will be displayed at the start of the slider
 * @slot label-end - Element will be displayed at the end of the slider
 */
export class Slider {
    constructor() {
        /**
         * Legal number intervals
         *
         * {@link https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/range#step}
         */
        this.step = 1;
        /**
         * Minimum slider value
         */
        this.min = 0;
        /**
         * Maximum slider value
         */
        this.max = 100;
        /**
         * Current value of the slider
         */
        this.value = 0;
        /**
         * Show a trace line
         */
        this.trace = false;
        /**
         * Define the start point of the trace line
         */
        this.traceReference = 0;
        /**
         * Show control as disabled
         */
        this.disabled = false;
        this.rangeInput = 0;
        this.rangeMin = 0;
        this.rangeMax = 100;
        this.rangeTraceReference = 0;
        this.showTooltip = false;
        this.thumbRef = makeRef();
        this.tooltipRef = makeRef();
    }
    get tooltip() {
        return this.tooltipRef.current;
    }
    get pseudoThumb() {
        return this.thumbRef.current;
    }
    get slider() {
        var _a;
        return (_a = this.hostElement.shadowRoot) === null || _a === void 0 ? void 0 : _a.getElementById('slider');
    }
    onShowTooltipChange() {
        var _a, _b;
        if (this.showTooltip && this.pseudoThumb) {
            (_a = this.tooltip) === null || _a === void 0 ? void 0 : _a.showTooltip(this.pseudoThumb);
            return;
        }
        (_b = this.tooltip) === null || _b === void 0 ? void 0 : _b.hideTooltip();
    }
    componentWillLoad() {
        this.a11yAttributes = a11yHostAttributes(this.hostElement, [
            'role',
            'aria-valuemin',
            'aria-valuemax',
            'aria-valuenow',
        ]);
        this.updateRangeVariables();
    }
    updateRangeVariables() {
        this.rangeInput = between(this.min, this.value, this.max);
        this.rangeTraceReference = between(this.min, this.traceReference, this.max);
        this.rangeMin = Math.min(this.min, this.max);
        this.rangeMax = Math.max(this.min, this.max);
    }
    onInput(event) {
        event.stopPropagation();
        const value = parseFloat(this.slider.value);
        if (!isNaN(value)) {
            const oldValue = this.rangeInput;
            this.rangeInput = value;
            const { defaultPrevented } = this.emitInputEvent();
            if (defaultPrevented) {
                this.rangeInput = oldValue;
                this.slider.value = oldValue.toString();
            }
        }
    }
    emitInputEvent() {
        return this.valueChange.emit(this.rangeInput);
    }
    isMarkerActive(markerValue) {
        const start = Math.min(this.traceReference, this.rangeInput);
        const end = Math.max(this.traceReference, this.rangeInput);
        const value = markerValue;
        return value >= start && value <= end;
    }
    // Listen globally on window because sometimes the event listener
    // of the DOM element input itself is not called if the release
    // click is not inside the element anymore
    onPointerUp() {
        this.showTooltip = false;
    }
    render() {
        const range = this.rangeMax - this.rangeMin;
        let traceReferenceInPercentage = (this.rangeTraceReference - this.rangeMin) / range;
        let valueInPercentage = (this.rangeInput - this.rangeMin) / range;
        const distance = valueInPercentage - traceReferenceInPercentage;
        let traceStart = traceReferenceInPercentage;
        let traceEnd = valueInPercentage;
        if (distance <= 0) {
            traceStart = valueInPercentage;
            traceEnd = traceReferenceInPercentage;
        }
        return (h(Host, { key: '03faa36448299850133ed4e7c1dbfd5f7a28a91b', class: {
                disabled: this.disabled,
                error: !!this.error,
            }, onPointerDown: () => setTimeout(() => (this.showTooltip = true)) }, h("div", { key: '686f462a50a42a4a29a20d22d3637af74a3799ae', class: "slider" }, h("div", { key: '8ef5374cd9e421bb2265b1c0b0b7f2c1f2ecf6ff', class: "track" }, h("div", { key: '969da1a4f687d3a254c3353fa0f36ee5ce61130a', ref: this.thumbRef, class: "thumb", style: {
                left: `calc(${valueInPercentage} * 100% - 8px)`,
            } }), h("div", { key: '7daba2cb0cf28c3d4b130eeebedae72827109cea', class: "ticks" }, this.marker
            ? this.marker.map((markerValue) => {
                if (markerValue > this.max || markerValue < this.min) {
                    return;
                }
                let left = (markerValue - this.rangeMin) / range;
                return (h("div", { class: {
                        tick: true,
                        'tick-active': this.isMarkerActive(markerValue) && this.trace,
                    }, style: {
                        '--tick-value': `${left}`,
                    } }));
            })
            : null)), h("input", Object.assign({ key: '2bcc70bf378b44230c3dc7223e6f24707a6a3265', id: "slider", type: "range", list: this.marker ? 'markers' : undefined, step: this.step, min: this.min, max: this.max, value: this.rangeInput, tabindex: this.disabled ? -1 : 0, onInput: (event) => this.onInput(event), style: {
                '--value': `${valueInPercentage}`,
                '--trace-reference': `${traceReferenceInPercentage}`,
                '--trace-start': `${traceStart}`,
                '--trace-end': `${traceEnd}`,
            }, class: {
                trace: this.trace && traceReferenceInPercentage !== valueInPercentage,
                'hide-trace-reference': this.trace &&
                    (this.traceReference <= this.min ||
                        this.traceReference >= this.max),
            }, onFocus: () => {
                this.showTooltip = true;
            }, onBlur: () => {
                this.showTooltip = false;
            }, role: "slider", "aria-valuenow": this.rangeInput, "aria-valuemin": this.min, "aria-valuemax": this.max }, this.a11yAttributes)), h("ix-tooltip", { key: 'c887ce7b3d18a92de4cf359fb9f8706393a82529', ref: this.tooltipRef, class: {
                'hide-tooltip': !this.showTooltip,
            }, animationFrame: true, for: this.thumbRef.waitForCurrent() }, this.rangeInput)), h("div", { key: '8de8c23aa94fc924842631eb4d312efc5ee7cc6f', class: "label" }, h("div", { key: '402d7e1bd4bdd648e361b88f211d7d80e5c907b3', class: "label-start" }, h("slot", { key: 'f9a3a0ecc744b3ab91dda8d425309ce78c57e1f1', name: "label-start" })), h("div", { key: '188c06d13b00c0124603b6a0513cfa24e5467b62', class: "label-end" }, h("slot", { key: 'cc4e9bd389f10ab7e293a80844012877070b2750', name: "label-end" }))), this.error ? (h("ix-typography", { class: 'label-error', textColor: "alarm" }, this.error)) : null));
    }
    static get is() { return "ix-slider"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["slider.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["slider.css"]
        };
    }
    static get properties() {
        return {
            "step": {
                "type": "number",
                "attribute": "step",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Legal number intervals\n\n{@link https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/range#step}"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "1"
            },
            "min": {
                "type": "number",
                "attribute": "min",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Minimum slider value"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "0"
            },
            "max": {
                "type": "number",
                "attribute": "max",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Maximum slider value"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "100"
            },
            "value": {
                "type": "number",
                "attribute": "value",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Current value of the slider"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "0"
            },
            "marker": {
                "type": "unknown",
                "attribute": "marker",
                "mutable": false,
                "complexType": {
                    "original": "SliderMarker",
                    "resolved": "number[] | undefined",
                    "references": {
                        "SliderMarker": {
                            "location": "import",
                            "path": "./slider.types",
                            "id": "src/components/slider/slider.types.ts::SliderMarker"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Define tick marker on the slider. Marker has to be within slider min/max"
                },
                "getter": false,
                "setter": false
            },
            "trace": {
                "type": "boolean",
                "attribute": "trace",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show a trace line"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "traceReference": {
                "type": "number",
                "attribute": "trace-reference",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Define the start point of the trace line"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "0"
            },
            "disabled": {
                "type": "boolean",
                "attribute": "disabled",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show control as disabled"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "error": {
                "type": "any",
                "attribute": "error",
                "mutable": false,
                "complexType": {
                    "original": "boolean | string",
                    "resolved": "boolean | string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Show error state and message"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            }
        };
    }
    static get states() {
        return {
            "rangeInput": {},
            "rangeMin": {},
            "rangeMax": {},
            "rangeTraceReference": {},
            "showTooltip": {}
        };
    }
    static get events() {
        return [{
                "method": "valueChange",
                "name": "valueChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "hostElement"; }
    static get watchers() {
        return [{
                "propName": "showTooltip",
                "methodName": "onShowTooltipChange"
            }, {
                "propName": "value",
                "methodName": "updateRangeVariables"
            }, {
                "propName": "max",
                "methodName": "updateRangeVariables"
            }, {
                "propName": "min",
                "methodName": "updateRangeVariables"
            }, {
                "propName": "traceReference",
                "methodName": "updateRangeVariables"
            }];
    }
}
__decorate([
    OnListener('pointerup', (self) => self.showTooltip)
], Slider.prototype, "onPointerUp", null);
