// @siemens/ix/dist/collection/components/pagination/pagination.js
/*
 * SPDX-FileCopyrightText: 2022 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host, } from "@stencil/core";
import { BaseButton } from "../button/base-button";
import { a11yBoolean } from "../utils/a11y";
import { iconChevronLeftSmall, iconChevronRightSmall, } from "@siemens/ix-icons/icons";
export class Pagination {
    constructor() {
        this.baseButtonConfig = {
            variant: 'subtle-tertiary',
            iconOnly: true,
            iconOval: false,
            disabled: false,
            icon: '',
            loading: false,
            selected: false,
            type: 'button',
        };
        this.maxCountPages = 7;
        /**
         * Advanced mode
         */
        this.advanced = false;
        /**
         * Number of items shown at once.
         * Can only be changed in advaced mode.
         */
        this.itemCount = 15;
        /**
         * Hide item count in advanced mode
         */
        this.hideItemCount = false;
        /**
         * Total number of pages
         */
        this.count = 0;
        /**
         * Zero based index of currently selected page
         */
        this.selectedPage = 0;
        /**
         * i18n label for 'Page'
         */
        this.i18nPage = 'Page';
        /**
         * i18n label for 'of'
         */
        this.i18nOf = 'of';
        /**
         * i18n label for 'Items'
         */
        this.i18nItems = 'Items';
        /**
         * ARIA label for the page selection input
         * Will be set as aria-label on the nested HTML input element
         *
         * @since 4.1.0
         */
        this.ariaLabelPageSelection = 'Page selection input';
    }
    selectPage(index) {
        const oldIndex = this.selectedPage;
        if (index < 0) {
            this.selectedPage = 0;
        }
        else if (index > this.count - 1) {
            this.selectedPage = this.count - 1;
        }
        else {
            this.selectedPage = index;
        }
        const { defaultPrevented } = this.pageSelected.emit(this.selectedPage);
        if (defaultPrevented) {
            this.selectedPage = oldIndex;
        }
    }
    increase() {
        if (this.selectedPage === this.count - 1) {
            return;
        }
        this.selectPage(this.selectedPage + 1);
    }
    decrease() {
        if (this.selectedPage === 0) {
            return;
        }
        this.selectPage(this.selectedPage - 1);
    }
    getPageButton(index) {
        const baseButtonProps = Object.assign(Object.assign({}, this.baseButtonConfig), { onClick: () => this.selectPage(index), selected: this.selectedPage === index, ariaAttributes: {
                'aria-pressed': a11yBoolean(this.selectedPage === index),
            } });
        return h(BaseButton, Object.assign({}, baseButtonProps), index + 1);
    }
    renderPageButtons() {
        const pagesBeforeOverflow = Math.floor(this.maxCountPages / 2);
        const hasOverflow = this.count > this.maxCountPages;
        const hasOverflowStart = hasOverflow && this.selectedPage > pagesBeforeOverflow;
        const hasOverflowEnd = hasOverflow && this.selectedPage < this.count - pagesBeforeOverflow - 1;
        const pageButtons = [];
        let start = 0;
        let end = Math.min(this.count, this.maxCountPages);
        let pageCount = Math.floor((this.maxCountPages - 4) / 2);
        if (hasOverflowStart) {
            const baseButtonProps = Object.assign(Object.assign({}, this.baseButtonConfig), { onClick: () => {
                    if (hasOverflowEnd) {
                        this.selectPage(this.selectedPage - Math.max(0, 2 * pageCount + 1));
                    }
                    else {
                        this.selectPage(this.count - this.maxCountPages);
                    }
                } });
            pageButtons.push(this.getPageButton(0));
            pageButtons.push(h(BaseButton, Object.assign({}, baseButtonProps), "..."));
            if (hasOverflowEnd) {
                start = this.count - this.maxCountPages + 2;
            }
            else {
                start = this.count - this.maxCountPages + 2;
                end = this.count;
            }
        }
        if (hasOverflowEnd) {
            if (hasOverflowStart) {
                start = this.selectedPage - pageCount;
                end = this.selectedPage + pageCount + 1;
            }
            else {
                end = this.maxCountPages - 2;
            }
        }
        for (let i = start; i < end; i++) {
            pageButtons.push(this.getPageButton(i));
        }
        if (hasOverflowEnd) {
            const baseButtonProps = Object.assign(Object.assign({}, this.baseButtonConfig), { onClick: () => {
                    if (hasOverflowStart) {
                        this.selectPage(this.selectedPage + Math.max(0, 2 * pageCount + 1));
                    }
                    else {
                        this.selectPage(this.maxCountPages - 1);
                    }
                } });
            pageButtons.push(h(BaseButton, Object.assign({}, baseButtonProps), "..."));
            pageButtons.push(this.getPageButton(this.count - 1));
        }
        return h("span", { class: "page-buttons" }, pageButtons);
    }
    handlePageInput(inputValue) {
        const value = Number.parseInt(inputValue, 10);
        if (!Number.isNaN(value)) {
            const clampedValue = Math.max(1, Math.min(value, this.count));
            this.selectPage(clampedValue - 1);
        }
    }
    render() {
        return (h(Host, { key: 'e9b053fdf801e514af1abef2e163a903f7d5f976' }, h("ix-icon-button", { key: 'a880a444f946f7d228a746b8d45d7328bdc26295', disabled: !this.count || this.selectedPage === 0, variant: "subtle-tertiary", icon: iconChevronLeftSmall, onClick: () => this.decrease(), "aria-label": this.ariaLabelChevronLeftIconButton }), this.advanced ? (h("div", { class: "advanced-pagination" }, h("ix-typography", { format: "body" }, this.i18nPage), h("input", { "aria-label": this.ariaLabelPageSelection, class: "ix-form-control page-selection", type: "number", min: "1", max: this.count, value: this.selectedPage + 1, onChange: (event) => {
                const inputElement = event.target;
                this.handlePageInput(inputElement.value);
            }, onKeyDown: (event) => {
                if (['e', 'E', '+', '-', '.'].includes(event.key))
                    event.preventDefault();
                if (event.key === 'Enter') {
                    const inputElement = event.target;
                    this.handlePageInput(inputElement.value);
                    inputElement.blur();
                }
            }, onBlur: (e) => {
                const inputElement = e.target;
                inputElement.value = (this.selectedPage + 1).toString();
            } }), h("span", { class: "total-count" }, h("ix-typography", { format: "body" }, this.i18nOf, " ", this.count)))) : (h("span", { class: "basic-pagination" }, this.renderPageButtons(), " ")), h("ix-icon-button", { key: 'dd75eb48382320194812f9eb4c644b7d37f0c84a', disabled: !this.count || this.selectedPage === this.count - 1, variant: "subtle-tertiary", icon: iconChevronRightSmall, onClick: () => this.increase(), "aria-label": this.ariaLabelChevronRightIconButton }), this.advanced && !this.hideItemCount && (h("span", { key: '018574c07720618970be857f7536fdf9b71e1542', class: "item-count" }, h("ix-typography", { key: '343a05e008f258ab9a56045cac1a8785c9855169', format: "body" }, this.i18nItems), h("ix-select", { key: '2d8a5efe84ea53ea85b9f4c75f1c84800c858846', hideListHeader: true, i18nPlaceholder: "", i18nSelectListHeader: "", value: `${this.itemCount}`, onValueChange: (e) => {
                const count = Number.parseInt(Array.isArray(e.detail) ? e.detail[0] : e.detail);
                this.itemCountChanged.emit(count);
            } }, h("ix-select-item", { key: '9106274e93a00e5cb62f7cd19082f43e2920076b', label: "10", value: "10" }), h("ix-select-item", { key: '452779779ea63e45d408d60b5faed2d30ea6bcc9', label: "15", value: "15" }), h("ix-select-item", { key: '0658cd6473b508ec6c01c83d99bce4bfa8a54940', label: "20", value: "20" }), h("ix-select-item", { key: '316b1e9a882ae3a516f8a3ff3997ab0fe95c20de', label: "40", value: "40" }), h("ix-select-item", { key: 'a5dbea62134178911100a1fb369d258b0ee74b8d', label: "100", value: "100" }))))));
    }
    static get is() { return "ix-pagination"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["pagination.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["pagination.css"]
        };
    }
    static get properties() {
        return {
            "advanced": {
                "type": "boolean",
                "attribute": "advanced",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Advanced mode"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "itemCount": {
                "type": "number",
                "attribute": "item-count",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Number of items shown at once.\nCan only be changed in advaced mode."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "15"
            },
            "hideItemCount": {
                "type": "boolean",
                "attribute": "hide-item-count",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Hide item count in advanced mode"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "count": {
                "type": "number",
                "attribute": "count",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Total number of pages"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "0"
            },
            "selectedPage": {
                "type": "number",
                "attribute": "selected-page",
                "mutable": true,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Zero based index of currently selected page"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "0"
            },
            "i18nPage": {
                "type": "string",
                "attribute": "i18n-page",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "i18n label for 'Page'"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'Page'"
            },
            "i18nOf": {
                "type": "string",
                "attribute": "i18n-of",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "i18n label for 'of'"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'of'"
            },
            "i18nItems": {
                "type": "string",
                "attribute": "i18n-items",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "i18n label for 'Items'"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'Items'"
            },
            "ariaLabelChevronLeftIconButton": {
                "type": "string",
                "attribute": "aria-label-chevron-left-icon-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.2.0"
                        }],
                    "text": "ARIA label for the chevron left icon button\nWill be set as aria-label on the nested HTML button element"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "ariaLabelChevronRightIconButton": {
                "type": "string",
                "attribute": "aria-label-chevron-right-icon-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.2.0"
                        }],
                    "text": "ARIA label for the chevron right icon button\nWill be set as aria-label on the nested HTML button element"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "ariaLabelPageSelection": {
                "type": "string",
                "attribute": "aria-label-page-selection",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.1.0"
                        }],
                    "text": "ARIA label for the page selection input\nWill be set as aria-label on the nested HTML input element"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'Page selection input'"
            }
        };
    }
    static get events() {
        return [{
                "method": "pageSelected",
                "name": "pageSelected",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Page selection event"
                },
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                }
            }, {
                "method": "itemCountChanged",
                "name": "itemCountChanged",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Item count change event"
                },
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "hostElement"; }
}
