// @siemens/ix/dist/collection/components/breadcrumb/breadcrumb.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { forceUpdate, h, Host, } from "@stencil/core";
import { a11yBoolean, a11yHostAttributes } from "../utils/a11y";
import { createMutationObserver } from "../utils/mutation-observer";
import { makeRef } from "../utils/make-ref";
let sequenceId = 0;
const createId = (prefix = 'breadcrumb-') => {
    return `${prefix}-${sequenceId++}`;
};
export class Breadcrumb {
    constructor() {
        /**
         * Excess items will get hidden inside of dropdown
         */
        this.visibleItemCount = 9;
        /**
         * Items will be accessible through a dropdown
         */
        this.nextItems = [];
        /**
         * Ghost breadcrumbs will not show solid backgrounds on individual crumbs unless there is a mouse event (e.g. hover)
         */
        this.subtle = false;
        /**
         * Accessibility label for the dropdown button (ellipsis icon) used to access the dropdown list
         * with conditionally hidden previous items
         */
        this.ariaLabelPreviousButton = 'previous';
        this.previousButtonRef = makeRef();
        this.nextButtonRef = makeRef();
        this.items = [];
        this.isPreviousDropdownExpanded = false;
        this.previousButtonId = createId();
        this.previousDropdownId = createId();
    }
    onNextItemsChange() {
        this.onChildMutation();
    }
    onItemClick(item) {
        this.itemClick.emit(item);
    }
    componentDidLoad() {
        this.mutationObserver = createMutationObserver(() => this.onChildMutation());
        this.mutationObserver.observe(this.hostElement, {
            subtree: true,
            childList: true,
        });
    }
    componentWillLoad() {
        this.onChildMutation();
    }
    disconnectedCallback() {
        var _a;
        (_a = this.mutationObserver) === null || _a === void 0 ? void 0 : _a.disconnect();
    }
    async onChildMutation() {
        const updatedItems = this.getItems();
        updatedItems.forEach((bc, index) => {
            const shouldShowDropdown = this.nextItems.length !== 0 && updatedItems.length - 1 === index;
            bc.subtle = this.subtle;
            bc.hideChevron = updatedItems.length - 1 === index && !shouldShowDropdown;
            bc.isDropdownTrigger = shouldShowDropdown;
            if (shouldShowDropdown) {
                this.nextButtonRef(bc);
            }
            if (updatedItems.length < this.visibleItemCount) {
                return;
            }
            bc.invisible = index < updatedItems.length - this.visibleItemCount;
        });
        this.items = updatedItems;
    }
    getItems() {
        return Array.from(this.hostElement.querySelectorAll('ix-breadcrumb-item'));
    }
    render() {
        var _a, _b, _c, _d;
        const a11y = a11yHostAttributes(this.hostElement);
        return (h(Host, { key: 'c5457acd786da77d474f83613a2c12910a193779' }, h("ix-dropdown", { key: '3f1e0ae1e4ac1d192ff7c0d5619329cfbcdf8a62', id: this.previousDropdownId, "aria-label": this.ariaLabelPreviousButton, trigger: ((_a = this.items) === null || _a === void 0 ? void 0 : _a.length) > this.visibleItemCount
                ? this.previousButtonRef.waitForCurrent()
                : undefined, onShowChanged: ({ detail }) => {
                this.isPreviousDropdownExpanded = detail;
                const previousButton = this.hostElement.shadowRoot.getElementById(this.previousButtonId);
                // Need to force update previous button to change `aria-expanded`
                if (previousButton) {
                    forceUpdate(previousButton);
                }
            } }, this.items
            .slice(0, this.items.length - this.visibleItemCount)
            .map((item) => {
            var _a;
            const label = (_a = item.label) !== null && _a !== void 0 ? _a : item.innerText;
            return (h("ix-dropdown-item", { label: label, onClick: () => {
                    this.onItemClick(label);
                }, onItemClick: (event) => event.stopPropagation() }));
        })), ((_b = this.items) === null || _b === void 0 ? void 0 : _b.length) > this.visibleItemCount ? (h("ix-breadcrumb-item", { id: this.previousButtonId, ref: this.previousButtonRef, label: "...", tabIndex: 1, onItemClick: (event) => event.stopPropagation(), "aria-describedby": this.previousDropdownId, "aria-controls": this.previousDropdownId, "aria-expanded": a11yBoolean(this.isPreviousDropdownExpanded), class: 'previous-button' })) : null, h("nav", { key: '11b4a0c0cbad59437fccee34a73cd4e89de39a1c', class: "crumb-items", "aria-label": (_c = a11y['aria-label']) !== null && _c !== void 0 ? _c : 'breadcrumbs' }, h("ol", { key: 'ad9d8c27e05b76f334f7cf6f79038d339e8d538c' }, h("slot", { key: 'd84c2c0028e22cbfd126b867177ba9b4271e8c24' }))), h("ix-dropdown", { key: '6ba681eb8f116ec287961a7e873b90d381baefdb', trigger: this.nextButtonRef.waitForCurrent() }, (_d = this.nextItems) === null || _d === void 0 ? void 0 : _d.map((item) => (h("ix-dropdown-item", { label: item, onClick: (e) => {
                this.nextClick.emit({
                    event: e,
                    item,
                });
            }, onItemClick: (event) => event.stopPropagation() }))))));
    }
    static get is() { return "ix-breadcrumb"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["breadcrumb.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["breadcrumb.css"]
        };
    }
    static get properties() {
        return {
            "visibleItemCount": {
                "type": "number",
                "attribute": "visible-item-count",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Excess items will get hidden inside of dropdown"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "9"
            },
            "nextItems": {
                "type": "unknown",
                "attribute": "next-items",
                "mutable": false,
                "complexType": {
                    "original": "string[]",
                    "resolved": "string[]",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Items will be accessible through a dropdown"
                },
                "getter": false,
                "setter": false,
                "defaultValue": "[]"
            },
            "subtle": {
                "type": "boolean",
                "attribute": "subtle",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Ghost breadcrumbs will not show solid backgrounds on individual crumbs unless there is a mouse event (e.g. hover)"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "ariaLabelPreviousButton": {
                "type": "string",
                "attribute": "aria-label-previous-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Accessibility label for the dropdown button (ellipsis icon) used to access the dropdown list\nwith conditionally hidden previous items"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'previous'"
            }
        };
    }
    static get states() {
        return {
            "items": {},
            "isPreviousDropdownExpanded": {}
        };
    }
    static get events() {
        return [{
                "method": "itemClick",
                "name": "itemClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Crumb item clicked event"
                },
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                }
            }, {
                "method": "nextClick",
                "name": "nextClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Next item clicked event"
                },
                "complexType": {
                    "original": "{ event: UIEvent; item: string }",
                    "resolved": "{ event: UIEvent; item: string; }",
                    "references": {
                        "UIEvent": {
                            "location": "global",
                            "id": "global::UIEvent"
                        }
                    }
                }
            }];
    }
    static get elementRef() { return "hostElement"; }
    static get watchers() {
        return [{
                "propName": "nextItems",
                "methodName": "onNextItemsChange"
            }];
    }
}
