// @siemens/ix/dist/collection/components/upload/upload.js
/*
 * SPDX-FileCopyrightText: 2024 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s)
        if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { h, Host, } from "@stencil/core";
import { UploadFileState } from "./upload-file-state";
import { iconError, iconSuccess } from "@siemens/ix-icons/icons";
import { a11yHostAttributes } from "../utils/a11y";
export class Upload {
    get inputElement() {
        return this.hostElement.shadowRoot.querySelector('#upload-browser');
    }
    constructor() {
        /**
         * If multiple is true the user can drop or select multiple files
         */
        this.multiple = false;
        /**
         * Whether the text should wrap to more than one line
         */
        this.multiline = false;
        /**
         * Disable all input events
         */
        this.disabled = false;
        /**
         * After a file is uploaded you can set the upload component to a defined state
         */
        this.state = UploadFileState.SELECT_FILE;
        /**
         * Will be used by state = UploadFileState.SELECT_FILE
         */
        this.selectFileText = '+ Drag files here or…';
        /**
         * Will be used by state = UploadFileState.LOADING
         */
        this.loadingText = 'Checking files…';
        /**
         * Will be used by state = UploadFileState.UPLOAD_FAILED
         */
        this.uploadFailedText = 'Upload failed. Please try again.';
        /**
         * Will be used by state = UploadFileState.UPLOAD_SUCCESSED
         */
        this.uploadSuccessText = 'Upload successful';
        /**
         * Label for upload file button
         */
        this.i18nUploadFile = 'Upload file…';
        /**
         * Text for disabled state
         */
        this.i18nUploadDisabled = 'File upload currently not possible.';
        this.isFileOver = false;
        this.a11y = {};
    }
    componentWillLoad() {
        this.a11y = a11yHostAttributes(this.hostElement);
    }
    fileDropped(evt) {
        evt.preventDefault();
        if (this.disabled) {
            return;
        }
        if (!evt.dataTransfer) {
            return;
        }
        const file = evt.dataTransfer.files;
        this.isFileOver = false;
        this.filesToUpload = this.convertToFileArray(file);
        this.filesChanged.emit(this.filesToUpload);
    }
    fileOver(event) {
        if (!event.dataTransfer) {
            return;
        }
        if (this.state !== UploadFileState.LOADING) {
            event.preventDefault();
            event.dataTransfer.dropEffect = 'move';
        }
        if (!this.multiple && event.dataTransfer.items.length > 1) {
            event.preventDefault();
            event.stopPropagation();
            event.dataTransfer.effectAllowed = 'none';
            event.dataTransfer.dropEffect = 'none';
        }
        else {
            this.isFileOver = true;
        }
    }
    fileLeave() {
        this.isFileOver = false;
    }
    fileChangeEvent(event) {
        if (this.disabled) {
            return;
        }
        if (!event.target) {
            return;
        }
        this.filesToUpload = this.convertToFileArray(event.target.files);
        this.filesChanged.emit(this.filesToUpload);
        // Workaround for bug in native input element, that prevents the user from uploading
        // a file with the same name as the most recent one, but with changed content.
        this.inputElement.type = '';
        this.inputElement.type = 'file';
    }
    convertToFileArray(filesFromEvent) {
        let files = [];
        if (!filesFromEvent) {
            return [];
        }
        if (filesFromEvent instanceof FileList) {
            files = Array.from(filesFromEvent);
        }
        else {
            files = [filesFromEvent];
        }
        return files;
    }
    renderUploadState() {
        if (this.disabled) {
            return (h("span", { class: "state" }, h("span", { class: "upload-text" }, this.i18nUploadDisabled)));
        }
        switch (this.state) {
            case UploadFileState.SELECT_FILE:
                return (h("span", { class: "state" }, h("span", { class: "upload-text" }, this.selectFileText)));
            case UploadFileState.LOADING:
                return (h("span", { class: "state" }, h("ix-spinner", { variant: "primary" }), h("span", { class: "upload-text" }, this.loadingText)));
            case UploadFileState.UPLOAD_FAILED:
                return (h("span", { class: "state" }, h("ix-icon", { name: iconError, class: "icon-error" }), h("span", { class: "upload-text" }, this.uploadFailedText)));
            case UploadFileState.UPLOAD_SUCCESSED:
                return (h("span", { class: "state" }, h("ix-icon", { name: iconSuccess, class: "icon-success" }), h("span", { class: "upload-text" }, this.uploadSuccessText)));
            default:
                return '';
        }
    }
    /**
     * Set files
     * @param obj
     */
    async setFilesToUpload(obj) {
        this.filesToUpload = obj;
    }
    render() {
        const disabled = this.disabled || this.state === UploadFileState.LOADING;
        const _a = this.a11y, { 'aria-label': ariaLabel = 'Upload files' } = _a, a11y = __rest(_a, ['aria-label']);
        return (h(Host, Object.assign({ key: '8b8c68a23b429ca57854751c8a1c9358f78cbc9f' }, a11y, { "aria-disabled": disabled }), h("div", { key: '1bb186087ab0d5c545a3f183ce3f6de033a837ad', class: {
                'file-upload-area': true,
                'file-over': this.state !== UploadFileState.LOADING && this.isFileOver,
                checking: this.state === UploadFileState.LOADING,
                disabled: this.disabled,
                multiline: this.multiline,
            }, onDrop: (e) => {
                if (this.state !== UploadFileState.LOADING) {
                    this.fileDropped(e);
                }
            }, onDragOver: (e) => this.fileOver(e), onDragLeave: () => this.fileLeave(), draggable: !this.disabled }, this.renderUploadState(), h("div", { key: '52209f325ad967197e48f651cc6f15365be9562f' }, h("input", { key: '91c11214c2f169a9dfbeaaac88a7d928ab389455', "aria-label": ariaLabel, "aria-disabled": disabled, multiple: this.multiple, type: "file", class: "upload-browser", id: "upload-browser", tabindex: "-1", onChange: (e) => {
                this.fileChangeEvent(e);
            }, accept: this.accept, disabled: disabled }), h("ix-button", { key: '807022699c0ba9ac3d81557e88cf1ae498a36cca', variant: "secondary", "aria-disabled": disabled, onClick: () => this.inputElement.click(), disabled: disabled }, this.i18nUploadFile)))));
    }
    static get is() { return "ix-upload"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["upload.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["upload.css"]
        };
    }
    static get properties() {
        return {
            "accept": {
                "type": "string",
                "attribute": "accept",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The accept attribute specifies the types of files that the server accepts (that can be submitted through a file upload).\nSee {@link https://www.w3schools.com/tags/att_input_accept.asp}"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "multiple": {
                "type": "boolean",
                "attribute": "multiple",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "If multiple is true the user can drop or select multiple files"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "multiline": {
                "type": "boolean",
                "attribute": "multiline",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Whether the text should wrap to more than one line"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "disabled": {
                "type": "boolean",
                "attribute": "disabled",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Disable all input events"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "state": {
                "type": "string",
                "attribute": "state",
                "mutable": false,
                "complexType": {
                    "original": "UploadFileState",
                    "resolved": "UploadFileState.LOADING | UploadFileState.SELECT_FILE | UploadFileState.UPLOAD_FAILED | UploadFileState.UPLOAD_SUCCESSED",
                    "references": {
                        "UploadFileState": {
                            "location": "import",
                            "path": "./upload-file-state",
                            "id": "src/components/upload/upload-file-state.ts::UploadFileState"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "After a file is uploaded you can set the upload component to a defined state"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "UploadFileState.SELECT_FILE"
            },
            "selectFileText": {
                "type": "string",
                "attribute": "select-file-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Will be used by state = UploadFileState.SELECT_FILE"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'+ Drag files here or\u2026'"
            },
            "loadingText": {
                "type": "string",
                "attribute": "loading-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Will be used by state = UploadFileState.LOADING"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'Checking files\u2026'"
            },
            "uploadFailedText": {
                "type": "string",
                "attribute": "upload-failed-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Will be used by state = UploadFileState.UPLOAD_FAILED"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'Upload failed. Please try again.'"
            },
            "uploadSuccessText": {
                "type": "string",
                "attribute": "upload-success-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Will be used by state = UploadFileState.UPLOAD_SUCCESSED"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'Upload successful'"
            },
            "i18nUploadFile": {
                "type": "string",
                "attribute": "i18n-upload-file",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Label for upload file button"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'Upload file\u2026'"
            },
            "i18nUploadDisabled": {
                "type": "string",
                "attribute": "i18n-upload-disabled",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Text for disabled state"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'File upload currently not possible.'"
            }
        };
    }
    static get states() {
        return {
            "isFileOver": {}
        };
    }
    static get events() {
        return [{
                "method": "filesChanged",
                "name": "filesChanged",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "You get an array of Files after drop-action or browse action is finished"
                },
                "complexType": {
                    "original": "Array<File>",
                    "resolved": "File[]",
                    "references": {
                        "Array": {
                            "location": "global",
                            "id": "global::Array"
                        },
                        "File": {
                            "location": "global",
                            "id": "global::File"
                        }
                    }
                }
            }];
    }
    static get methods() {
        return {
            "setFilesToUpload": {
                "complexType": {
                    "signature": "(obj: any) => Promise<void>",
                    "parameters": [{
                            "name": "obj",
                            "type": "any",
                            "docs": ""
                        }],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Set files",
                    "tags": [{
                            "name": "param",
                            "text": "obj"
                        }]
                }
            }
        };
    }
    static get elementRef() { return "hostElement"; }
}
