// @siemens/ix/dist/collection/components/col/col.js
/*
 * SPDX-FileCopyrightText: 2024 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { forceUpdate, h, Host } from "@stencil/core";
import { matchBreakpoint } from "../utils/breakpoints";
export class Col {
    onResize() {
        forceUpdate(this);
    }
    getSize(breakpoint) {
        if (breakpoint === '') {
            return this.size;
        }
        if (breakpoint === 'sm') {
            return this.sizeSm;
        }
        if (breakpoint === 'md') {
            return this.sizeMd;
        }
        if (breakpoint === 'lg') {
            return this.sizeLg;
        }
    }
    getColumnSize() {
        let size;
        Col.Breakpoints.forEach((breakpoint) => {
            const isMediaQueryActive = breakpoint !== '' ? matchBreakpoint(breakpoint) : true;
            if (!isMediaQueryActive) {
                return;
            }
            const currentSize = this.getSize(breakpoint);
            if (currentSize) {
                size = currentSize;
            }
        });
        return size;
    }
    getColumnSizeStyling() {
        const size = this.getColumnSize();
        if (!size) {
            return;
        }
        if (size === 'auto') {
            return {
                flex: '0 0 auto',
                width: 'auto',
                'max-width': 'auto',
            };
        }
        const colSize = `calc(calc(${size} / var(--ix-layout-grid-columns)) * 100%)`;
        return {
            flex: `0 0 ${colSize}`,
            width: `${colSize}`,
            'max-width': `${colSize}`,
        };
    }
    render() {
        return (h(Host, { key: '552bf87dd4905b2cfc67dbda7c33e1edf89a9cf8', style: Object.assign({}, this.getColumnSizeStyling()) }, h("slot", { key: '293331e4b2de069f550f0d47e1525228e49fc159' })));
    }
    static get is() { return "ix-col"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["col.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["col.css"]
        };
    }
    static get properties() {
        return {
            "size": {
                "type": "string",
                "attribute": "size",
                "mutable": false,
                "complexType": {
                    "original": "ColumnSize",
                    "resolved": "\"1\" | \"10\" | \"11\" | \"12\" | \"2\" | \"3\" | \"4\" | \"5\" | \"6\" | \"7\" | \"8\" | \"9\" | \"auto\" | undefined",
                    "references": {
                        "ColumnSize": {
                            "location": "import",
                            "path": "./col.types",
                            "id": "src/components/col/col.types.ts::ColumnSize"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Size of the column"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "sizeSm": {
                "type": "string",
                "attribute": "size-sm",
                "mutable": false,
                "complexType": {
                    "original": "ColumnSize",
                    "resolved": "\"1\" | \"10\" | \"11\" | \"12\" | \"2\" | \"3\" | \"4\" | \"5\" | \"6\" | \"7\" | \"8\" | \"9\" | \"auto\" | undefined",
                    "references": {
                        "ColumnSize": {
                            "location": "import",
                            "path": "./col.types",
                            "id": "src/components/col/col.types.ts::ColumnSize"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Size of the column for sm screens"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "sizeMd": {
                "type": "string",
                "attribute": "size-md",
                "mutable": false,
                "complexType": {
                    "original": "ColumnSize",
                    "resolved": "\"1\" | \"10\" | \"11\" | \"12\" | \"2\" | \"3\" | \"4\" | \"5\" | \"6\" | \"7\" | \"8\" | \"9\" | \"auto\" | undefined",
                    "references": {
                        "ColumnSize": {
                            "location": "import",
                            "path": "./col.types",
                            "id": "src/components/col/col.types.ts::ColumnSize"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Size of the column for md screens"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "sizeLg": {
                "type": "string",
                "attribute": "size-lg",
                "mutable": false,
                "complexType": {
                    "original": "ColumnSize",
                    "resolved": "\"1\" | \"10\" | \"11\" | \"12\" | \"2\" | \"3\" | \"4\" | \"5\" | \"6\" | \"7\" | \"8\" | \"9\" | \"auto\" | undefined",
                    "references": {
                        "ColumnSize": {
                            "location": "import",
                            "path": "./col.types",
                            "id": "src/components/col/col.types.ts::ColumnSize"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Size of the column for lg screens"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            }
        };
    }
    static get listeners() {
        return [{
                "name": "resize",
                "method": "onResize",
                "target": "window",
                "capture": false,
                "passive": true
            }];
    }
}
Col.Breakpoints = ['', 'sm', 'md', 'lg'];
