// @siemens/ix/dist/collection/components/card-list/card-list.js
import { h, Host, } from "@stencil/core";
import { createMutationObserver } from "../utils/mutation-observer";
import { iconChevronDown, iconMoreMenu } from "@siemens/ix-icons/icons";
function CardListTitle(props) {
    if (!props.label) {
        return null;
    }
    return (h("div", { class: "CardList_Title" }, h("ix-icon-button", { variant: "tertiary", icon: iconChevronDown, onClick: props.onClick, iconColor: "color-primary", class: {
            CardList__Title__Button: true,
            CardList__Title__Button__Collapsed: props.isCollapsed,
        }, "aria-label": props.ariaLabelExpandButton }), h("ix-typography", { class: "CardList_Title__Label", format: "body-lg" }, props.label), !props.hideShowAll && (h("ix-button", { class: "CardList__Title__Show__All", variant: "tertiary", onClick: props.onShowAllClick }, h("span", null, props.showAllLabel), h("span", null, !isNaN(props.showAllCounter) ? ` (${props.showAllCounter})` : null)))));
}
export class CardList {
    constructor() {
        /**
         * Collapse the list
         */
        this.collapse = false;
        /**
         * List style
         */
        this.listStyle = 'stack';
        /**
         * Maximal visible cards
         *
         * @internal
         */
        this.maxVisibleCards = 12;
        /**
         * Suppress the overflow handling of child elements
         */
        this.suppressOverflowHandling = false;
        /**
         * Hide the show all button
         */
        this.hideShowAll = false;
        /**
         * i18n Show all button
         */
        this.i18nShowAll = 'Show all';
        /**
         * i18n More cards available
         */
        this.i18nMoreCards = 'There are more cards available';
        this.hasOverflowingElements = false;
        this.numberOfOverflowingElements = 0;
        this.numberOfAllChildElements = 0;
        this.leftScrollDistance = 0;
        this.rightScrollDistance = 0;
    }
    onCardListVisibilityToggle() {
        this.collapse = !this.collapse;
        this.collapseChanged.emit(this.collapse);
    }
    onShowAllClick(event) {
        this.showAllClick.emit({
            nativeEvent: event,
        });
    }
    getListChildren() {
        const slot = this.hostElement.shadowRoot.querySelector('.CardList__Content > slot');
        return slot.assignedElements({ flatten: true });
    }
    changeVisibilityOfSlotChildren() {
        const childElements = this.getListChildren();
        childElements.forEach((element, index) => {
            if (element instanceof HTMLElement) {
                if (index > this.maxVisibleCards - 1) {
                    element.classList.add('display-none');
                    return;
                }
                element.classList.remove('display-none');
            }
        });
        this.hasOverflowingElements = childElements.length > this.maxVisibleCards;
        this.numberOfOverflowingElements =
            childElements.length - this.maxVisibleCards;
        this.numberOfAllChildElements = childElements.length;
        this.detectOverflow();
    }
    registerOverflowHandler() {
        this.observer = createMutationObserver(() => {
            this.changeVisibilityOfSlotChildren();
        });
        this.observer.observe(this.hostElement.shadowRoot.querySelector('.CardList__Content'), {
            childList: true,
            subtree: true,
        });
        requestAnimationFrame(() => {
            this.changeVisibilityOfSlotChildren();
        });
    }
    shouldHandleOverflow() {
        if (this.suppressOverflowHandling) {
            return false;
        }
        if (this.listStyle === 'stack' || this.listStyle === 'scroll') {
            return true;
        }
    }
    get listElement() {
        return this.hostElement.shadowRoot.querySelector('.CardList__Content');
    }
    onCardListScroll() {
        this.detectOverflow();
    }
    isShowMoreCardVisible() {
        return (this.suppressOverflowHandling === false && this.hasOverflowingElements);
    }
    getOpacityFromScrollDistance(distance) {
        if (!this.listElement) {
            return 0;
        }
        if (distance === 0) {
            return 0;
        }
        if (distance > 100) {
            return 1;
        }
        return distance / 100;
    }
    computeMaskLayer() {
        const maxOverflowWidth = 80;
        const maskLayer = `linear-gradient(
      90deg,
      transparent 0px,
      black ${maxOverflowWidth *
            (this.getOpacityFromScrollDistance(this.leftScrollDistance) > 0 ? 1 : 0)}px,
      black calc(100% - ${maxOverflowWidth *
            (this.getOpacityFromScrollDistance(this.rightScrollDistance) > 0
                ? 1
                : 0)}px),
      transparent 100%
    )`;
        return {
            '--ix-card-list-overflow': maskLayer,
        };
    }
    detectOverflow() {
        if (!this.listElement) {
            return;
        }
        const { clientWidth, scrollWidth, scrollLeft } = this.listElement;
        this.leftScrollDistance = scrollLeft;
        this.rightScrollDistance = scrollWidth - scrollLeft - clientWidth;
    }
    componentDidLoad() {
        if (this.shouldHandleOverflow()) {
            this.registerOverflowHandler();
        }
    }
    disconnectedCallback() {
        if (this.observer) {
            this.observer.disconnect();
        }
    }
    render() {
        return (h(Host, { key: '37a1bb98b5703b76b6f0f5038561917bef13916b' }, h(CardListTitle, { key: '8f90c6052259ff8e1bbb377a9fa9cc465c951aab', isCollapsed: this.collapse, label: this.label, showAllLabel: this.i18nShowAll, showAllCounter: this.showAllCount === undefined
                ? this.numberOfAllChildElements
                : this.showAllCount, onClick: () => this.onCardListVisibilityToggle(), onShowAllClick: (e) => this.onShowAllClick(e), hideShowAll: this.hideShowAll }), h("div", { key: '1d1162c53d8fb1d97dfb4b00ae1c1a5942a0b0f7', class: {
                CardList__Overflow: true,
            }, style: this.computeMaskLayer() }, h("div", { key: '298741cb6e8f1617e9a3ca1f0b49d713c3421977', class: {
                CardList__Content: true,
                CardList__Content__Collapsed: this.collapse,
                CardList__Style__Flexbox__Scroll: this.listStyle === 'stack',
                CardList__Style__Infinite__Scroll: this.listStyle === 'scroll',
            }, onScroll: () => this.onCardListScroll() }, h("slot", { key: 'abccd55d11dae4c83b6d7dcdabc77aec0a2f7b32', onSlotchange: () => this.changeVisibilityOfSlotChildren() }), this.isShowMoreCardVisible() ? (h("ix-card", { class: {
                Show__All__Card: true,
            }, onClick: (event) => this.showMoreCardClick.emit({
                nativeEvent: event,
            }) }, h("ix-card-content", null, h("div", { class: "Show__All__Card__Content" }, h("ix-icon", { name: iconMoreMenu, size: '32', class: 'Show__All__Card__Icon' }), h("span", { class: "Show__All__Card__Text" }, this.i18nMoreCards, " (", this.numberOfOverflowingElements, ")"))))) : null))));
    }
    static get is() { return "ix-card-list"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["card-list.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["card-list.css"]
        };
    }
    static get properties() {
        return {
            "ariaLabelExpandButton": {
                "type": "string",
                "attribute": "aria-label-expand-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.2.0"
                        }],
                    "text": "ARIA label for the card's expand button.\nWill be set as aria-label on the nested HTML button element"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "label": {
                "type": "string",
                "attribute": "label",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Name the card list"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "collapse": {
                "type": "boolean",
                "attribute": "collapse",
                "mutable": true,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Collapse the list"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "listStyle": {
                "type": "string",
                "attribute": "list-style",
                "mutable": false,
                "complexType": {
                    "original": "'stack' | 'scroll'",
                    "resolved": "\"scroll\" | \"stack\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "List style"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'stack'"
            },
            "maxVisibleCards": {
                "type": "number",
                "attribute": "max-visible-cards",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }],
                    "text": "Maximal visible cards"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "12"
            },
            "showAllCount": {
                "type": "number",
                "attribute": "show-all-count",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Overwrite the default show all count."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "suppressOverflowHandling": {
                "type": "boolean",
                "attribute": "suppress-overflow-handling",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Suppress the overflow handling of child elements"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "hideShowAll": {
                "type": "boolean",
                "attribute": "hide-show-all",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Hide the show all button"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "i18nShowAll": {
                "type": "string",
                "attribute": "i18n-show-all",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "i18n Show all button"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'Show all'"
            },
            "i18nMoreCards": {
                "type": "string",
                "attribute": "i18n-more-cards",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "i18n More cards available"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'There are more cards available'"
            }
        };
    }
    static get states() {
        return {
            "hasOverflowingElements": {},
            "numberOfOverflowingElements": {},
            "numberOfAllChildElements": {},
            "leftScrollDistance": {},
            "rightScrollDistance": {}
        };
    }
    static get events() {
        return [{
                "method": "collapseChanged",
                "name": "collapseChanged",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Fire event when the collapse state is changed by the user"
                },
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                }
            }, {
                "method": "showAllClick",
                "name": "showAllClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Fire event when the collapse state is changed by the user"
                },
                "complexType": {
                    "original": "{\n    nativeEvent: MouseEvent;\n  }",
                    "resolved": "{ nativeEvent: MouseEvent; }",
                    "references": {
                        "MouseEvent": {
                            "location": "global",
                            "id": "global::MouseEvent"
                        }
                    }
                }
            }, {
                "method": "showMoreCardClick",
                "name": "showMoreCardClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Fire event when the show more card is clicked."
                },
                "complexType": {
                    "original": "{\n    nativeEvent: MouseEvent;\n  }",
                    "resolved": "{ nativeEvent: MouseEvent; }",
                    "references": {
                        "MouseEvent": {
                            "location": "global",
                            "id": "global::MouseEvent"
                        }
                    }
                }
            }];
    }
    static get elementRef() { return "hostElement"; }
    static get listeners() {
        return [{
                "name": "resize",
                "method": "detectOverflow",
                "target": "window",
                "capture": false,
                "passive": true
            }];
    }
}
