// @siemens/ix/dist/esm/index.js
export { I as InputState, L as LogicalFilterOperator } from './logical-filter-operator-BH3f5fa3.js';
export { d as defaultRefreshTreeOptions, r as renderDefaultItem } from './tree.types-DSTYytot.js';
export { U as UploadFileState } from './upload-file-state-BGzrnl_l.js';
import { g as getCoreDelegate } from './modal-BTxff2hq.js';
export { c as closeModal, d as dismissModal, r as registerFrameworkDelegate, a as resolveDelegate, b as showMessage, s as showModal } from './modal-BTxff2hq.js';
import { A as Animation } from './animation-C5MWUgDN.js';
export { T as TypedEvent } from './typed-event-BdCnOrqW.js';
export { a as convertToAbbreviationString, c as convertToRemString } from './rwd.util-pXYAoEyc.js';
export { g as getCurrentSystemAppearance, t as themeSwitcher } from './theme-switcher-DRqJGlG2.js';
export { h as handlePlatformHelpers } from './setup-BHVEOpQp.js';
import './index-8HpPmDK_.js';
import './index-CtoreFVF.js';

/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
class FilterState {
    constructor() {
        this.tokens = [];
        this.categories = [];
    }
}

function getToastContainer() {
    const containerList = Array.from(document.querySelectorAll('ix-toast-container'));
    const [container] = containerList;
    if (containerList.length > 1) {
        console.warn('Multiple toast containers were found. Only the first one will be used.');
        return container;
    }
    if (!container) {
        const toastContainer = document.createElement('ix-toast-container');
        document.body.appendChild(toastContainer);
        return toastContainer;
    }
    return container;
}
function setToastPosition(position) {
    const container = getToastContainer();
    container.setAttribute('position', position);
}
function toast(config) {
    const container = getToastContainer();
    return container.showToast(config);
}
toast.info = (config) => {
    return toast(Object.assign(Object.assign({}, config), { type: 'info' }));
};
toast.error = (config) => {
    return toast(Object.assign(Object.assign({}, config), { type: 'error' }));
};
toast.success = (config) => {
    return toast(Object.assign(Object.assign({}, config), { type: 'success' }));
};
toast.warning = (config) => {
    return toast(Object.assign(Object.assign({}, config), { type: 'warning' }));
};

/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
function showModalLoading(message) {
    const modal = document.createElement('ix-modal');
    modal.disableEscapeClose = true;
    const loading = document.createElement('ix-modal-loading');
    loading.innerText = message;
    modal.appendChild(loading);
    getCoreDelegate().attachView(modal);
    modal.showModal();
    return {
        update: (text) => (loading.innerHTML = text),
        finish: (text, timeout = 250) => {
            if (text !== undefined) {
                loading.innerText = text;
            }
            else {
                timeout = 0;
            }
            setTimeout(() => {
                modal.closeModal(null);
                setTimeout(async () => await getCoreDelegate().removeView(modal), Animation.mediumTime);
            }, timeout);
        },
    };
}

export { FilterState, getCoreDelegate, getToastContainer, setToastPosition, showModalLoading, toast };
