// @siemens/ix/dist/collection/components/date-input/date-input.js
/*
 * SPDX-FileCopyrightText: 2024 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
        r = Reflect.decorate(decorators, target, key, desc);
    else
        for (var i = decorators.length - 1; i >= 0; i--)
            if (d = decorators[i])
                r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { iconCalendar } from "@siemens/ix-icons/icons";
import { Host, h, } from "@stencil/core";
import { DateTime } from "luxon";
import { SlotEnd, SlotStart } from "../input/input.fc";
import { addDisposableChangesAndVisibilityObservers, adjustPaddingForStartAndEnd, handleSubmitOnEnterKeydown, } from "../input/input.util";
import { HookValidationLifecycle, createClassMutationObserver, getValidationText, } from "../utils/input";
import { makeRef } from "../utils/make-ref";
import { closeDropdown as closeDropdownUtil, createValidityState, handleIconClick, openDropdown as openDropdownUtil, } from "../utils/input/picker-input.util";
/**
 * @form-ready
 *
 * @slot start - Element will be displayed at the start of the input
 * @slot end - Element will be displayed at the end of the input
 */
export class DateInput {
    constructor() {
        /**
         * Value of the input element
         */
        this.value = '';
        /**
         * The earliest date that can be selected by the date input/picker.
         * If not set there will be no restriction.
         */
        this.minDate = '';
        /**
         * The latest date that can be selected by the date input/picker.
         * If not set there will be no restriction.
         */
        this.maxDate = '';
        /**
         * Date format string.
         * See {@link https://moment.github.io/luxon/#/formatting?id=table-of-tokens} for all available tokens.
         */
        this.format = 'yyyy/LL/dd';
        /**
         * Readonly attribute
         */
        this.readonly = false;
        /**
         * Disabled attribute
         */
        this.disabled = false;
        /**
         * I18n string for the error message when the date is not parsable
         */
        this.i18nErrorDateUnparsable = 'Date is not valid';
        /**
         * Shows week numbers displayed on the left side of the date picker
         *
         * @since 3.0.0
         */
        this.showWeekNumbers = false;
        /**
         * The index of which day to start the week on, based on the Locale#weekdays array.
         * E.g. if the locale is en-us, weekStartIndex = 1 results in starting the week on monday.
         */
        this.weekStartIndex = 0;
        /**
         * If false, pressing Enter will submit the form (if inside a form).
         * Set to true to suppress submit on Enter.
         */
        this.suppressSubmitOnEnter = false;
        /**
         * Text alignment within the date input. 'start' aligns the text to the start of the input, 'end' aligns the text to the end of the input.
         */
        this.textAlignment = 'start';
        this.show = false;
        this.from = null;
        this.isInputInvalid = false;
        this.isInvalid = false;
        this.isValid = false;
        this.isInfo = false;
        this.isWarning = false;
        this.focus = false;
        this.slotStartRef = makeRef();
        this.slotEndRef = makeRef();
        this.datepickerRef = makeRef();
        this.inputElementRef = makeRef();
        this.dropdownElementRef = makeRef();
        this.touched = false;
    }
    watchValuePropHandler(newValue) {
        this.onInput(newValue);
    }
    updateFormInternalValue(value) {
        if (value) {
            this.formInternals.setFormValue(value);
        }
        else {
            this.formInternals.setFormValue(null);
        }
        this.value = value;
    }
    connectedCallback() {
        this.classObserver = createClassMutationObserver(this.hostElement, () => this.checkClassList());
        this.disposableChangesAndVisibilityObservers =
            addDisposableChangesAndVisibilityObservers(this.hostElement, this.updatePaddings.bind(this));
    }
    componentWillLoad() {
        this.onInput(this.value);
        if (this.isInputInvalid) {
            this.from = null;
        }
        else {
            this.watchValue();
        }
        this.checkClassList();
        this.updateFormInternalValue(this.value);
    }
    updatePaddings() {
        adjustPaddingForStartAndEnd(this.slotStartRef.current, this.slotEndRef.current, this.inputElementRef.current);
    }
    disconnectedCallback() {
        var _a, _b;
        (_a = this.classObserver) === null || _a === void 0 ? void 0 : _a.destroy();
        (_b = this.disposableChangesAndVisibilityObservers) === null || _b === void 0 ? void 0 : _b.call(this);
    }
    watchValue() {
        this.from = this.value;
    }
    /** @internal */
    hasValidValue() {
        return Promise.resolve(!!this.value);
    }
    /** @internal */
    getAssociatedFormElement() {
        return Promise.resolve(this.formInternals.form);
    }
    async onInput(value) {
        this.value = value;
        if (!value) {
            this.valueChange.emit(value);
            return;
        }
        if (!this.format) {
            return;
        }
        const date = DateTime.fromFormat(value, this.format);
        const minDate = DateTime.fromFormat(this.minDate, this.format);
        const maxDate = DateTime.fromFormat(this.maxDate, this.format);
        this.isInputInvalid = !date.isValid || date < minDate || date > maxDate;
        if (this.isInputInvalid) {
            this.invalidReason = date.invalidReason || undefined;
            this.from = undefined;
        }
        else {
            this.updateFormInternalValue(value);
            this.closeDropdown();
        }
        this.valueChange.emit(value);
    }
    onCalenderClick(event) {
        handleIconClick(event, this.show, () => this.openDropdown(), this.inputElementRef);
    }
    async openDropdown() {
        return openDropdownUtil(this.dropdownElementRef);
    }
    async closeDropdown() {
        return closeDropdownUtil(this.dropdownElementRef);
    }
    checkClassList() {
        this.isInvalid = this.hostElement.classList.contains('ix-invalid');
    }
    handleInputKeyDown(event) {
        handleSubmitOnEnterKeydown(event, this.suppressSubmitOnEnter, this.formInternals.form);
    }
    renderInput() {
        var _a;
        return (h("div", { class: "input-wrapper" }, h(SlotStart, { slotStartRef: this.slotStartRef, onSlotChange: () => this.updatePaddings() }), h("input", { autoComplete: "off", class: {
                'is-invalid': this.isInputInvalid,
            }, disabled: this.disabled, readOnly: this.readonly, readonly: this.readonly, required: this.required, ref: this.inputElementRef, type: "text", value: (_a = this.value) !== null && _a !== void 0 ? _a : '', placeholder: this.placeholder, name: this.name, onInput: (event) => {
                const target = event.target;
                this.onInput(target.value);
            }, onClick: (event) => {
                if (this.show) {
                    event.stopPropagation();
                    event.preventDefault();
                }
            }, onFocus: async () => {
                this.openDropdown();
                this.ixFocus.emit();
            }, onBlur: () => {
                this.ixBlur.emit();
                this.touched = true;
            }, onKeyDown: (event) => this.handleInputKeyDown(event), style: {
                textAlign: this.textAlignment,
            } }), h(SlotEnd, { slotEndRef: this.slotEndRef, onSlotChange: () => this.updatePaddings() }, h("ix-icon-button", { "data-testid": "open-calendar", class: { 'calendar-hidden': this.disabled || this.readonly }, variant: "subtle-tertiary", icon: iconCalendar, onClick: (event) => this.onCalenderClick(event), "aria-label": this.ariaLabelCalendarButton }))));
    }
    hookValidationLifecycle({ isInfo, isInvalid, isInvalidByRequired, isValid, isWarning, }) {
        this.isInvalid = isInvalid || isInvalidByRequired || this.isInputInvalid;
        this.isInfo = isInfo;
        this.isValid = isValid;
        this.isWarning = isWarning;
    }
    async onInputValidationChange() {
        const state = await this.getValidityState();
        this.validityStateChange.emit({
            patternMismatch: state.patternMismatch,
            invalidReason: this.invalidReason,
        });
    }
    /** @internal */
    getValidityState() {
        return Promise.resolve(createValidityState(this.isInputInvalid, !!this.required, this.value));
    }
    /**
     * Get the native input element
     */
    getNativeInputElement() {
        return this.inputElementRef.waitForCurrent();
    }
    /**
     * Focuses the input field
     */
    async focusInput() {
        return (await this.getNativeInputElement()).focus();
    }
    /**
     * Returns whether the text field has been touched.
     * @internal
     */
    isTouched() {
        return Promise.resolve(this.touched);
    }
    render() {
        var _a;
        const invalidText = getValidationText(this.isInputInvalid, this.invalidText, this.i18nErrorDateUnparsable);
        return (h(Host, { key: '4a7f986c62a82763e46faa0203a5b6ff82415aaf', class: {
                disabled: this.disabled,
                readonly: this.readonly,
            } }, h("ix-field-wrapper", { key: 'c366d9ef3a538fd71e8d0f33d305e46da824421e', label: this.label, helperText: this.helperText, isInvalid: this.isInvalid, invalidText: invalidText, infoText: this.infoText, isInfo: this.isInfo, isWarning: this.isWarning, warningText: this.warningText, isValid: this.isValid, validText: this.validText, showTextAsTooltip: this.showTextAsTooltip, required: this.required, controlRef: this.inputElementRef }, this.renderInput()), h("ix-dropdown", { key: '6f10990b1e667f0e079f951d920707c15e24e50e', "data-testid": "date-dropdown", trigger: this.inputElementRef.waitForCurrent(), ref: this.dropdownElementRef, closeBehavior: "outside", suppressOverflowBehavior: true, show: this.show, onShowChanged: (event) => {
                this.show = event.detail;
            } }, h("ix-date-picker", { key: 'df14f590271c8177fa10360e72b0338088ed79be', ref: this.datepickerRef, format: this.format, locale: this.locale, singleSelection: true, from: (_a = this.from) !== null && _a !== void 0 ? _a : '', minDate: this.minDate, maxDate: this.maxDate, onDateChange: (event) => {
                const { from } = event.detail;
                this.onInput(from);
            }, showWeekNumbers: this.showWeekNumbers, ariaLabelNextMonthButton: this.ariaLabelNextMonthButton, ariaLabelPreviousMonthButton: this.ariaLabelPreviousMonthButton, embedded: true }))));
    }
    static get is() { return "ix-date-input"; }
    static get encapsulation() { return "shadow"; }
    static get formAssociated() { return true; }
    static get originalStyleUrls() {
        return {
            "$": ["date-input.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["date-input.css"]
        };
    }
    static get properties() {
        return {
            "name": {
                "type": "string",
                "attribute": "name",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Name of the input element"
                },
                "getter": false,
                "setter": false,
                "reflect": true
            },
            "placeholder": {
                "type": "string",
                "attribute": "placeholder",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Placeholder of the input element"
                },
                "getter": false,
                "setter": false,
                "reflect": true
            },
            "value": {
                "type": "string",
                "attribute": "value",
                "mutable": true,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Value of the input element"
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "defaultValue": "''"
            },
            "minDate": {
                "type": "string",
                "attribute": "min-date",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The earliest date that can be selected by the date input/picker.\nIf not set there will be no restriction."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "maxDate": {
                "type": "string",
                "attribute": "max-date",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The latest date that can be selected by the date input/picker.\nIf not set there will be no restriction."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "''"
            },
            "locale": {
                "type": "string",
                "attribute": "locale",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Locale identifier (e.g. 'en' or 'de').\nThe locale is used to translate the labels for weekdays and months.\nIt also determines the default order of weekdays based on the locale's conventions.\nWhen the locale changes, the weekday labels are rotated according to the `weekStartIndex`.\nIt does not affect the values returned by methods and events."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "format": {
                "type": "string",
                "attribute": "format",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Date format string.\nSee {@link https://moment.github.io/luxon/#/formatting?id=table-of-tokens} for all available tokens."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'yyyy/LL/dd'"
            },
            "required": {
                "type": "boolean",
                "attribute": "required",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Required attribute"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "helperText": {
                "type": "string",
                "attribute": "helper-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Helper text below the input field"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "label": {
                "type": "string",
                "attribute": "label",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Label of the input field"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "ariaLabelCalendarButton": {
                "type": "string",
                "attribute": "aria-label-calendar-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.2.0"
                        }],
                    "text": "ARIA label for the calendar icon button\nWill be set as aria-label on the nested HTML button element"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "invalidText": {
                "type": "string",
                "attribute": "invalid-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Error text below the input field"
                },
                "getter": false,
                "setter": false,
                "reflect": true
            },
            "readonly": {
                "type": "boolean",
                "attribute": "readonly",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Readonly attribute"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "disabled": {
                "type": "boolean",
                "attribute": "disabled",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Disabled attribute"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "infoText": {
                "type": "string",
                "attribute": "info-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Info text below the input field"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "warningText": {
                "type": "string",
                "attribute": "warning-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Warning text below the input field"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "validText": {
                "type": "string",
                "attribute": "valid-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Valid text below the input field"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "showTextAsTooltip": {
                "type": "boolean",
                "attribute": "show-text-as-tooltip",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Show text as tooltip"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "i18nErrorDateUnparsable": {
                "type": "string",
                "attribute": "i18n-error-date-unparsable",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "I18n string for the error message when the date is not parsable"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'Date is not valid'"
            },
            "showWeekNumbers": {
                "type": "boolean",
                "attribute": "show-week-numbers",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.0.0"
                        }],
                    "text": "Shows week numbers displayed on the left side of the date picker"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "weekStartIndex": {
                "type": "number",
                "attribute": "week-start-index",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The index of which day to start the week on, based on the Locale#weekdays array.\nE.g. if the locale is en-us, weekStartIndex = 1 results in starting the week on monday."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "0"
            },
            "ariaLabelPreviousMonthButton": {
                "type": "string",
                "attribute": "aria-label-previous-month-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "ARIA label for the previous month icon button\nWill be set as aria-label on the nested HTML button element"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "ariaLabelNextMonthButton": {
                "type": "string",
                "attribute": "aria-label-next-month-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "ARIA label for the next month icon button\nWill be set as aria-label on the nested HTML button element"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "suppressSubmitOnEnter": {
                "type": "boolean",
                "attribute": "suppress-submit-on-enter",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "If false, pressing Enter will submit the form (if inside a form).\nSet to true to suppress submit on Enter."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "defaultValue": "false"
            },
            "textAlignment": {
                "type": "string",
                "attribute": "text-alignment",
                "mutable": false,
                "complexType": {
                    "original": "'start' | 'end'",
                    "resolved": "\"end\" | \"start\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Text alignment within the date input. 'start' aligns the text to the start of the input, 'end' aligns the text to the end of the input."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'start'"
            }
        };
    }
    static get states() {
        return {
            "show": {},
            "from": {},
            "isInputInvalid": {},
            "isInvalid": {},
            "isValid": {},
            "isInfo": {},
            "isWarning": {},
            "focus": {}
        };
    }
    static get events() {
        return [{
                "method": "valueChange",
                "name": "valueChange",
                "bubbles": true,
                "cancelable": false,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Input change event."
                },
                "complexType": {
                    "original": "string | undefined",
                    "resolved": "string | undefined",
                    "references": {}
                }
            }, {
                "method": "validityStateChange",
                "name": "validityStateChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Validation state change event."
                },
                "complexType": {
                    "original": "DateInputValidityState",
                    "resolved": "{ patternMismatch: boolean; invalidReason?: string | undefined; }",
                    "references": {
                        "DateInputValidityState": {
                            "location": "import",
                            "path": "./date-input.types",
                            "id": "src/components/date-input/date-input.types.ts::DateInputValidityState"
                        }
                    }
                }
            }, {
                "method": "ixFocus",
                "name": "ixFocus",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }],
                    "text": ""
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }, {
                "method": "ixBlur",
                "name": "ixBlur",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }],
                    "text": ""
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }];
    }
    static get methods() {
        return {
            "hasValidValue": {
                "complexType": {
                    "signature": "() => Promise<boolean>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<boolean>"
                },
                "docs": {
                    "text": "",
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }]
                }
            },
            "getAssociatedFormElement": {
                "complexType": {
                    "signature": "() => Promise<HTMLFormElement | null>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "HTMLFormElement": {
                            "location": "global",
                            "id": "global::HTMLFormElement"
                        }
                    },
                    "return": "Promise<HTMLFormElement | null>"
                },
                "docs": {
                    "text": "",
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }]
                }
            },
            "getValidityState": {
                "complexType": {
                    "signature": "() => Promise<ValidityState>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "ValidityState": {
                            "location": "global",
                            "id": "global::ValidityState"
                        }
                    },
                    "return": "Promise<ValidityState>"
                },
                "docs": {
                    "text": "",
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }]
                }
            },
            "getNativeInputElement": {
                "complexType": {
                    "signature": "() => Promise<HTMLInputElement>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "HTMLInputElement": {
                            "location": "global",
                            "id": "global::HTMLInputElement"
                        }
                    },
                    "return": "Promise<HTMLInputElement>"
                },
                "docs": {
                    "text": "Get the native input element",
                    "tags": []
                }
            },
            "focusInput": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Focuses the input field",
                    "tags": []
                }
            },
            "isTouched": {
                "complexType": {
                    "signature": "() => Promise<boolean>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<boolean>"
                },
                "docs": {
                    "text": "Returns whether the text field has been touched.",
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }]
                }
            }
        };
    }
    static get elementRef() { return "hostElement"; }
    static get watchers() {
        return [{
                "propName": "value",
                "methodName": "watchValuePropHandler"
            }, {
                "propName": "value",
                "methodName": "watchValue"
            }, {
                "propName": "isInputInvalid",
                "methodName": "onInputValidationChange"
            }];
    }
    static get attachInternalsMemberName() { return "formInternals"; }
}
__decorate([
    HookValidationLifecycle()
], DateInput.prototype, "hookValidationLifecycle", null);
