// @siemens/ix/dist/collection/components/tile/tile.js
/*
 * SPDX-FileCopyrightText: 2024 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host } from "@stencil/core";
export class Tile {
    constructor() {
        /**
         * Size of the tile - one of 'small', 'medium' or 'large'
         */
        this.size = 'medium';
        this.hasHeaderSlot = false;
        this.hasFooterSlot = false;
    }
    handleHeaderSlotChange() {
        this.hasHeaderSlot = !!this.hostElement.querySelector('[slot="header"]');
    }
    handleFooterSlotChange() {
        this.hasFooterSlot = !!this.hostElement.querySelector('[slot="footer"]');
    }
    render() {
        return (h(Host, { key: 'f5c50da791dcbdfeb884d33071b6d557360c1a60', class: {
                'tile-small': this.size === 'small',
                'tile-medium': this.size === 'medium',
                'tile-big': this.size === 'big',
            } }, h("div", { key: '651e393dbd8f7f7cd55373c3bdadc195dfc88cbd', class: {
                'tile-header': true,
                'has-content': this.hasHeaderSlot,
            } }, h("slot", { key: '1b94a241077664e5bcb535321f816b3b3faa9fad', name: "header", onSlotchange: () => this.handleHeaderSlotChange() })), h("div", { key: 'd2ce3123c9ffa767e07edf25248567ee9406fc95', class: "tile-subheader" }, h("slot", { key: '025ac544a3750fbfb145d1350e7091e969c2e369', name: "subheader" })), h("div", { key: '6ce8df1e1b426df28b42f490db11dc1d951715e6', class: "tile-content" }, h("slot", { key: '7bb66b49df63f613c3b7be8ff0da64bc16dde6b9' })), h("div", { key: '55f6e978190c3c8581310559a7a46ec6c4cc0e2f', class: {
                'tile-footer': true,
                'has-content': this.hasFooterSlot,
            } }, h("slot", { key: 'd63334865e6a32a6bb082939b44f3bf0b041b93a', name: "footer", onSlotchange: () => this.handleFooterSlotChange() }))));
    }
    static get is() { return "ix-tile"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tile.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tile.css"]
        };
    }
    static get properties() {
        return {
            "size": {
                "type": "string",
                "attribute": "size",
                "mutable": false,
                "complexType": {
                    "original": "'small' | 'medium' | 'big'",
                    "resolved": "\"big\" | \"medium\" | \"small\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Size of the tile - one of 'small', 'medium' or 'large'"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'medium'"
            }
        };
    }
    static get states() {
        return {
            "hasHeaderSlot": {},
            "hasFooterSlot": {}
        };
    }
    static get elementRef() { return "hostElement"; }
}
