// @siemens/ix/dist/collection/components/message-bar/message-bar.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host, } from "@stencil/core";
import { animate } from "animejs";
import { iconClose, iconError, iconInfo, iconNotification, iconSuccess, iconWarning, iconWarningRhomb, } from "@siemens/ix-icons/icons";
export class MessageBar {
    constructor() {
        /**
         * Specifies the type of the alert.
         */
        this.type = 'info';
        /**
         * If true, close button is disabled and alert cannot be dismissed by the user
         */
        this.persistent = false;
    }
    componentWillRender() {
        const config = MessageBar.messageTypeConfigs[this.type];
        if (config) {
            this.icon = config.icon;
            this.color = config.color;
        }
    }
    closeAlert(el) {
        const { defaultPrevented } = this.closedChange.emit();
        if (!defaultPrevented) {
            animate(el, {
                duration: MessageBar.duration,
                opacity: [1, 0],
                easing: 'easeOutSine',
                onComplete: () => {
                    el.classList.add('message-bar-hidden');
                    this.closeAnimationCompleted.emit();
                },
            });
        }
    }
    render() {
        return (h(Host, { key: '538fcc08ad07fe808f424e067a361f4a56f4c0ec' }, h("div", { key: '4a75052efe1b22c247dcfcd298aabbd1098ff7b6', class: { 'message-container': true, [this.type]: true }, role: "alert", ref: (el) => (this.divElement = el) }, h("ix-icon", { key: 'd348829692e38a1023e044834aefa45516b4d916', color: this.color, name: this.icon }), h("div", { key: '0ac0a509efe1dd246f7732660e4cf16416c8655d', class: "message-content" }, h("slot", { key: '6143bfa9cf39d32d149c2893b7712358337833c9' })), !this.persistent && (h("ix-icon-button", { key: 'b118c91956af0fd00d02cfefdb7da36d46f2d360', icon: iconClose, iconColor: "color-soft-text", size: "24", variant: "tertiary", onClick: () => {
                if (this.divElement) {
                    this.closeAlert(this.divElement);
                }
            }, "data-testid": "close-btn" })))));
    }
    static get is() { return "ix-message-bar"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["message-bar.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["message-bar.css"]
        };
    }
    static get properties() {
        return {
            "type": {
                "type": "string",
                "attribute": "type",
                "mutable": false,
                "complexType": {
                    "original": "| 'alarm'\n    | 'critical'\n    | 'warning'\n    | 'success'\n    | 'info'\n    | 'neutral'\n    | 'primary'",
                    "resolved": "\"alarm\" | \"critical\" | \"info\" | \"neutral\" | \"primary\" | \"success\" | \"warning\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Specifies the type of the alert."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'info'"
            },
            "persistent": {
                "type": "boolean",
                "attribute": "persistent",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "If true, close button is disabled and alert cannot be dismissed by the user"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
    static get states() {
        return {
            "icon": {},
            "color": {}
        };
    }
    static get events() {
        return [{
                "method": "closedChange",
                "name": "closedChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "An event emitted when the close button is clicked"
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }, {
                "method": "closeAnimationCompleted",
                "name": "closeAnimationCompleted",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "An event emitted when the close animation is completed"
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }];
    }
}
MessageBar.duration = 300;
MessageBar.messageTypeConfigs = {
    //TODO(IX-3400): Replace icon colors with proper CSS variables when available
    alarm: { icon: iconError, color: 'color-alarm' },
    critical: { icon: iconWarningRhomb, color: 'color-critical' },
    warning: { icon: iconWarning, color: 'color-warning-text' },
    success: { icon: iconSuccess, color: 'color-success' },
    info: { icon: iconInfo, color: 'color-info' },
    neutral: { icon: iconNotification, color: 'color-neutral' },
    primary: { icon: iconNotification, color: 'color-primary' },
};
