// @siemens/ix/dist/collection/components/toast/toast.js
/*
 * SPDX-FileCopyrightText: 2024 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host, } from "@stencil/core";
import { iconClose, iconError, iconInfo, iconSuccess, iconWarning, } from "@siemens/ix-icons/icons";
export class Toast {
    constructor() {
        /**
         * Toast type
         */
        this.type = 'info';
        /**
         * Autoclose title after delay
         */
        this.autoCloseDelay = 5000;
        /**
         * Autoclose behavior
         */
        this.preventAutoClose = false;
        this.progress = 0;
        this.touched = false;
        this.paused = false;
    }
    getIcon() {
        if (this.icon) {
            return (h("ix-icon", { "data-testid": "toast-icon", name: this.icon, color: this.iconColor, size: "24" }));
        }
        switch (this.type) {
            case 'info':
                return (h("ix-icon", { "data-testid": "toast-icon", name: iconInfo, size: "24", color: "color-std-text" }));
            case 'error':
                return (h("ix-icon", { "data-testid": "toast-icon", name: iconError, size: "24", color: "color-alarm" }));
            case 'success':
                return (h("ix-icon", { "data-testid": "toast-icon", name: iconSuccess, size: "24", color: "color-success" }));
            case 'warning':
                //TODO(IX-3400): Replace icon colors with proper CSS variables when available
                return (h("ix-icon", { "data-testid": "toast-icon", name: iconWarning, size: "24", color: "color-warning-text" }));
            default:
                return '';
        }
    }
    close() {
        if (this.hostElement) {
            this.hostElement.classList.add('animate__fadeOut');
        }
        setTimeout(() => {
            this.closeToast.emit();
        }, 250);
    }
    /**
     * Pause the toast's auto-close progress bar and timer.
     */
    async pause() {
        this.paused = true;
    }
    /**
     * Resume the toast's auto-close progress bar and timer if previously paused.
     */
    async resume() {
        this.paused = false;
    }
    /**
     * Returns whether the toast is currently paused (auto-close is paused).
     */
    async isPaused() {
        return this.paused || this.touched;
    }
    render() {
        let progressBarStyle = {};
        const progressBarClass = ['toast-progress-bar'];
        progressBarStyle = {
            animationDuration: `${this.autoCloseDelay}ms`,
            animationPlayState: this.touched || this.paused ? 'paused' : 'running',
        };
        progressBarClass.push('toast-progress-bar--animated');
        return (h(Host, { key: 'b006843e5552e82ed437b10ab31c5cdff7ec90a6', class: "animate__animated animate__fadeIn" }, h("div", { key: '8d4c07604138e38235b3a2f042ce8550966df392', class: "toast-body", onPointerLeave: () => {
                this.touched = false;
            }, onPointerEnter: () => {
                this.touched = true;
            } }, this.type || this.icon ? (h("div", { class: "toast-icon" }, this.getIcon())) : null, h("div", { key: 'b5235db039ca12fa5a3823861f41e86fbf312661', class: "toast-content" }, this.toastTitle ? (h("ix-typography", { class: "toast-title", format: "h5" }, this.toastTitle)) : null, h("div", { key: '6577c06f5530655aafc974684aedbac7d48da649', class: "toast-message" }, h("slot", { key: '16675e05c803fe4ef8e246463f30c194e76a8428' })), h("div", { key: '84eeb321fc81b49d14468e7d517dd8cfc8769c8d', class: "toast-action" }, h("slot", { key: 'f55d9f57a23bf58006e1f3bc1d3e7fab85440dab', name: "action" }))), h("div", { key: 'd409d2622bce56496f6081cf55ab9ad1137dc222', class: "toast-close" }, h("ix-icon-button", { key: '873c0a73f6c5a9ae47cd39d63ff9c68dde219cbd', icon: iconClose, iconColor: "color-soft-text", size: "24", variant: "tertiary", onClick: () => this.closeToast.emit(), "aria-label": this.ariaLabelCloseIconButton }))), !this.preventAutoClose && (h("div", { key: '15a6b8556e7d22e9c9061a27e31b3138cb86380b', class: progressBarClass.join(' '), style: progressBarStyle, onAnimationEnd: () => {
                this.close();
            }, onTransitionEnd: () => {
                if (this.progress === 0) {
                    this.close();
                }
            } }))));
    }
    static get is() { return "ix-toast"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["toast.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["toast.css"]
        };
    }
    static get properties() {
        return {
            "type": {
                "type": "string",
                "attribute": "type",
                "mutable": false,
                "complexType": {
                    "original": "ToastType",
                    "resolved": "\"error\" | \"info\" | \"success\" | \"warning\"",
                    "references": {
                        "ToastType": {
                            "location": "import",
                            "path": "./toast-utils",
                            "id": "src/components/toast/toast-utils.ts::ToastType"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Toast type"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'info'"
            },
            "toastTitle": {
                "type": "string",
                "attribute": "toast-title",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Toast title"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "autoCloseDelay": {
                "type": "number",
                "attribute": "auto-close-delay",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Autoclose title after delay"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "5000"
            },
            "preventAutoClose": {
                "type": "boolean",
                "attribute": "prevent-auto-close",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Autoclose behavior"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "icon": {
                "type": "string",
                "attribute": "icon",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Icon of toast"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "iconColor": {
                "type": "string",
                "attribute": "icon-color",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Icon color of toast"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "ariaLabelCloseIconButton": {
                "type": "string",
                "attribute": "aria-label-close-icon-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.2.0"
                        }],
                    "text": "ARIA label for the close icon button\nWill be set as aria-label on the nested HTML button element"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            }
        };
    }
    static get states() {
        return {
            "progress": {},
            "touched": {},
            "paused": {}
        };
    }
    static get events() {
        return [{
                "method": "closeToast",
                "name": "closeToast",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Toast closed"
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }];
    }
    static get methods() {
        return {
            "pause": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Pause the toast's auto-close progress bar and timer.",
                    "tags": []
                }
            },
            "resume": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Resume the toast's auto-close progress bar and timer if previously paused.",
                    "tags": []
                }
            },
            "isPaused": {
                "complexType": {
                    "signature": "() => Promise<boolean>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<boolean>"
                },
                "docs": {
                    "text": "Returns whether the toast is currently paused (auto-close is paused).",
                    "tags": []
                }
            }
        };
    }
    static get elementRef() { return "hostElement"; }
}
