// @siemens/ix/dist/collection/components/blind/blind.js
/*
 * SPDX-FileCopyrightText: 2024 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Fragment, h, Host, } from "@stencil/core";
import { animate } from "animejs";
import { a11yBoolean } from "../utils/a11y";
import { iconChevronDownSmall } from "@siemens/ix-icons/icons";
import Animation from "../utils/animation";
let sequentialInstanceId = 0;
export class Blind {
    constructor() {
        /**
         * Collapsed state
         */
        this.collapsed = false;
        /**
         * Blind variant
         */
        this.variant = 'filled';
        this.blindId = ++sequentialInstanceId;
    }
    onHeaderClick() {
        this.collapsed = !this.collapsed;
        this.collapsedChange.emit(this.collapsed);
    }
    componentDidLoad() {
        this.animateCollapse(this.collapsed);
    }
    get content() {
        return this.hostElement.shadowRoot.querySelector('.blind-content');
    }
    animation(isCollapsed) {
        this.animateCollapse(isCollapsed);
    }
    animateCollapse(isCollapsed) {
        if (isCollapsed) {
            this.rotateChevronDown();
        }
        else {
            this.rotateChevronUp();
        }
    }
    rotateChevronUp() {
        animate(this.chevronRef, {
            duration: Animation.defaultTime,
            easing: 'easeInOutSine',
            rotateZ: 180,
        });
        animate(this.content, {
            duration: Animation.defaultTime,
            easing: 'easeInOutSine',
            opacity: 1,
        });
    }
    rotateChevronDown() {
        animate(this.chevronRef, {
            duration: Animation.defaultTime,
            easing: 'easeInOutSine',
            rotateZ: 0,
        });
        animate(this.content, {
            duration: Animation.defaultTime,
            easing: 'easeInOutSine',
            opacity: 0,
        });
    }
    render() {
        return (h(Host, { key: '671ad2c0686ea67b8dd5a3b6259edd8cd2667b19', class: {
                [`blind-${this.variant}`]: true,
            } }, h("div", { key: '40fa7a9d37b674e2979022aadc84a4496ea9a9f5', class: 'blind-header-wrapper' }, h("button", { key: '29c3e84c41ec924f2b3c5a373778130afa943626', class: {
                'blind-header': true,
                [`blind-${this.variant}`]: true,
                closed: this.collapsed,
            }, type: "button", "aria-labelledby": `ix-blind-header-title-${this.blindId}`, "aria-controls": `ix-blind-content-section-${this.blindId}`, "aria-expanded": a11yBoolean(!this.collapsed), onClick: () => this.onHeaderClick() }, h("slot", { key: 'b252b735380d02d028f14e7d9fbe4c5eba8c6602', name: "custom-header" })), h("div", { key: '6bf0d3c1465c63a94c8f022734ed826c4028855a', class: 'blind-header-content' }, h("ix-icon", { key: 'd21dd7db4cb813e4becce868c4dbe340a12b1e65', class: "collapse-icon", name: iconChevronDownSmall, color: this.variant === 'filled' || this.variant === 'outline'
                ? 'color-std-text'
                : `color-${this.variant}--contrast`, ref: (ref) => (this.chevronRef = ref) }), h("div", { key: 'da2e1435ca1070eed1794226ca721e7e3fb22390', class: "blind-header-title", id: `ix-blind-header-title-${this.blindId}` }, this.label !== undefined ? (h(Fragment, null, this.icon && (h("ix-icon", { class: "blind-header-title-icon", name: this.icon, color: this.variant === 'filled' || this.variant === 'outline'
                ? 'color-std-text'
                : `color-${this.variant}--contrast` })), h("div", { class: 'blind-header-title-row' }, h("div", { class: "blind-header-title-col" }, h("ix-typography", { title: this.label, format: "label-lg", bold: true }, h("div", { class: "blind-header-title-label", title: this.label }, this.label))), this.sublabel && (h("div", { class: "blind-header-title-col" }, h("ix-typography", { title: this.sublabel }, h("div", { class: "blind-header-title-sublabel" }, this.sublabel))))), h("div", { class: "header-actions" }, h("slot", { name: "header-actions" })))) : null))), h("section", { key: '854dd19e82a63ccb31f090dadfbcfa46d7346d21', id: `ix-blind-content-section-${this.blindId}`, "aria-labelledby": `ix-blind-header-title-${this.blindId}` }, h("div", { key: '12f77fdbb29c89d5ab26877e9db8ad9ec79f1e6f', class: {
                'blind-content': true,
                hide: this.collapsed,
            } }, !this.collapsed ? h("slot", null) : null))));
    }
    static get is() { return "ix-blind"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["blind.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["blind.css"]
        };
    }
    static get properties() {
        return {
            "collapsed": {
                "type": "boolean",
                "attribute": "collapsed",
                "mutable": true,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Collapsed state"
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "defaultValue": "false"
            },
            "label": {
                "type": "string",
                "attribute": "label",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Label of blind"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "sublabel": {
                "type": "string",
                "attribute": "sublabel",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Secondary label inside blind header"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "icon": {
                "type": "string",
                "attribute": "icon",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Optional icon to be displayed next to the header label"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "variant": {
                "type": "string",
                "attribute": "variant",
                "mutable": false,
                "complexType": {
                    "original": "BlindVariant",
                    "resolved": "\"alarm\" | \"critical\" | \"filled\" | \"info\" | \"neutral\" | \"outline\" | \"primary\" | \"success\" | \"warning\"",
                    "references": {
                        "BlindVariant": {
                            "location": "import",
                            "path": "./blind.types",
                            "id": "src/components/blind/blind.types.ts::BlindVariant"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Blind variant"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'filled'"
            }
        };
    }
    static get events() {
        return [{
                "method": "collapsedChange",
                "name": "collapsedChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Collapsed state changed"
                },
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "hostElement"; }
    static get watchers() {
        return [{
                "propName": "collapsed",
                "methodName": "animation"
            }];
    }
}
