// @siemens/ix/dist/collection/components/flip-tile-content/flip-tile-content.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host } from "@stencil/core";
export class FlipTileContent {
    constructor() {
        /**
         * Controls the visibility of the content
         *
         * @internal
         */
        this.contentVisible = false;
    }
    render() {
        return h(Host, { key: '264e3539d6c5abcd8716968a4128da99ca191375' }, this.contentVisible ? h("slot", null) : null);
    }
    static get is() { return "ix-flip-tile-content"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["flip-tile-content.css"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["flip-tile-content.css"]
        };
    }
    static get properties() {
        return {
            "contentVisible": {
                "type": "boolean",
                "attribute": "content-visible",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }],
                    "text": "Controls the visibility of the content"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
}
