// @siemens/ix/dist/collection/components/dropdown-button/dropdown-button.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host } from "@stencil/core";
import { iconChevronDownSmall, iconChevronUpSmall, } from "@siemens/ix-icons/icons";
import { makeRef } from "../utils/make-ref";
export class DropdownButton {
    constructor() {
        /**
         * Button variant
         */
        this.variant = 'primary';
        /**
         * Disable button
         */
        this.disabled = false;
        /**
         * Controls if the dropdown will be closed in response to a click event depending on the position of the event relative to the dropdown.
         */
        this.closeBehavior = 'both';
        this.dropdownShow = false;
        this.dropdownAnchor = makeRef();
        this.onDropdownShowChanged = (event) => {
            this.dropdownShow = event.detail;
        };
    }
    getTriangle() {
        return (h("div", { class: {
                triangle: true,
                [this.variant]: true,
                hide: !!this.label,
                disabled: this.disabled,
            } }));
    }
    render() {
        return (h(Host, { key: '4e4aca39eff13b8f665682b3f2dc9ace533c7ca0', class: {
                disabled: this.disabled,
            }, ref: this.dropdownAnchor }, h("div", { key: '9583ed2274621e67d779e9668fead87f663b7558', class: "dropdown-button" }, this.label ? (h("ix-button", { variant: this.variant, disabled: this.disabled, alignment: "start", ariaLabel: this.ariaLabelDropdownButton }, h("div", { class: 'content' }, this.icon ? (h("ix-icon", { name: this.icon, size: "24", class: 'dropdown-icon' })) : null, h("div", { class: 'button-label' }, this.label), h("ix-icon", { name: this.dropdownShow
                ? iconChevronUpSmall
                : iconChevronDownSmall, size: "24" })))) : (h("div", null, h("ix-icon-button", { icon: this.icon, variant: this.variant, disabled: this.disabled }), this.getTriangle()))), h("ix-dropdown", { key: 'e4fd6fb8451d8883fac84764758a90ce6dc4d433', class: "dropdown", trigger: this.dropdownAnchor.waitForCurrent(), placement: this.placement, closeBehavior: this.closeBehavior, onShowChanged: this.onDropdownShowChanged }, h("slot", { key: 'be8d3427e80c326bc7a6993f53190e290b824967' }))));
    }
    static get is() { return "ix-dropdown-button"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["dropdown-button.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["dropdown-button.css"]
        };
    }
    static get properties() {
        return {
            "variant": {
                "type": "string",
                "attribute": "variant",
                "mutable": false,
                "complexType": {
                    "original": "DropdownButtonVariant",
                    "resolved": "\"danger-primary\" | \"danger-secondary\" | \"danger-tertiary\" | \"primary\" | \"secondary\" | \"subtle-primary\" | \"subtle-secondary\" | \"subtle-tertiary\" | \"tertiary\"",
                    "references": {
                        "DropdownButtonVariant": {
                            "location": "import",
                            "path": "./dropdown-button.types",
                            "id": "src/components/dropdown-button/dropdown-button.types.ts::DropdownButtonVariant"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Button variant"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'primary'"
            },
            "disabled": {
                "type": "boolean",
                "attribute": "disabled",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Disable button"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "label": {
                "type": "string",
                "attribute": "label",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Set label"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "icon": {
                "type": "string",
                "attribute": "icon",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Button icon"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "closeBehavior": {
                "type": "any",
                "attribute": "close-behavior",
                "mutable": false,
                "complexType": {
                    "original": "'inside' | 'outside' | 'both' | boolean",
                    "resolved": "\"both\" | \"inside\" | \"outside\" | boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Controls if the dropdown will be closed in response to a click event depending on the position of the event relative to the dropdown."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'both'"
            },
            "placement": {
                "type": "string",
                "attribute": "placement",
                "mutable": false,
                "complexType": {
                    "original": "AlignedPlacement",
                    "resolved": "\"bottom-end\" | \"bottom-start\" | \"left-end\" | \"left-start\" | \"right-end\" | \"right-start\" | \"top-end\" | \"top-start\" | undefined",
                    "references": {
                        "AlignedPlacement": {
                            "location": "import",
                            "path": "../dropdown/placement",
                            "id": "src/components/dropdown/placement.ts::AlignedPlacement"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Placement of the dropdown"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "ariaLabelDropdownButton": {
                "type": "string",
                "attribute": "aria-label-dropdown-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.2.0"
                        }],
                    "text": "ARIA label for the dropdown button\nWill be set as aria-label on the nested HTML button element"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            }
        };
    }
    static get states() {
        return {
            "dropdownShow": {}
        };
    }
    static get elementRef() { return "hostElement"; }
}
