// @siemens/ix/dist/collection/components/typography/typography.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host } from "@stencil/core";
export class IxTypography {
    constructor() {
        /**
         * Display text bold
         */
        this.bold = false;
        /**
         * Text decoration
         */
        this.textDecoration = 'none';
    }
    static getTextColor(color) {
        if (color.startsWith('inv-') || !color.endsWith('-contrast')) {
            return `var(--theme-color-${color}-text)`;
        }
        return `var(--theme-color-${color.replace('-', '--')})`;
    }
    render() {
        var _a;
        let typographyClass = {};
        typographyClass[`typography-${(_a = this.format) !== null && _a !== void 0 ? _a : 'body'}`] = true;
        if (this.textDecoration !== 'none') {
            typographyClass[`typography-decoration-${this.textDecoration}`] = true;
        }
        typographyClass['typography-weight-bold'] = this.bold;
        let style = {};
        if (this.textColor) {
            style = {
                color: IxTypography.getTextColor(this.textColor),
            };
        }
        return (h(Host, { key: 'de007c46746d98dffe4647fca37b0d8dd1967415', class: typographyClass, style: style }, h("slot", { key: 'a4ada9ab3f8a64839f4a6c7938b3769f6ef0d6c7' })));
    }
    static get is() { return "ix-typography"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["typography.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["typography.css"]
        };
    }
    static get properties() {
        return {
            "format": {
                "type": "string",
                "attribute": "format",
                "mutable": false,
                "complexType": {
                    "original": "TypographyFormat",
                    "resolved": "TypographyFormat | undefined",
                    "references": {
                        "TypographyFormat": {
                            "location": "import",
                            "path": "./typography.types",
                            "id": "src/components/typography/typography.types.ts::TypographyFormat"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Text format"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "textColor": {
                "type": "string",
                "attribute": "text-color",
                "mutable": false,
                "complexType": {
                    "original": "TypographyColors",
                    "resolved": "\"alarm\" | \"alarm-contrast\" | \"contrast\" | \"critical-contrast\" | \"info-contrast\" | \"inv-contrast\" | \"inv-soft\" | \"inv-std\" | \"inv-weak\" | \"neutral-contrast\" | \"primary-contrast\" | \"soft\" | \"std\" | \"success-contrast\" | \"warning-contrast\" | \"weak\" | undefined",
                    "references": {
                        "TypographyColors": {
                            "location": "import",
                            "path": "./typography.types",
                            "id": "src/components/typography/typography.types.ts::TypographyColors"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Text color based on theme variables"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "bold": {
                "type": "boolean",
                "attribute": "bold",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Display text bold"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "textDecoration": {
                "type": "string",
                "attribute": "text-decoration",
                "mutable": false,
                "complexType": {
                    "original": "TextDecoration",
                    "resolved": "\"line-through\" | \"none\" | \"underline\"",
                    "references": {
                        "TextDecoration": {
                            "location": "import",
                            "path": "./typography.types",
                            "id": "src/components/typography/typography.types.ts::TextDecoration"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Text decoration"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'none'"
            }
        };
    }
}
