// @siemens/ix-react/dist/components.js
"use client";
import { defineCustomElement as defineCustomElement$1, IxActionCard as IxActionCard$1 } from '@siemens/ix/components/ix-action-card.js';
import { defineCustomElement as defineCustomElement$3, IxApplicationHeader as IxApplicationHeader$1 } from '@siemens/ix/components/ix-application-header.js';
import { defineCustomElement as defineCustomElement$2, IxApplication as IxApplication$1 } from '@siemens/ix/components/ix-application.js';
import { defineCustomElement as defineCustomElement$4, IxAvatar as IxAvatar$1 } from '@siemens/ix/components/ix-avatar.js';
import { defineCustomElement as defineCustomElement$5, IxBlind as IxBlind$1 } from '@siemens/ix/components/ix-blind.js';
import { defineCustomElement as defineCustomElement$7, IxBreadcrumbItem as IxBreadcrumbItem$1 } from '@siemens/ix/components/ix-breadcrumb-item.js';
import { defineCustomElement as defineCustomElement$6, IxBreadcrumb as IxBreadcrumb$1 } from '@siemens/ix/components/ix-breadcrumb.js';
import { defineCustomElement as defineCustomElement$8, IxButton as IxButton$1 } from '@siemens/ix/components/ix-button.js';
import { defineCustomElement as defineCustomElement$a, IxCardAccordion as IxCardAccordion$1 } from '@siemens/ix/components/ix-card-accordion.js';
import { defineCustomElement as defineCustomElement$b, IxCardContent as IxCardContent$1 } from '@siemens/ix/components/ix-card-content.js';
import { defineCustomElement as defineCustomElement$c, IxCardList as IxCardList$1 } from '@siemens/ix/components/ix-card-list.js';
import { defineCustomElement as defineCustomElement$d, IxCardTitle as IxCardTitle$1 } from '@siemens/ix/components/ix-card-title.js';
import { defineCustomElement as defineCustomElement$9, IxCard as IxCard$1 } from '@siemens/ix/components/ix-card.js';
import { defineCustomElement as defineCustomElement$e, IxCategoryFilter as IxCategoryFilter$1 } from '@siemens/ix/components/ix-category-filter.js';
import { defineCustomElement as defineCustomElement$g, IxCheckboxGroup as IxCheckboxGroup$1 } from '@siemens/ix/components/ix-checkbox-group.js';
import { defineCustomElement as defineCustomElement$f, IxCheckbox as IxCheckbox$1 } from '@siemens/ix/components/ix-checkbox.js';
import { defineCustomElement as defineCustomElement$h, IxChip as IxChip$1 } from '@siemens/ix/components/ix-chip.js';
import { defineCustomElement as defineCustomElement$i, IxCol as IxCol$1 } from '@siemens/ix/components/ix-col.js';
import { defineCustomElement as defineCustomElement$k, IxContentHeader as IxContentHeader$1 } from '@siemens/ix/components/ix-content-header.js';
import { defineCustomElement as defineCustomElement$j, IxContent as IxContent$1 } from '@siemens/ix/components/ix-content.js';
import { defineCustomElement as defineCustomElement$l, IxCustomField as IxCustomField$1 } from '@siemens/ix/components/ix-custom-field.js';
import { defineCustomElement as defineCustomElement$m, IxDateDropdown as IxDateDropdown$1 } from '@siemens/ix/components/ix-date-dropdown.js';
import { defineCustomElement as defineCustomElement$n, IxDateInput as IxDateInput$1 } from '@siemens/ix/components/ix-date-input.js';
import { defineCustomElement as defineCustomElement$o, IxDatePicker as IxDatePicker$1 } from '@siemens/ix/components/ix-date-picker.js';
import { defineCustomElement as defineCustomElement$p, IxDatetimePicker as IxDatetimePicker$1 } from '@siemens/ix/components/ix-datetime-picker.js';
import { defineCustomElement as defineCustomElement$q, IxDivider as IxDivider$1 } from '@siemens/ix/components/ix-divider.js';
import { defineCustomElement as defineCustomElement$r, IxDrawer as IxDrawer$1 } from '@siemens/ix/components/ix-drawer.js';
import { defineCustomElement as defineCustomElement$t, IxDropdownButton as IxDropdownButton$1 } from '@siemens/ix/components/ix-dropdown-button.js';
import { defineCustomElement as defineCustomElement$u, IxDropdownHeader as IxDropdownHeader$1 } from '@siemens/ix/components/ix-dropdown-header.js';
import { defineCustomElement as defineCustomElement$v, IxDropdownItem as IxDropdownItem$1 } from '@siemens/ix/components/ix-dropdown-item.js';
import { defineCustomElement as defineCustomElement$w, IxDropdownQuickActions as IxDropdownQuickActions$1 } from '@siemens/ix/components/ix-dropdown-quick-actions.js';
import { defineCustomElement as defineCustomElement$s, IxDropdown as IxDropdown$1 } from '@siemens/ix/components/ix-dropdown.js';
import { defineCustomElement as defineCustomElement$x, IxEmptyState as IxEmptyState$1 } from '@siemens/ix/components/ix-empty-state.js';
import { defineCustomElement as defineCustomElement$z, IxEventListItem as IxEventListItem$1 } from '@siemens/ix/components/ix-event-list-item.js';
import { defineCustomElement as defineCustomElement$y, IxEventList as IxEventList$1 } from '@siemens/ix/components/ix-event-list.js';
import { defineCustomElement as defineCustomElement$A, IxExpandingSearch as IxExpandingSearch$1 } from '@siemens/ix/components/ix-expanding-search.js';
import { defineCustomElement as defineCustomElement$B, IxFieldLabel as IxFieldLabel$1 } from '@siemens/ix/components/ix-field-label.js';
import { defineCustomElement as defineCustomElement$C, IxFilterChip as IxFilterChip$1 } from '@siemens/ix/components/ix-filter-chip.js';
import { defineCustomElement as defineCustomElement$E, IxFlipTileContent as IxFlipTileContent$1 } from '@siemens/ix/components/ix-flip-tile-content.js';
import { defineCustomElement as defineCustomElement$D, IxFlipTile as IxFlipTile$1 } from '@siemens/ix/components/ix-flip-tile.js';
import { defineCustomElement as defineCustomElement$G, IxGroupContextMenu as IxGroupContextMenu$1 } from '@siemens/ix/components/ix-group-context-menu.js';
import { defineCustomElement as defineCustomElement$H, IxGroupItem as IxGroupItem$1 } from '@siemens/ix/components/ix-group-item.js';
import { defineCustomElement as defineCustomElement$F, IxGroup as IxGroup$1 } from '@siemens/ix/components/ix-group.js';
import { defineCustomElement as defineCustomElement$I, IxHelperText as IxHelperText$1 } from '@siemens/ix/components/ix-helper-text.js';
import { defineCustomElement as defineCustomElement$J, IxIconButton as IxIconButton$1 } from '@siemens/ix/components/ix-icon-button.js';
import { defineCustomElement as defineCustomElement$K, IxIconToggleButton as IxIconToggleButton$1 } from '@siemens/ix/components/ix-icon-toggle-button.js';
import { defineCustomElement as defineCustomElement$M, IxInputGroup as IxInputGroup$1 } from '@siemens/ix/components/ix-input-group.js';
import { defineCustomElement as defineCustomElement$L, IxInput as IxInput$1 } from '@siemens/ix/components/ix-input.js';
import { defineCustomElement as defineCustomElement$O, IxKeyValueList as IxKeyValueList$1 } from '@siemens/ix/components/ix-key-value-list.js';
import { defineCustomElement as defineCustomElement$N, IxKeyValue as IxKeyValue$1 } from '@siemens/ix/components/ix-key-value.js';
import { defineCustomElement as defineCustomElement$P, IxKpi as IxKpi$1 } from '@siemens/ix/components/ix-kpi.js';
import { defineCustomElement as defineCustomElement$Q, IxLayoutAuto as IxLayoutAuto$1 } from '@siemens/ix/components/ix-layout-auto.js';
import { defineCustomElement as defineCustomElement$R, IxLayoutGrid as IxLayoutGrid$1 } from '@siemens/ix/components/ix-layout-grid.js';
import { defineCustomElement as defineCustomElement$S, IxLinkButton as IxLinkButton$1 } from '@siemens/ix/components/ix-link-button.js';
import { defineCustomElement as defineCustomElement$V, IxMenuAboutItem as IxMenuAboutItem$1 } from '@siemens/ix/components/ix-menu-about-item.js';
import { defineCustomElement as defineCustomElement$W, IxMenuAboutNews as IxMenuAboutNews$1 } from '@siemens/ix/components/ix-menu-about-news.js';
import { defineCustomElement as defineCustomElement$U, IxMenuAbout as IxMenuAbout$1 } from '@siemens/ix/components/ix-menu-about.js';
import { defineCustomElement as defineCustomElement$Y, IxMenuAvatarItem as IxMenuAvatarItem$1 } from '@siemens/ix/components/ix-menu-avatar-item.js';
import { defineCustomElement as defineCustomElement$X, IxMenuAvatar as IxMenuAvatar$1 } from '@siemens/ix/components/ix-menu-avatar.js';
import { defineCustomElement as defineCustomElement$Z, IxMenuCategory as IxMenuCategory$1 } from '@siemens/ix/components/ix-menu-category.js';
import { defineCustomElement as defineCustomElement$_, IxMenuItem as IxMenuItem$1 } from '@siemens/ix/components/ix-menu-item.js';
import { defineCustomElement as defineCustomElement$10, IxMenuSettingsItem as IxMenuSettingsItem$1 } from '@siemens/ix/components/ix-menu-settings-item.js';
import { defineCustomElement as defineCustomElement$$, IxMenuSettings as IxMenuSettings$1 } from '@siemens/ix/components/ix-menu-settings.js';
import { defineCustomElement as defineCustomElement$T, IxMenu as IxMenu$1 } from '@siemens/ix/components/ix-menu.js';
import { defineCustomElement as defineCustomElement$11, IxMessageBar as IxMessageBar$1 } from '@siemens/ix/components/ix-message-bar.js';
import { defineCustomElement as defineCustomElement$12, IxModalContent as IxModalContent$1 } from '@siemens/ix/components/ix-modal-content.js';
import { defineCustomElement as defineCustomElement$13, IxModalFooter as IxModalFooter$1 } from '@siemens/ix/components/ix-modal-footer.js';
import { defineCustomElement as defineCustomElement$14, IxModalHeader as IxModalHeader$1 } from '@siemens/ix/components/ix-modal-header.js';
import { defineCustomElement, IxModal as IxModal$1 } from '@siemens/ix/components/ix-modal.js';
import { defineCustomElement as defineCustomElement$15, IxNumberInput as IxNumberInput$1 } from '@siemens/ix/components/ix-number-input.js';
import { defineCustomElement as defineCustomElement$16, IxPagination as IxPagination$1 } from '@siemens/ix/components/ix-pagination.js';
import { defineCustomElement as defineCustomElement$18, IxPaneLayout as IxPaneLayout$1 } from '@siemens/ix/components/ix-pane-layout.js';
import { defineCustomElement as defineCustomElement$17, IxPane as IxPane$1 } from '@siemens/ix/components/ix-pane.js';
import { defineCustomElement as defineCustomElement$19, IxPill as IxPill$1 } from '@siemens/ix/components/ix-pill.js';
import { defineCustomElement as defineCustomElement$1a, IxProgressIndicator as IxProgressIndicator$1 } from '@siemens/ix/components/ix-progress-indicator.js';
import { defineCustomElement as defineCustomElement$1b, IxPushCard as IxPushCard$1 } from '@siemens/ix/components/ix-push-card.js';
import { defineCustomElement as defineCustomElement$1d, IxRadioGroup as IxRadioGroup$1 } from '@siemens/ix/components/ix-radio-group.js';
import { defineCustomElement as defineCustomElement$1c, IxRadio as IxRadio$1 } from '@siemens/ix/components/ix-radio.js';
import { defineCustomElement as defineCustomElement$1e, IxRow as IxRow$1 } from '@siemens/ix/components/ix-row.js';
import { defineCustomElement as defineCustomElement$1g, IxSelectItem as IxSelectItem$1 } from '@siemens/ix/components/ix-select-item.js';
import { defineCustomElement as defineCustomElement$1f, IxSelect as IxSelect$1 } from '@siemens/ix/components/ix-select.js';
import { defineCustomElement as defineCustomElement$1h, IxSlider as IxSlider$1 } from '@siemens/ix/components/ix-slider.js';
import { defineCustomElement as defineCustomElement$1i, IxSpinner as IxSpinner$1 } from '@siemens/ix/components/ix-spinner.js';
import { defineCustomElement as defineCustomElement$1j, IxSplitButton as IxSplitButton$1 } from '@siemens/ix/components/ix-split-button.js';
import { defineCustomElement as defineCustomElement$1k, IxTabItem as IxTabItem$1 } from '@siemens/ix/components/ix-tab-item.js';
import { defineCustomElement as defineCustomElement$1l, IxTabs as IxTabs$1 } from '@siemens/ix/components/ix-tabs.js';
import { defineCustomElement as defineCustomElement$1m, IxTextarea as IxTextarea$1 } from '@siemens/ix/components/ix-textarea.js';
import { defineCustomElement as defineCustomElement$1n, IxTile as IxTile$1 } from '@siemens/ix/components/ix-tile.js';
import { defineCustomElement as defineCustomElement$1o, IxTimeInput as IxTimeInput$1 } from '@siemens/ix/components/ix-time-input.js';
import { defineCustomElement as defineCustomElement$1p, IxTimePicker as IxTimePicker$1 } from '@siemens/ix/components/ix-time-picker.js';
import { defineCustomElement as defineCustomElement$1r, IxToastContainer as IxToastContainer$1 } from '@siemens/ix/components/ix-toast-container.js';
import { defineCustomElement as defineCustomElement$1q, IxToast as IxToast$1 } from '@siemens/ix/components/ix-toast.js';
import { defineCustomElement as defineCustomElement$1t, IxToggleButton as IxToggleButton$1 } from '@siemens/ix/components/ix-toggle-button.js';
import { defineCustomElement as defineCustomElement$1s, IxToggle as IxToggle$1 } from '@siemens/ix/components/ix-toggle.js';
import { defineCustomElement as defineCustomElement$1u, IxTooltip as IxTooltip$1 } from '@siemens/ix/components/ix-tooltip.js';
import { defineCustomElement as defineCustomElement$1v, IxTypography as IxTypography$1 } from '@siemens/ix/components/ix-typography.js';
import { defineCustomElement as defineCustomElement$1w, IxUpload as IxUpload$1 } from '@siemens/ix/components/ix-upload.js';
import { defineCustomElement as defineCustomElement$1x, IxValidationTooltip as IxValidationTooltip$1 } from '@siemens/ix/components/ix-validation-tooltip.js';
import { defineCustomElement as defineCustomElement$1y, IxWorkflowStep as IxWorkflowStep$1 } from '@siemens/ix/components/ix-workflow-step.js';
import { defineCustomElement as defineCustomElement$1z, IxWorkflowSteps as IxWorkflowSteps$1 } from '@siemens/ix/components/ix-workflow-steps.js';
import { createComponent } from '@stencil/react-output-target/runtime';
import React from 'react';

const IxActionCard = /*@__PURE__*/ createComponent({
    tagName: 'ix-action-card',
    elementClass: IxActionCard$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$1
});
const IxApplication = /*@__PURE__*/ createComponent({
    tagName: 'ix-application',
    elementClass: IxApplication$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$2
});
const IxApplicationHeader = /*@__PURE__*/ createComponent({
    tagName: 'ix-application-header',
    elementClass: IxApplicationHeader$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onMenuToggle: 'menuToggle',
        onOpenAppSwitch: 'openAppSwitch'
    },
    defineCustomElement: defineCustomElement$3
});
const IxAvatar = /*@__PURE__*/ createComponent({
    tagName: 'ix-avatar',
    elementClass: IxAvatar$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$4
});
const IxBlind = /*@__PURE__*/ createComponent({
    tagName: 'ix-blind',
    elementClass: IxBlind$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onCollapsedChange: 'collapsedChange' },
    defineCustomElement: defineCustomElement$5
});
const IxBreadcrumb = /*@__PURE__*/ createComponent({
    tagName: 'ix-breadcrumb',
    elementClass: IxBreadcrumb$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onItemClick: 'itemClick',
        onNextClick: 'nextClick'
    },
    defineCustomElement: defineCustomElement$6
});
const IxBreadcrumbItem = /*@__PURE__*/ createComponent({
    tagName: 'ix-breadcrumb-item',
    elementClass: IxBreadcrumbItem$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$7
});
const IxButton = /*@__PURE__*/ createComponent({
    tagName: 'ix-button',
    elementClass: IxButton$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$8
});
const IxCard = /*@__PURE__*/ createComponent({
    tagName: 'ix-card',
    elementClass: IxCard$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$9
});
const IxCardAccordion = /*@__PURE__*/ createComponent({
    tagName: 'ix-card-accordion',
    elementClass: IxCardAccordion$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$a
});
const IxCardContent = /*@__PURE__*/ createComponent({
    tagName: 'ix-card-content',
    elementClass: IxCardContent$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$b
});
const IxCardList = /*@__PURE__*/ createComponent({
    tagName: 'ix-card-list',
    elementClass: IxCardList$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onCollapseChanged: 'collapseChanged',
        onShowAllClick: 'showAllClick',
        onShowMoreCardClick: 'showMoreCardClick'
    },
    defineCustomElement: defineCustomElement$c
});
const IxCardTitle = /*@__PURE__*/ createComponent({
    tagName: 'ix-card-title',
    elementClass: IxCardTitle$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$d
});
const IxCategoryFilter = /*@__PURE__*/ createComponent({
    tagName: 'ix-category-filter',
    elementClass: IxCategoryFilter$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onCategoryChanged: 'categoryChanged',
        onInputChanged: 'inputChanged',
        onFilterChanged: 'filterChanged',
        onFilterCleared: 'filterCleared'
    },
    defineCustomElement: defineCustomElement$e
});
const IxCheckbox = /*@__PURE__*/ createComponent({
    tagName: 'ix-checkbox',
    elementClass: IxCheckbox$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onCheckedChange: 'checkedChange',
        onValueChange: 'valueChange',
        onIxBlur: 'ixBlur'
    },
    defineCustomElement: defineCustomElement$f
});
const IxCheckboxGroup = /*@__PURE__*/ createComponent({
    tagName: 'ix-checkbox-group',
    elementClass: IxCheckboxGroup$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$g
});
const IxChip = /*@__PURE__*/ createComponent({
    tagName: 'ix-chip',
    elementClass: IxChip$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onCloseChip: 'closeChip' },
    defineCustomElement: defineCustomElement$h
});
const IxCol = /*@__PURE__*/ createComponent({
    tagName: 'ix-col',
    elementClass: IxCol$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$i
});
const IxContent = /*@__PURE__*/ createComponent({
    tagName: 'ix-content',
    elementClass: IxContent$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$j
});
const IxContentHeader = /*@__PURE__*/ createComponent({
    tagName: 'ix-content-header',
    elementClass: IxContentHeader$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onBackButtonClick: 'backButtonClick' },
    defineCustomElement: defineCustomElement$k
});
const IxCustomField = /*@__PURE__*/ createComponent({
    tagName: 'ix-custom-field',
    elementClass: IxCustomField$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$l
});
const IxDateDropdown = /*@__PURE__*/ createComponent({
    tagName: 'ix-date-dropdown',
    elementClass: IxDateDropdown$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onDateRangeChange: 'dateRangeChange' },
    defineCustomElement: defineCustomElement$m
});
const IxDateInput = /*@__PURE__*/ createComponent({
    tagName: 'ix-date-input',
    elementClass: IxDateInput$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onValueChange: 'valueChange',
        onValidityStateChange: 'validityStateChange'
    },
    defineCustomElement: defineCustomElement$n
});
const IxDatePicker = /*@__PURE__*/ createComponent({
    tagName: 'ix-date-picker',
    elementClass: IxDatePicker$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onDateChange: 'dateChange',
        onDateRangeChange: 'dateRangeChange',
        onDateSelect: 'dateSelect'
    },
    defineCustomElement: defineCustomElement$o
});
const IxDatetimePicker = /*@__PURE__*/ createComponent({
    tagName: 'ix-datetime-picker',
    elementClass: IxDatetimePicker$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onTimeChange: 'timeChange',
        onDateChange: 'dateChange',
        onDateSelect: 'dateSelect'
    },
    defineCustomElement: defineCustomElement$p
});
const IxDivider = /*@__PURE__*/ createComponent({
    tagName: 'ix-divider',
    elementClass: IxDivider$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$q
});
const IxDrawer = /*@__PURE__*/ createComponent({
    tagName: 'ix-drawer',
    elementClass: IxDrawer$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onOpen: 'open',
        onDrawerClose: 'drawerClose'
    },
    defineCustomElement: defineCustomElement$r
});
const IxDropdown = /*@__PURE__*/ createComponent({
    tagName: 'ix-dropdown',
    elementClass: IxDropdown$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onShowChanged: 'showChanged' },
    defineCustomElement: defineCustomElement$s
});
const IxDropdownButton = /*@__PURE__*/ createComponent({
    tagName: 'ix-dropdown-button',
    elementClass: IxDropdownButton$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$t
});
const IxDropdownHeader = /*@__PURE__*/ createComponent({
    tagName: 'ix-dropdown-header',
    elementClass: IxDropdownHeader$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$u
});
const IxDropdownItem = /*@__PURE__*/ createComponent({
    tagName: 'ix-dropdown-item',
    elementClass: IxDropdownItem$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$v
});
const IxDropdownQuickActions = /*@__PURE__*/ createComponent({
    tagName: 'ix-dropdown-quick-actions',
    elementClass: IxDropdownQuickActions$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$w
});
const IxEmptyState = /*@__PURE__*/ createComponent({
    tagName: 'ix-empty-state',
    elementClass: IxEmptyState$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onActionClick: 'actionClick' },
    defineCustomElement: defineCustomElement$x
});
const IxEventList = /*@__PURE__*/ createComponent({
    tagName: 'ix-event-list',
    elementClass: IxEventList$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$y
});
const IxEventListItem = /*@__PURE__*/ createComponent({
    tagName: 'ix-event-list-item',
    elementClass: IxEventListItem$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onItemClick: 'itemClick' },
    defineCustomElement: defineCustomElement$z
});
const IxExpandingSearch = /*@__PURE__*/ createComponent({
    tagName: 'ix-expanding-search',
    elementClass: IxExpandingSearch$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onValueChange: 'valueChange' },
    defineCustomElement: defineCustomElement$A
});
const IxFieldLabel = /*@__PURE__*/ createComponent({
    tagName: 'ix-field-label',
    elementClass: IxFieldLabel$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$B
});
const IxFilterChip = /*@__PURE__*/ createComponent({
    tagName: 'ix-filter-chip',
    elementClass: IxFilterChip$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onCloseClick: 'closeClick' },
    defineCustomElement: defineCustomElement$C
});
const IxFlipTile = /*@__PURE__*/ createComponent({
    tagName: 'ix-flip-tile',
    elementClass: IxFlipTile$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onToggle: 'toggle' },
    defineCustomElement: defineCustomElement$D
});
const IxFlipTileContent = /*@__PURE__*/ createComponent({
    tagName: 'ix-flip-tile-content',
    elementClass: IxFlipTileContent$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$E
});
const IxGroup = /*@__PURE__*/ createComponent({
    tagName: 'ix-group',
    elementClass: IxGroup$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onSelectGroup: 'selectGroup',
        onSelectItem: 'selectItem',
        onExpandedChanged: 'expandedChanged'
    },
    defineCustomElement: defineCustomElement$F
});
const IxGroupContextMenu = /*@__PURE__*/ createComponent({
    tagName: 'ix-group-context-menu',
    elementClass: IxGroupContextMenu$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$G
});
const IxGroupItem = /*@__PURE__*/ createComponent({
    tagName: 'ix-group-item',
    elementClass: IxGroupItem$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onSelectedChanged: 'selectedChanged' },
    defineCustomElement: defineCustomElement$H
});
const IxHelperText = /*@__PURE__*/ createComponent({
    tagName: 'ix-helper-text',
    elementClass: IxHelperText$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$I
});
const IxIconButton = /*@__PURE__*/ createComponent({
    tagName: 'ix-icon-button',
    elementClass: IxIconButton$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$J
});
const IxIconToggleButton = /*@__PURE__*/ createComponent({
    tagName: 'ix-icon-toggle-button',
    elementClass: IxIconToggleButton$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onPressedChange: 'pressedChange' },
    defineCustomElement: defineCustomElement$K
});
const IxInput = /*@__PURE__*/ createComponent({
    tagName: 'ix-input',
    elementClass: IxInput$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onValueChange: 'valueChange',
        onValidityStateChange: 'validityStateChange',
        onIxBlur: 'ixBlur'
    },
    defineCustomElement: defineCustomElement$L
});
const IxInputGroup = /*@__PURE__*/ createComponent({
    tagName: 'ix-input-group',
    elementClass: IxInputGroup$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$M
});
const IxKeyValue = /*@__PURE__*/ createComponent({
    tagName: 'ix-key-value',
    elementClass: IxKeyValue$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$N
});
const IxKeyValueList = /*@__PURE__*/ createComponent({
    tagName: 'ix-key-value-list',
    elementClass: IxKeyValueList$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$O
});
const IxKpi = /*@__PURE__*/ createComponent({
    tagName: 'ix-kpi',
    elementClass: IxKpi$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$P
});
const IxLayoutAuto = /*@__PURE__*/ createComponent({
    tagName: 'ix-layout-auto',
    elementClass: IxLayoutAuto$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$Q
});
const IxLayoutGrid = /*@__PURE__*/ createComponent({
    tagName: 'ix-layout-grid',
    elementClass: IxLayoutGrid$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$R
});
const IxLinkButton = /*@__PURE__*/ createComponent({
    tagName: 'ix-link-button',
    elementClass: IxLinkButton$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$S
});
const IxMenu = /*@__PURE__*/ createComponent({
    tagName: 'ix-menu',
    elementClass: IxMenu$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onExpandChange: 'expandChange',
        onMapExpandChange: 'mapExpandChange',
        onOpenAppSwitch: 'openAppSwitch',
        onOpenSettings: 'openSettings',
        onOpenAbout: 'openAbout'
    },
    defineCustomElement: defineCustomElement$T
});
const IxMenuAbout = /*@__PURE__*/ createComponent({
    tagName: 'ix-menu-about',
    elementClass: IxMenuAbout$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onTabChange: 'tabChange',
        onClose: 'close'
    },
    defineCustomElement: defineCustomElement$U
});
const IxMenuAboutItem = /*@__PURE__*/ createComponent({
    tagName: 'ix-menu-about-item',
    elementClass: IxMenuAboutItem$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onLabelChange: 'labelChange' },
    defineCustomElement: defineCustomElement$V
});
const IxMenuAboutNews = /*@__PURE__*/ createComponent({
    tagName: 'ix-menu-about-news',
    elementClass: IxMenuAboutNews$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onShowMore: 'showMore',
        onClosePopover: 'closePopover'
    },
    defineCustomElement: defineCustomElement$W
});
const IxMenuAvatar = /*@__PURE__*/ createComponent({
    tagName: 'ix-menu-avatar',
    elementClass: IxMenuAvatar$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onLogoutClick: 'logoutClick' },
    defineCustomElement: defineCustomElement$X
});
const IxMenuAvatarItem = /*@__PURE__*/ createComponent({
    tagName: 'ix-menu-avatar-item',
    elementClass: IxMenuAvatarItem$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onItemClick: 'itemClick' },
    defineCustomElement: defineCustomElement$Y
});
const IxMenuCategory = /*@__PURE__*/ createComponent({
    tagName: 'ix-menu-category',
    elementClass: IxMenuCategory$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$Z
});
const IxMenuItem = /*@__PURE__*/ createComponent({
    tagName: 'ix-menu-item',
    elementClass: IxMenuItem$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$_
});
const IxMenuSettings = /*@__PURE__*/ createComponent({
    tagName: 'ix-menu-settings',
    elementClass: IxMenuSettings$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onTabChange: 'tabChange',
        onClose: 'close'
    },
    defineCustomElement: defineCustomElement$$
});
const IxMenuSettingsItem = /*@__PURE__*/ createComponent({
    tagName: 'ix-menu-settings-item',
    elementClass: IxMenuSettingsItem$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onLabelChange: 'labelChange' },
    defineCustomElement: defineCustomElement$10
});
const IxMessageBar = /*@__PURE__*/ createComponent({
    tagName: 'ix-message-bar',
    elementClass: IxMessageBar$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onClosedChange: 'closedChange',
        onCloseAnimationCompleted: 'closeAnimationCompleted'
    },
    defineCustomElement: defineCustomElement$11
});
const IxModal = /*@__PURE__*/ createComponent({
    tagName: 'ix-modal',
    elementClass: IxModal$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onDialogClose: 'dialogClose',
        onDialogDismiss: 'dialogDismiss'
    },
    defineCustomElement: defineCustomElement
});
const IxModalContent = /*@__PURE__*/ createComponent({
    tagName: 'ix-modal-content',
    elementClass: IxModalContent$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$12
});
const IxModalFooter = /*@__PURE__*/ createComponent({
    tagName: 'ix-modal-footer',
    elementClass: IxModalFooter$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$13
});
const IxModalHeader = /*@__PURE__*/ createComponent({
    tagName: 'ix-modal-header',
    elementClass: IxModalHeader$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onCloseClick: 'closeClick' },
    defineCustomElement: defineCustomElement$14
});
const IxNumberInput = /*@__PURE__*/ createComponent({
    tagName: 'ix-number-input',
    elementClass: IxNumberInput$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onValueChange: 'valueChange',
        onValidityStateChange: 'validityStateChange',
        onIxBlur: 'ixBlur'
    },
    defineCustomElement: defineCustomElement$15
});
const IxPagination = /*@__PURE__*/ createComponent({
    tagName: 'ix-pagination',
    elementClass: IxPagination$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onPageSelected: 'pageSelected',
        onItemCountChanged: 'itemCountChanged'
    },
    defineCustomElement: defineCustomElement$16
});
const IxPane = /*@__PURE__*/ createComponent({
    tagName: 'ix-pane',
    elementClass: IxPane$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onExpandedChanged: 'expandedChanged',
        onVariantChanged: 'variantChanged',
        onBorderlessChanged: 'borderlessChanged'
    },
    defineCustomElement: defineCustomElement$17
});
const IxPaneLayout = /*@__PURE__*/ createComponent({
    tagName: 'ix-pane-layout',
    elementClass: IxPaneLayout$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$18
});
const IxPill = /*@__PURE__*/ createComponent({
    tagName: 'ix-pill',
    elementClass: IxPill$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$19
});
const IxProgressIndicator = /*@__PURE__*/ createComponent({
    tagName: 'ix-progress-indicator',
    elementClass: IxProgressIndicator$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$1a
});
const IxPushCard = /*@__PURE__*/ createComponent({
    tagName: 'ix-push-card',
    elementClass: IxPushCard$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$1b
});
const IxRadio = /*@__PURE__*/ createComponent({
    tagName: 'ix-radio',
    elementClass: IxRadio$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onCheckedChange: 'checkedChange',
        onValueChange: 'valueChange',
        onIxBlur: 'ixBlur'
    },
    defineCustomElement: defineCustomElement$1c
});
const IxRadioGroup = /*@__PURE__*/ createComponent({
    tagName: 'ix-radio-group',
    elementClass: IxRadioGroup$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onValueChange: 'valueChange' },
    defineCustomElement: defineCustomElement$1d
});
const IxRow = /*@__PURE__*/ createComponent({
    tagName: 'ix-row',
    elementClass: IxRow$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$1e
});
const IxSelect = /*@__PURE__*/ createComponent({
    tagName: 'ix-select',
    elementClass: IxSelect$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onValueChange: 'valueChange',
        onInputChange: 'inputChange',
        onAddItem: 'addItem',
        onIxBlur: 'ixBlur'
    },
    defineCustomElement: defineCustomElement$1f
});
const IxSelectItem = /*@__PURE__*/ createComponent({
    tagName: 'ix-select-item',
    elementClass: IxSelectItem$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onItemClick: 'itemClick' },
    defineCustomElement: defineCustomElement$1g
});
const IxSlider = /*@__PURE__*/ createComponent({
    tagName: 'ix-slider',
    elementClass: IxSlider$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onValueChange: 'valueChange' },
    defineCustomElement: defineCustomElement$1h
});
const IxSpinner = /*@__PURE__*/ createComponent({
    tagName: 'ix-spinner',
    elementClass: IxSpinner$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$1i
});
const IxSplitButton = /*@__PURE__*/ createComponent({
    tagName: 'ix-split-button',
    elementClass: IxSplitButton$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onButtonClick: 'buttonClick' },
    defineCustomElement: defineCustomElement$1j
});
const IxTabItem = /*@__PURE__*/ createComponent({
    tagName: 'ix-tab-item',
    elementClass: IxTabItem$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onTabClick: 'tabClick' },
    defineCustomElement: defineCustomElement$1k
});
const IxTabs = /*@__PURE__*/ createComponent({
    tagName: 'ix-tabs',
    elementClass: IxTabs$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onSelectedChange: 'selectedChange' },
    defineCustomElement: defineCustomElement$1l
});
const IxTextarea = /*@__PURE__*/ createComponent({
    tagName: 'ix-textarea',
    elementClass: IxTextarea$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onValueChange: 'valueChange',
        onValidityStateChange: 'validityStateChange',
        onIxBlur: 'ixBlur'
    },
    defineCustomElement: defineCustomElement$1m
});
const IxTile = /*@__PURE__*/ createComponent({
    tagName: 'ix-tile',
    elementClass: IxTile$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$1n
});
const IxTimeInput = /*@__PURE__*/ createComponent({
    tagName: 'ix-time-input',
    elementClass: IxTimeInput$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onValueChange: 'valueChange',
        onValidityStateChange: 'validityStateChange'
    },
    defineCustomElement: defineCustomElement$1o
});
const IxTimePicker = /*@__PURE__*/ createComponent({
    tagName: 'ix-time-picker',
    elementClass: IxTimePicker$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onTimeSelect: 'timeSelect',
        onTimeChange: 'timeChange'
    },
    defineCustomElement: defineCustomElement$1p
});
const IxToast = /*@__PURE__*/ createComponent({
    tagName: 'ix-toast',
    elementClass: IxToast$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onCloseToast: 'closeToast' },
    defineCustomElement: defineCustomElement$1q
});
const IxToastContainer = /*@__PURE__*/ createComponent({
    tagName: 'ix-toast-container',
    elementClass: IxToastContainer$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$1r
});
const IxToggle = /*@__PURE__*/ createComponent({
    tagName: 'ix-toggle',
    elementClass: IxToggle$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {
        onCheckedChange: 'checkedChange',
        onIxBlur: 'ixBlur'
    },
    defineCustomElement: defineCustomElement$1s
});
const IxToggleButton = /*@__PURE__*/ createComponent({
    tagName: 'ix-toggle-button',
    elementClass: IxToggleButton$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onPressedChange: 'pressedChange' },
    defineCustomElement: defineCustomElement$1t
});
const IxTooltip = /*@__PURE__*/ createComponent({
    tagName: 'ix-tooltip',
    elementClass: IxTooltip$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$1u
});
const IxTypography = /*@__PURE__*/ createComponent({
    tagName: 'ix-typography',
    elementClass: IxTypography$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$1v
});
const IxUpload = /*@__PURE__*/ createComponent({
    tagName: 'ix-upload',
    elementClass: IxUpload$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onFilesChanged: 'filesChanged' },
    defineCustomElement: defineCustomElement$1w
});
const IxValidationTooltip = /*@__PURE__*/ createComponent({
    tagName: 'ix-validation-tooltip',
    elementClass: IxValidationTooltip$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$1x
});
const IxWorkflowStep = /*@__PURE__*/ createComponent({
    tagName: 'ix-workflow-step',
    elementClass: IxWorkflowStep$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: {},
    defineCustomElement: defineCustomElement$1y
});
const IxWorkflowSteps = /*@__PURE__*/ createComponent({
    tagName: 'ix-workflow-steps',
    elementClass: IxWorkflowSteps$1,
    // @ts-ignore - ignore potential React type mismatches between the Stencil Output Target and your project.
    react: React,
    events: { onStepSelected: 'stepSelected' },
    defineCustomElement: defineCustomElement$1z
});

export { IxActionCard, IxApplication, IxApplicationHeader, IxAvatar, IxBlind, IxBreadcrumb, IxBreadcrumbItem, IxButton, IxCard, IxCardAccordion, IxCardContent, IxCardList, IxCardTitle, IxCategoryFilter, IxCheckbox, IxCheckboxGroup, IxChip, IxCol, IxContent, IxContentHeader, IxCustomField, IxDateDropdown, IxDateInput, IxDatePicker, IxDatetimePicker, IxDivider, IxDrawer, IxDropdown, IxDropdownButton, IxDropdownHeader, IxDropdownItem, IxDropdownQuickActions, IxEmptyState, IxEventList, IxEventListItem, IxExpandingSearch, IxFieldLabel, IxFilterChip, IxFlipTile, IxFlipTileContent, IxGroup, IxGroupContextMenu, IxGroupItem, IxHelperText, IxIconButton, IxIconToggleButton, IxInput, IxInputGroup, IxKeyValue, IxKeyValueList, IxKpi, IxLayoutAuto, IxLayoutGrid, IxLinkButton, IxMenu, IxMenuAbout, IxMenuAboutItem, IxMenuAboutNews, IxMenuAvatar, IxMenuAvatarItem, IxMenuCategory, IxMenuItem, IxMenuSettings, IxMenuSettingsItem, IxMessageBar, IxModal, IxModalContent, IxModalFooter, IxModalHeader, IxNumberInput, IxPagination, IxPane, IxPaneLayout, IxPill, IxProgressIndicator, IxPushCard, IxRadio, IxRadioGroup, IxRow, IxSelect, IxSelectItem, IxSlider, IxSpinner, IxSplitButton, IxTabItem, IxTabs, IxTextarea, IxTile, IxTimeInput, IxTimePicker, IxToast, IxToastContainer, IxToggle, IxToggleButton, IxTooltip, IxTypography, IxUpload, IxValidationTooltip, IxWorkflowStep, IxWorkflowSteps };
