// @siemens/ix-react/dist/components.server.js
import * as clientComponents from '@siemens/ix';
import { createComponent } from '@stencil/react-output-target/ssr';

/**
 * This file was automatically generated by the Stencil React Output Target.
 * Changes to this file may cause incorrect behavior and will be lost if the code is regenerated.
 */
/* eslint-disable */
// @ts-ignore - ignore potential type issues as the project is importing itself
const serializeShadowRoot = { "scoped": [], "default": "declarative-shadow-dom" };
const IxActionCard = /*@__PURE__*/ createComponent({
    tagName: 'ix-action-card',
    properties: {
        variant: 'variant',
        icon: 'icon',
        ariaLabelIcon: 'aria-label-icon',
        heading: 'heading',
        subheading: 'subheading',
        selected: 'selected',
        ariaLabelCard: 'aria-label-card'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxActionCard,
    serializeShadowRoot,
});
const IxApplication = /*@__PURE__*/ createComponent({
    tagName: 'ix-application',
    properties: {
        theme: 'theme',
        themeSystemAppearance: 'theme-system-appearance',
        forceBreakpoint: 'force-breakpoint',
        breakpoints: 'breakpoints',
        appSwitchConfig: 'app-switch-config'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxApplication,
    serializeShadowRoot,
});
const IxApplicationHeader = /*@__PURE__*/ createComponent({
    tagName: 'ix-application-header',
    properties: {
        name: 'name',
        nameSuffix: 'name-suffix',
        companyLogo: 'company-logo',
        companyLogoAlt: 'company-logo-alt',
        appIcon: 'app-icon',
        appIconAlt: 'app-icon-alt',
        appIconOutline: 'app-icon-outline',
        hideBottomBorder: 'hide-bottom-border',
        showMenu: 'show-menu',
        ariaLabelMenuExpandIconButton: 'aria-label-menu-expand-icon-button',
        ariaLabelAppSwitchIconButton: 'aria-label-app-switch-icon-button',
        ariaLabelMoreMenuIconButton: 'aria-label-more-menu-icon-button'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxApplicationHeader,
    serializeShadowRoot,
});
const IxAvatar = /*@__PURE__*/ createComponent({
    tagName: 'ix-avatar',
    properties: {
        a11yLabel: 'a11y-label',
        image: 'image',
        initials: 'initials',
        username: 'username',
        extra: 'extra',
        tooltipText: 'tooltip-text',
        ariaLabelTooltip: 'aria-label-tooltip'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxAvatar,
    serializeShadowRoot,
});
const IxBlind = /*@__PURE__*/ createComponent({
    tagName: 'ix-blind',
    properties: {
        collapsed: 'collapsed',
        label: 'label',
        sublabel: 'sublabel',
        icon: 'icon',
        variant: 'variant'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxBlind,
    serializeShadowRoot,
});
const IxBreadcrumb = /*@__PURE__*/ createComponent({
    tagName: 'ix-breadcrumb',
    properties: {
        visibleItemCount: 'visible-item-count',
        nextItems: 'next-items',
        subtle: 'subtle',
        ariaLabelPreviousButton: 'aria-label-previous-button'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxBreadcrumb,
    serializeShadowRoot,
});
const IxBreadcrumbItem = /*@__PURE__*/ createComponent({
    tagName: 'ix-breadcrumb-item',
    properties: {
        ariaLabelButton: 'aria-label-button',
        label: 'label',
        icon: 'icon',
        href: 'href',
        target: 'target',
        rel: 'rel',
        subtle: 'subtle',
        invisible: 'invisible',
        hideChevron: 'hide-chevron',
        isDropdownTrigger: 'is-dropdown-trigger'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxBreadcrumbItem,
    serializeShadowRoot,
});
const IxButton = /*@__PURE__*/ createComponent({
    tagName: 'ix-button',
    properties: {
        ariaLabelButton: 'aria-label-button',
        variant: 'variant',
        disabled: 'disabled',
        type: 'type',
        loading: 'loading',
        form: 'form',
        icon: 'icon',
        iconRight: 'icon-right',
        alignment: 'alignment',
        iconSize: 'icon-size',
        href: 'href',
        target: 'target',
        rel: 'rel'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxButton,
    serializeShadowRoot,
});
const IxCard = /*@__PURE__*/ createComponent({
    tagName: 'ix-card',
    properties: {
        variant: 'variant',
        selected: 'selected'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxCard,
    serializeShadowRoot,
});
const IxCardAccordion = /*@__PURE__*/ createComponent({
    tagName: 'ix-card-accordion',
    properties: {
        ariaLabelExpandButton: 'aria-label-expand-button',
        collapse: 'collapse',
        variant: 'variant'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxCardAccordion,
    serializeShadowRoot,
});
const IxCardContent = /*@__PURE__*/ createComponent({
    tagName: 'ix-card-content',
    properties: {},
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxCardContent,
    serializeShadowRoot,
});
const IxCardList = /*@__PURE__*/ createComponent({
    tagName: 'ix-card-list',
    properties: {
        ariaLabelExpandButton: 'aria-label-expand-button',
        label: 'label',
        collapse: 'collapse',
        listStyle: 'list-style',
        maxVisibleCards: 'max-visible-cards',
        showAllCount: 'show-all-count',
        suppressOverflowHandling: 'suppress-overflow-handling',
        hideShowAll: 'hide-show-all',
        i18nShowAll: 'i18n-show-all',
        i18nMoreCards: 'i18n-more-cards'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxCardList,
    serializeShadowRoot,
});
const IxCardTitle = /*@__PURE__*/ createComponent({
    tagName: 'ix-card-title',
    properties: {},
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxCardTitle,
    serializeShadowRoot,
});
const IxCategoryFilter = /*@__PURE__*/ createComponent({
    tagName: 'ix-category-filter',
    properties: {
        disabled: 'disabled',
        readonly: 'readonly',
        filterState: 'filter-state',
        placeholder: 'placeholder',
        categories: 'categories',
        nonSelectableCategories: 'non-selectable-categories',
        suggestions: 'suggestions',
        icon: 'icon',
        hideIcon: 'hide-icon',
        staticOperator: 'static-operator',
        uniqueCategories: 'unique-categories',
        labelCategories: 'label-categories',
        i18nPlainText: 'i18n-plain-text',
        ariaLabelResetButton: 'aria-label-reset-button',
        ariaLabelOperatorButton: 'aria-label-operator-button',
        ariaLabelFilterInput: 'aria-label-filter-input'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxCategoryFilter,
    serializeShadowRoot,
});
const IxCheckbox = /*@__PURE__*/ createComponent({
    tagName: 'ix-checkbox',
    properties: {
        name: 'name',
        value: 'value',
        label: 'label',
        checked: 'checked',
        disabled: 'disabled',
        indeterminate: 'indeterminate',
        required: 'required'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxCheckbox,
    serializeShadowRoot,
});
const IxCheckboxGroup = /*@__PURE__*/ createComponent({
    tagName: 'ix-checkbox-group',
    properties: {
        helperText: 'helper-text',
        label: 'label',
        direction: 'direction',
        invalidText: 'invalid-text',
        infoText: 'info-text',
        validText: 'valid-text',
        warningText: 'warning-text',
        showTextAsTooltip: 'show-text-as-tooltip',
        required: 'required'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxCheckboxGroup,
    serializeShadowRoot,
});
const IxChip = /*@__PURE__*/ createComponent({
    tagName: 'ix-chip',
    properties: {
        variant: 'variant',
        inactive: 'inactive',
        closable: 'closable',
        icon: 'icon',
        background: 'background',
        chipColor: 'chip-color',
        outline: 'outline',
        tooltipText: 'tooltip-text',
        centerContent: 'center-content',
        ariaLabelCloseButton: 'aria-label-close-button'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxChip,
    serializeShadowRoot,
});
const IxCol = /*@__PURE__*/ createComponent({
    tagName: 'ix-col',
    properties: {
        size: 'size',
        sizeSm: 'size-sm',
        sizeMd: 'size-md',
        sizeLg: 'size-lg'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxCol,
    serializeShadowRoot,
});
const IxContent = /*@__PURE__*/ createComponent({
    tagName: 'ix-content',
    properties: {},
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxContent,
    serializeShadowRoot,
});
const IxContentHeader = /*@__PURE__*/ createComponent({
    tagName: 'ix-content-header',
    properties: {
        variant: 'variant',
        headerTitle: 'header-title',
        headerSubtitle: 'header-subtitle',
        hasBackButton: 'has-back-button'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxContentHeader,
    serializeShadowRoot,
});
const IxCustomField = /*@__PURE__*/ createComponent({
    tagName: 'ix-custom-field',
    properties: {
        required: 'required',
        label: 'label',
        helperText: 'helper-text',
        infoText: 'info-text',
        warningText: 'warning-text',
        invalidText: 'invalid-text',
        validText: 'valid-text',
        showTextAsTooltip: 'show-text-as-tooltip'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxCustomField,
    serializeShadowRoot,
});
const IxDateDropdown = /*@__PURE__*/ createComponent({
    tagName: 'ix-date-dropdown',
    properties: {
        disabled: 'disabled',
        format: 'format',
        singleSelection: 'single-selection',
        from: 'from',
        to: 'to',
        minDate: 'min-date',
        maxDate: 'max-date',
        dateRangeId: 'date-range-id',
        variant: 'variant',
        loading: 'loading',
        showWeekNumbers: 'show-week-numbers',
        ariaLabelDropdownButton: 'aria-label-dropdown-button',
        customRangeDisabled: 'custom-range-disabled',
        dateRangeOptions: 'date-range-options',
        locale: 'locale',
        weekStartIndex: 'week-start-index',
        i18nCustomItem: 'i18n-custom-item',
        i18nDone: 'i18n-done',
        i18nNoRange: 'i18n-no-range',
        today: 'today'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxDateDropdown,
    serializeShadowRoot,
});
const IxDateInput = /*@__PURE__*/ createComponent({
    tagName: 'ix-date-input',
    properties: {
        name: 'name',
        placeholder: 'placeholder',
        value: 'value',
        minDate: 'min-date',
        maxDate: 'max-date',
        locale: 'locale',
        format: 'format',
        required: 'required',
        helperText: 'helper-text',
        label: 'label',
        ariaLabelCalendarButton: 'aria-label-calendar-button',
        invalidText: 'invalid-text',
        readonly: 'readonly',
        disabled: 'disabled',
        infoText: 'info-text',
        warningText: 'warning-text',
        validText: 'valid-text',
        showTextAsTooltip: 'show-text-as-tooltip',
        i18nErrorDateUnparsable: 'i18n-error-date-unparsable',
        showWeekNumbers: 'show-week-numbers',
        weekStartIndex: 'week-start-index',
        ariaLabelPreviousMonthButton: 'aria-label-previous-month-button',
        ariaLabelNextMonthButton: 'aria-label-next-month-button',
        suppressSubmitOnEnter: 'suppress-submit-on-enter',
        textAlignment: 'text-alignment'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxDateInput,
    serializeShadowRoot,
});
const IxDatePicker = /*@__PURE__*/ createComponent({
    tagName: 'ix-date-picker',
    properties: {
        format: 'format',
        singleSelection: 'single-selection',
        corners: 'corners',
        from: 'from',
        to: 'to',
        minDate: 'min-date',
        maxDate: 'max-date',
        i18nDone: 'i18n-done',
        ariaLabelPreviousMonthButton: 'aria-label-previous-month-button',
        ariaLabelNextMonthButton: 'aria-label-next-month-button',
        weekStartIndex: 'week-start-index',
        locale: 'locale',
        showWeekNumbers: 'show-week-numbers',
        embedded: 'embedded',
        today: 'today'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxDatePicker,
    serializeShadowRoot,
});
const IxDatetimePicker = /*@__PURE__*/ createComponent({
    tagName: 'ix-datetime-picker',
    properties: {
        singleSelection: 'single-selection',
        minDate: 'min-date',
        maxDate: 'max-date',
        dateFormat: 'date-format',
        timeFormat: 'time-format',
        from: 'from',
        to: 'to',
        time: 'time',
        showTimeReference: 'show-time-reference',
        timeReference: 'time-reference',
        i18nDone: 'i18n-done',
        i18nTime: 'i18n-time',
        ariaLabelPreviousMonthButton: 'aria-label-previous-month-button',
        ariaLabelNextMonthButton: 'aria-label-next-month-button',
        weekStartIndex: 'week-start-index',
        locale: 'locale',
        showWeekNumbers: 'show-week-numbers'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxDatetimePicker,
    serializeShadowRoot,
});
const IxDivider = /*@__PURE__*/ createComponent({
    tagName: 'ix-divider',
    properties: {},
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxDivider,
    serializeShadowRoot,
});
const IxDrawer = /*@__PURE__*/ createComponent({
    tagName: 'ix-drawer',
    properties: {
        show: 'show',
        closeOnClickOutside: 'close-on-click-outside',
        fullHeight: 'full-height',
        minWidth: 'min-width',
        maxWidth: 'max-width',
        width: 'width',
        ariaLabelCloseButton: 'aria-label-close-button'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxDrawer,
    serializeShadowRoot,
});
const IxDropdown = /*@__PURE__*/ createComponent({
    tagName: 'ix-dropdown',
    properties: {
        suppressAutomaticPlacement: 'suppress-automatic-placement',
        show: 'show',
        trigger: 'trigger',
        anchor: 'anchor',
        closeBehavior: 'close-behavior',
        placement: 'placement',
        positioningStrategy: 'positioning-strategy',
        header: 'header',
        offset: 'offset',
        overwriteDropdownStyle: 'overwrite-dropdown-style',
        discoverAllSubmenus: 'discover-all-submenus',
        ignoreRelatedSubmenu: 'ignore-related-submenu',
        suppressOverflowBehavior: 'suppress-overflow-behavior'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxDropdown,
    serializeShadowRoot,
});
const IxDropdownButton = /*@__PURE__*/ createComponent({
    tagName: 'ix-dropdown-button',
    properties: {
        variant: 'variant',
        disabled: 'disabled',
        label: 'label',
        icon: 'icon',
        closeBehavior: 'close-behavior',
        placement: 'placement',
        ariaLabelDropdownButton: 'aria-label-dropdown-button'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxDropdownButton,
    serializeShadowRoot,
});
const IxDropdownHeader = /*@__PURE__*/ createComponent({
    tagName: 'ix-dropdown-header',
    properties: { label: 'label' },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxDropdownHeader,
    serializeShadowRoot,
});
const IxDropdownItem = /*@__PURE__*/ createComponent({
    tagName: 'ix-dropdown-item',
    properties: {
        label: 'label',
        icon: 'icon',
        ariaLabelIcon: 'aria-label-icon',
        ariaLabelButton: 'aria-label-button',
        hover: 'hover',
        disabled: 'disabled',
        checked: 'checked',
        isSubMenu: 'is-sub-menu',
        suppressChecked: 'suppress-checked'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxDropdownItem,
    serializeShadowRoot,
});
const IxDropdownQuickActions = /*@__PURE__*/ createComponent({
    tagName: 'ix-dropdown-quick-actions',
    properties: {},
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxDropdownQuickActions,
    serializeShadowRoot,
});
const IxEmptyState = /*@__PURE__*/ createComponent({
    tagName: 'ix-empty-state',
    properties: {
        layout: 'layout',
        icon: 'icon',
        header: 'header',
        subHeader: 'sub-header',
        action: 'action',
        ariaLabelEmptyStateIcon: 'aria-label-empty-state-icon'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxEmptyState,
    serializeShadowRoot,
});
const IxEventList = /*@__PURE__*/ createComponent({
    tagName: 'ix-event-list',
    properties: {
        itemHeight: 'item-height',
        compact: 'compact',
        animated: 'animated',
        chevron: 'chevron'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxEventList,
    serializeShadowRoot,
});
const IxEventListItem = /*@__PURE__*/ createComponent({
    tagName: 'ix-event-list-item',
    properties: {
        variant: 'variant',
        itemColor: 'item-color',
        selected: 'selected',
        disabled: 'disabled',
        chevron: 'chevron'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxEventListItem,
    serializeShadowRoot,
});
const IxExpandingSearch = /*@__PURE__*/ createComponent({
    tagName: 'ix-expanding-search',
    properties: {
        icon: 'icon',
        placeholder: 'placeholder',
        value: 'value',
        fullWidth: 'full-width',
        variant: 'variant',
        ariaLabelSearchIconButton: 'aria-label-search-icon-button',
        ariaLabelClearIconButton: 'aria-label-clear-icon-button',
        ariaLabelSearchInput: 'aria-label-search-input'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxExpandingSearch,
    serializeShadowRoot,
});
const IxFieldLabel = /*@__PURE__*/ createComponent({
    tagName: 'ix-field-label',
    properties: {
        required: 'required',
        htmlFor: 'html-for',
        controlRef: 'control-ref',
        isInvalid: 'is-invalid'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxFieldLabel,
    serializeShadowRoot,
});
const IxFilterChip = /*@__PURE__*/ createComponent({
    tagName: 'ix-filter-chip',
    properties: {
        disabled: 'disabled',
        readonly: 'readonly',
        ariaLabelCloseIconButton: 'aria-label-close-icon-button'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxFilterChip,
    serializeShadowRoot,
});
const IxFlipTile = /*@__PURE__*/ createComponent({
    tagName: 'ix-flip-tile',
    properties: {
        variant: 'variant',
        height: 'height',
        width: 'width',
        index: 'index',
        ariaLabelEyeIconButton: 'aria-label-eye-icon-button'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxFlipTile,
    serializeShadowRoot,
});
const IxFlipTileContent = /*@__PURE__*/ createComponent({
    tagName: 'ix-flip-tile-content',
    properties: { contentVisible: 'content-visible' },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxFlipTileContent,
    serializeShadowRoot,
});
const IxGroup = /*@__PURE__*/ createComponent({
    tagName: 'ix-group',
    properties: {
        suppressHeaderSelection: 'suppress-header-selection',
        header: 'header',
        subHeader: 'sub-header',
        expanded: 'expanded',
        selected: 'selected',
        index: 'index',
        expandOnHeaderClick: 'expand-on-header-click'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxGroup,
    serializeShadowRoot,
});
const IxGroupContextMenu = /*@__PURE__*/ createComponent({
    tagName: 'ix-group-context-menu',
    properties: {},
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxGroupContextMenu,
    serializeShadowRoot,
});
const IxGroupItem = /*@__PURE__*/ createComponent({
    tagName: 'ix-group-item',
    properties: {
        icon: 'icon',
        ariaLabelIcon: 'aria-label-icon',
        text: 'text',
        secondaryText: 'secondary-text',
        suppressSelection: 'suppress-selection',
        groupFooter: 'group-footer',
        selected: 'selected',
        disabled: 'disabled',
        index: 'index'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxGroupItem,
    serializeShadowRoot,
});
const IxHelperText = /*@__PURE__*/ createComponent({
    tagName: 'ix-helper-text',
    properties: {
        htmlFor: 'html-for',
        helperText: 'helper-text',
        invalidText: 'invalid-text',
        validText: 'valid-text',
        infoText: 'info-text',
        warningText: 'warning-text'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxHelperText,
    serializeShadowRoot,
});
const IxIconButton = /*@__PURE__*/ createComponent({
    tagName: 'ix-icon-button',
    properties: {
        a11yLabel: 'a11y-label',
        variant: 'variant',
        oval: 'oval',
        icon: 'icon',
        size: 'size',
        iconColor: 'icon-color',
        disabled: 'disabled',
        type: 'type',
        loading: 'loading'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxIconButton,
    serializeShadowRoot,
});
const IxIconToggleButton = /*@__PURE__*/ createComponent({
    tagName: 'ix-icon-toggle-button',
    properties: {
        variant: 'variant',
        outline: 'outline',
        ghost: 'ghost',
        icon: 'icon',
        oval: 'oval',
        pressed: 'pressed',
        size: 'size',
        disabled: 'disabled',
        loading: 'loading',
        ariaLabelIconButton: 'aria-label-icon-button'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxIconToggleButton,
    serializeShadowRoot,
});
const IxInput = /*@__PURE__*/ createComponent({
    tagName: 'ix-input',
    properties: {
        type: 'type',
        name: 'name',
        placeholder: 'placeholder',
        value: 'value',
        required: 'required',
        disabled: 'disabled',
        readonly: 'readonly',
        helperText: 'helper-text',
        infoText: 'info-text',
        showTextAsTooltip: 'show-text-as-tooltip',
        validText: 'valid-text',
        warningText: 'warning-text',
        label: 'label',
        invalidText: 'invalid-text',
        pattern: 'pattern',
        maxLength: 'max-length',
        minLength: 'min-length',
        allowedCharactersPattern: 'allowed-characters-pattern',
        suppressSubmitOnEnter: 'suppress-submit-on-enter',
        textAlignment: 'text-alignment'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxInput,
    serializeShadowRoot,
});
const IxInputGroup = /*@__PURE__*/ createComponent({
    tagName: 'ix-input-group',
    properties: {},
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxInputGroup,
    serializeShadowRoot,
});
const IxKeyValue = /*@__PURE__*/ createComponent({
    tagName: 'ix-key-value',
    properties: {
        icon: 'icon',
        ariaLabelIcon: 'aria-label-icon',
        label: 'label',
        labelPosition: 'label-position',
        value: 'value'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxKeyValue,
    serializeShadowRoot,
});
const IxKeyValueList = /*@__PURE__*/ createComponent({
    tagName: 'ix-key-value-list',
    properties: { striped: 'striped' },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxKeyValueList,
    serializeShadowRoot,
});
const IxKpi = /*@__PURE__*/ createComponent({
    tagName: 'ix-kpi',
    properties: {
        label: 'label',
        ariaLabelAlarmIcon: 'aria-label-alarm-icon',
        ariaLabelWarningIcon: 'aria-label-warning-icon',
        value: 'value',
        unit: 'unit',
        state: 'state',
        orientation: 'orientation'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxKpi,
    serializeShadowRoot,
});
const IxLayoutAuto = /*@__PURE__*/ createComponent({
    tagName: 'ix-layout-auto',
    properties: { layout: 'layout' },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxLayoutAuto,
    serializeShadowRoot,
});
const IxLayoutGrid = /*@__PURE__*/ createComponent({
    tagName: 'ix-layout-grid',
    properties: {
        noMargin: 'no-margin',
        gap: 'gap',
        columns: 'columns'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxLayoutGrid,
    serializeShadowRoot,
});
const IxLinkButton = /*@__PURE__*/ createComponent({
    tagName: 'ix-link-button',
    properties: {
        disabled: 'disabled',
        url: 'url',
        target: 'target'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxLinkButton,
    serializeShadowRoot,
});
const IxMenu = /*@__PURE__*/ createComponent({
    tagName: 'ix-menu',
    properties: {
        showSettings: 'show-settings',
        showAbout: 'show-about',
        enableToggleTheme: 'enable-toggle-theme',
        enableMapExpand: 'enable-map-expand',
        applicationName: 'application-name',
        applicationDescription: 'application-description',
        i18nExpandSidebar: 'i18n-expand-sidebar',
        expand: 'expand',
        startExpanded: 'start-expanded',
        pinned: 'pinned',
        i18nLegal: 'i18n-legal',
        i18nSettings: 'i18n-settings',
        i18nToggleTheme: 'i18n-toggle-theme',
        i18nExpand: 'i18n-expand',
        i18nCollapse: 'i18n-collapse'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxMenu,
    serializeShadowRoot,
});
const IxMenuAbout = /*@__PURE__*/ createComponent({
    tagName: 'ix-menu-about',
    properties: {
        activeTabLabel: 'active-tab-label',
        label: 'label',
        show: 'show'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxMenuAbout,
    serializeShadowRoot,
});
const IxMenuAboutItem = /*@__PURE__*/ createComponent({
    tagName: 'ix-menu-about-item',
    properties: { label: 'label' },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxMenuAboutItem,
    serializeShadowRoot,
});
const IxMenuAboutNews = /*@__PURE__*/ createComponent({
    tagName: 'ix-menu-about-news',
    properties: {
        show: 'show',
        label: 'label',
        i18nShowMore: 'i18n-show-more',
        aboutItemLabel: 'about-item-label',
        expanded: 'expanded'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxMenuAboutNews,
    serializeShadowRoot,
});
const IxMenuAvatar = /*@__PURE__*/ createComponent({
    tagName: 'ix-menu-avatar',
    properties: {
        top: 'top',
        bottom: 'bottom',
        image: 'image',
        initials: 'initials',
        i18nLogout: 'i18n-logout',
        hideLogoutButton: 'hide-logout-button'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxMenuAvatar,
    serializeShadowRoot,
});
const IxMenuAvatarItem = /*@__PURE__*/ createComponent({
    tagName: 'ix-menu-avatar-item',
    properties: {
        icon: 'icon',
        label: 'label'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxMenuAvatarItem,
    serializeShadowRoot,
});
const IxMenuCategory = /*@__PURE__*/ createComponent({
    tagName: 'ix-menu-category',
    properties: {
        label: 'label',
        icon: 'icon',
        notifications: 'notifications',
        tooltipText: 'tooltip-text'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxMenuCategory,
    serializeShadowRoot,
});
const IxMenuItem = /*@__PURE__*/ createComponent({
    tagName: 'ix-menu-item',
    properties: {
        label: 'label',
        home: 'home',
        bottom: 'bottom',
        icon: 'icon',
        notifications: 'notifications',
        active: 'active',
        disabled: 'disabled',
        tooltipText: 'tooltip-text',
        href: 'href',
        target: 'target',
        rel: 'rel',
        isCategory: 'is-category'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxMenuItem,
    serializeShadowRoot,
});
const IxMenuSettings = /*@__PURE__*/ createComponent({
    tagName: 'ix-menu-settings',
    properties: {
        activeTabLabel: 'active-tab-label',
        label: 'label',
        show: 'show'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxMenuSettings,
    serializeShadowRoot,
});
const IxMenuSettingsItem = /*@__PURE__*/ createComponent({
    tagName: 'ix-menu-settings-item',
    properties: { label: 'label' },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxMenuSettingsItem,
    serializeShadowRoot,
});
const IxMessageBar = /*@__PURE__*/ createComponent({
    tagName: 'ix-message-bar',
    properties: {
        type: 'type',
        persistent: 'persistent'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxMessageBar,
    serializeShadowRoot,
});
const IxModal = /*@__PURE__*/ createComponent({
    tagName: 'ix-modal',
    properties: {
        size: 'size',
        disableAnimation: 'disable-animation',
        hideBackdrop: 'hide-backdrop',
        closeOnBackdropClick: 'close-on-backdrop-click',
        beforeDismiss: 'before-dismiss',
        centered: 'centered',
        disableEscapeClose: 'disable-escape-close'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxModal,
    serializeShadowRoot,
});
const IxModalContent = /*@__PURE__*/ createComponent({
    tagName: 'ix-modal-content',
    properties: {},
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxModalContent,
    serializeShadowRoot,
});
const IxModalFooter = /*@__PURE__*/ createComponent({
    tagName: 'ix-modal-footer',
    properties: {},
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxModalFooter,
    serializeShadowRoot,
});
const IxModalHeader = /*@__PURE__*/ createComponent({
    tagName: 'ix-modal-header',
    properties: {
        hideClose: 'hide-close',
        icon: 'icon',
        ariaLabelIcon: 'aria-label-icon',
        ariaLabelCloseIconButton: 'aria-label-close-icon-button',
        iconColor: 'icon-color'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxModalHeader,
    serializeShadowRoot,
});
const IxNumberInput = /*@__PURE__*/ createComponent({
    tagName: 'ix-number-input',
    properties: {
        name: 'name',
        placeholder: 'placeholder',
        value: 'value',
        required: 'required',
        disabled: 'disabled',
        readonly: 'readonly',
        helperText: 'helper-text',
        infoText: 'info-text',
        showTextAsTooltip: 'show-text-as-tooltip',
        validText: 'valid-text',
        warningText: 'warning-text',
        label: 'label',
        invalidText: 'invalid-text',
        pattern: 'pattern',
        min: 'min',
        max: 'max',
        allowedCharactersPattern: 'allowed-characters-pattern',
        showStepperButtons: 'show-stepper-buttons',
        step: 'step',
        suppressSubmitOnEnter: 'suppress-submit-on-enter',
        textAlignment: 'text-alignment',
        allowEmptyValueChange: 'allow-empty-value-change'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxNumberInput,
    serializeShadowRoot,
});
const IxPagination = /*@__PURE__*/ createComponent({
    tagName: 'ix-pagination',
    properties: {
        advanced: 'advanced',
        itemCount: 'item-count',
        hideItemCount: 'hide-item-count',
        count: 'count',
        selectedPage: 'selected-page',
        i18nPage: 'i18n-page',
        i18nOf: 'i18n-of',
        i18nItems: 'i18n-items',
        ariaLabelChevronLeftIconButton: 'aria-label-chevron-left-icon-button',
        ariaLabelChevronRightIconButton: 'aria-label-chevron-right-icon-button',
        ariaLabelPageSelection: 'aria-label-page-selection'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxPagination,
    serializeShadowRoot,
});
const IxPane = /*@__PURE__*/ createComponent({
    tagName: 'ix-pane',
    properties: {
        heading: 'heading',
        variant: 'variant',
        hideOnCollapse: 'hide-on-collapse',
        size: 'size',
        borderless: 'borderless',
        expanded: 'expanded',
        composition: 'composition',
        icon: 'icon',
        closeOnClickOutside: 'close-on-click-outside',
        ariaLabelIcon: 'aria-label-icon',
        ariaLabelCollapseCloseButton: 'aria-label-collapse-close-button',
        ignoreLayoutSettings: 'ignore-layout-settings',
        isMobile: 'is-mobile'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxPane,
    serializeShadowRoot,
});
const IxPaneLayout = /*@__PURE__*/ createComponent({
    tagName: 'ix-pane-layout',
    properties: {
        layout: 'layout',
        variant: 'variant',
        borderless: 'borderless'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxPaneLayout,
    serializeShadowRoot,
});
const IxPill = /*@__PURE__*/ createComponent({
    tagName: 'ix-pill',
    properties: {
        variant: 'variant',
        outline: 'outline',
        icon: 'icon',
        ariaLabelIcon: 'aria-label-icon',
        background: 'background',
        pillColor: 'pill-color',
        alignLeft: 'align-left',
        tooltipText: 'tooltip-text'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxPill,
    serializeShadowRoot,
});
const IxProgressIndicator = /*@__PURE__*/ createComponent({
    tagName: 'ix-progress-indicator',
    properties: {
        type: 'type',
        size: 'size',
        value: 'value',
        min: 'min',
        max: 'max',
        status: 'status',
        label: 'label',
        helperText: 'helper-text',
        textAlignment: 'text-alignment',
        showTextAsTooltip: 'show-text-as-tooltip'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxProgressIndicator,
    serializeShadowRoot,
});
const IxPushCard = /*@__PURE__*/ createComponent({
    tagName: 'ix-push-card',
    properties: {
        icon: 'icon',
        ariaLabelIcon: 'aria-label-icon',
        notification: 'notification',
        heading: 'heading',
        subheading: 'subheading',
        variant: 'variant',
        expanded: 'expanded'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxPushCard,
    serializeShadowRoot,
});
const IxRadio = /*@__PURE__*/ createComponent({
    tagName: 'ix-radio',
    properties: {
        name: 'name',
        value: 'value',
        label: 'label',
        disabled: 'disabled',
        checked: 'checked',
        required: 'required'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxRadio,
    serializeShadowRoot,
});
const IxRadioGroup = /*@__PURE__*/ createComponent({
    tagName: 'ix-radio-group',
    properties: {
        helperText: 'helper-text',
        label: 'label',
        value: 'value',
        invalidText: 'invalid-text',
        infoText: 'info-text',
        warningText: 'warning-text',
        validText: 'valid-text',
        showTextAsTooltip: 'show-text-as-tooltip',
        direction: 'direction',
        required: 'required'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxRadioGroup,
    serializeShadowRoot,
});
const IxRow = /*@__PURE__*/ createComponent({
    tagName: 'ix-row',
    properties: {},
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxRow,
    serializeShadowRoot,
});
const IxSelect = /*@__PURE__*/ createComponent({
    tagName: 'ix-select',
    properties: {
        name: 'name',
        required: 'required',
        label: 'label',
        ariaLabelChevronDownIconButton: 'aria-label-chevron-down-icon-button',
        ariaLabelClearIconButton: 'aria-label-clear-icon-button',
        warningText: 'warning-text',
        infoText: 'info-text',
        invalidText: 'invalid-text',
        validText: 'valid-text',
        helperText: 'helper-text',
        showTextAsTooltip: 'show-text-as-tooltip',
        value: 'value',
        allowClear: 'allow-clear',
        mode: 'mode',
        editable: 'editable',
        disabled: 'disabled',
        readonly: 'readonly',
        i18nPlaceholder: 'i18n-placeholder',
        i18nPlaceholderEditable: 'i18n-placeholder-editable',
        i18nSelectListHeader: 'i18n-select-list-header',
        i18nNoMatches: 'i18n-no-matches',
        i18nAllSelected: 'i18n-all-selected',
        hideListHeader: 'hide-list-header',
        dropdownWidth: 'dropdown-width',
        dropdownMaxWidth: 'dropdown-max-width',
        collapseMultipleSelection: 'collapse-multiple-selection'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxSelect,
    serializeShadowRoot,
});
const IxSelectItem = /*@__PURE__*/ createComponent({
    tagName: 'ix-select-item',
    properties: {
        label: 'label',
        value: 'value',
        selected: 'selected',
        hover: 'hover'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxSelectItem,
    serializeShadowRoot,
});
const IxSlider = /*@__PURE__*/ createComponent({
    tagName: 'ix-slider',
    properties: {
        step: 'step',
        min: 'min',
        max: 'max',
        value: 'value',
        marker: 'marker',
        trace: 'trace',
        traceReference: 'trace-reference',
        disabled: 'disabled',
        error: 'error'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxSlider,
    serializeShadowRoot,
});
const IxSpinner = /*@__PURE__*/ createComponent({
    tagName: 'ix-spinner',
    properties: {
        variant: 'variant',
        size: 'size',
        hideTrack: 'hide-track'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxSpinner,
    serializeShadowRoot,
});
const IxSplitButton = /*@__PURE__*/ createComponent({
    tagName: 'ix-split-button',
    properties: {
        variant: 'variant',
        closeBehavior: 'close-behavior',
        label: 'label',
        ariaLabelButton: 'aria-label-button',
        icon: 'icon',
        splitIcon: 'split-icon',
        ariaLabelSplitIconButton: 'aria-label-split-icon-button',
        disabled: 'disabled',
        disableButton: 'disable-button',
        disableDropdownButton: 'disable-dropdown-button',
        placement: 'placement'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxSplitButton,
    serializeShadowRoot,
});
const IxTabItem = /*@__PURE__*/ createComponent({
    tagName: 'ix-tab-item',
    properties: {
        selected: 'selected',
        disabled: 'disabled',
        small: 'small',
        icon: 'icon',
        rounded: 'rounded',
        counter: 'counter',
        layout: 'layout',
        placement: 'placement'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxTabItem,
    serializeShadowRoot,
});
const IxTabs = /*@__PURE__*/ createComponent({
    tagName: 'ix-tabs',
    properties: {
        small: 'small',
        rounded: 'rounded',
        selected: 'selected',
        layout: 'layout',
        placement: 'placement',
        ariaLabelChevronLeftIconButton: 'aria-label-chevron-left-icon-button',
        ariaLabelChevronRightIconButton: 'aria-label-chevron-right-icon-button'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxTabs,
    serializeShadowRoot,
});
const IxTextarea = /*@__PURE__*/ createComponent({
    tagName: 'ix-textarea',
    properties: {
        name: 'name',
        placeholder: 'placeholder',
        value: 'value',
        required: 'required',
        disabled: 'disabled',
        readonly: 'readonly',
        helperText: 'helper-text',
        infoText: 'info-text',
        showTextAsTooltip: 'show-text-as-tooltip',
        validText: 'valid-text',
        warningText: 'warning-text',
        label: 'label',
        invalidText: 'invalid-text',
        textareaHeight: 'textarea-height',
        textareaWidth: 'textarea-width',
        textareaRows: 'textarea-rows',
        textareaCols: 'textarea-cols',
        resizeBehavior: 'resize-behavior',
        maxLength: 'max-length',
        minLength: 'min-length'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxTextarea,
    serializeShadowRoot,
});
const IxTile = /*@__PURE__*/ createComponent({
    tagName: 'ix-tile',
    properties: { size: 'size' },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxTile,
    serializeShadowRoot,
});
const IxTimeInput = /*@__PURE__*/ createComponent({
    tagName: 'ix-time-input',
    properties: {
        name: 'name',
        placeholder: 'placeholder',
        value: 'value',
        format: 'format',
        required: 'required',
        helperText: 'helper-text',
        label: 'label',
        invalidText: 'invalid-text',
        readonly: 'readonly',
        disabled: 'disabled',
        infoText: 'info-text',
        warningText: 'warning-text',
        validText: 'valid-text',
        showTextAsTooltip: 'show-text-as-tooltip',
        i18nErrorTimeUnparsable: 'i18n-error-time-unparsable',
        hourInterval: 'hour-interval',
        minuteInterval: 'minute-interval',
        secondInterval: 'second-interval',
        millisecondInterval: 'millisecond-interval',
        i18nSelectTime: 'i18n-select-time',
        i18nTime: 'i18n-time',
        i18nHourColumnHeader: 'i18n-hour-column-header',
        i18nMinuteColumnHeader: 'i18n-minute-column-header',
        i18nSecondColumnHeader: 'i18n-second-column-header',
        i18nMillisecondColumnHeader: 'i18n-millisecond-column-header',
        suppressSubmitOnEnter: 'suppress-submit-on-enter',
        hideHeader: 'hide-header',
        textAlignment: 'text-alignment'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxTimeInput,
    serializeShadowRoot,
});
const IxTimePicker = /*@__PURE__*/ createComponent({
    tagName: 'ix-time-picker',
    properties: {
        format: 'format',
        corners: 'corners',
        embedded: 'embedded',
        dateTimePickerAppearance: 'date-time-picker-appearance',
        hideHeader: 'hide-header',
        hourInterval: 'hour-interval',
        minuteInterval: 'minute-interval',
        secondInterval: 'second-interval',
        millisecondInterval: 'millisecond-interval',
        time: 'time',
        i18nConfirmTime: 'i18n-confirm-time',
        i18nHeader: 'i18n-header',
        i18nHourColumnHeader: 'i18n-column-header',
        i18nMinuteColumnHeader: 'i18n-minute-column-header',
        i18nSecondColumnHeader: 'i18n-second-column-header',
        i18nMillisecondColumnHeader: 'i18n-millisecond-column-header'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxTimePicker,
    serializeShadowRoot,
});
const IxToast = /*@__PURE__*/ createComponent({
    tagName: 'ix-toast',
    properties: {
        type: 'type',
        toastTitle: 'toast-title',
        autoCloseDelay: 'auto-close-delay',
        preventAutoClose: 'prevent-auto-close',
        icon: 'icon',
        iconColor: 'icon-color',
        ariaLabelCloseIconButton: 'aria-label-close-icon-button'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxToast,
    serializeShadowRoot,
});
const IxToastContainer = /*@__PURE__*/ createComponent({
    tagName: 'ix-toast-container',
    properties: {
        containerId: 'container-id',
        containerClass: 'container-class',
        position: 'position'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxToastContainer,
    serializeShadowRoot,
});
const IxToggle = /*@__PURE__*/ createComponent({
    tagName: 'ix-toggle',
    properties: {
        name: 'name',
        value: 'value',
        checked: 'checked',
        disabled: 'disabled',
        indeterminate: 'indeterminate',
        textOn: 'text-on',
        textOff: 'text-off',
        textIndeterminate: 'text-indeterminate',
        hideText: 'hide-text',
        required: 'required'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxToggle,
    serializeShadowRoot,
});
const IxToggleButton = /*@__PURE__*/ createComponent({
    tagName: 'ix-toggle-button',
    properties: {
        variant: 'variant',
        disabled: 'disabled',
        loading: 'loading',
        icon: 'icon',
        iconRight: 'icon-right',
        pressed: 'pressed',
        ariaLabelButton: 'aria-label-button'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxToggleButton,
    serializeShadowRoot,
});
const IxTooltip = /*@__PURE__*/ createComponent({
    tagName: 'ix-tooltip',
    properties: {
        for: 'for',
        titleContent: 'title-content',
        interactive: 'interactive',
        placement: 'placement',
        showDelay: 'show-delay',
        hideDelay: 'hide-delay',
        animationFrame: 'animation-frame'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxTooltip,
    serializeShadowRoot,
});
const IxTypography = /*@__PURE__*/ createComponent({
    tagName: 'ix-typography',
    properties: {
        format: 'format',
        textColor: 'text-color',
        bold: 'bold',
        textDecoration: 'text-decoration'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxTypography,
    serializeShadowRoot,
});
const IxUpload = /*@__PURE__*/ createComponent({
    tagName: 'ix-upload',
    properties: {
        accept: 'accept',
        multiple: 'multiple',
        multiline: 'multiline',
        disabled: 'disabled',
        state: 'state',
        selectFileText: 'select-file-text',
        loadingText: 'loading-text',
        uploadFailedText: 'upload-failed-text',
        uploadSuccessText: 'upload-success-text',
        i18nUploadFile: 'i18n-upload-file',
        i18nUploadDisabled: 'i18n-upload-disabled'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxUpload,
    serializeShadowRoot,
});
const IxValidationTooltip = /*@__PURE__*/ createComponent({
    tagName: 'ix-validation-tooltip',
    properties: {
        message: 'message',
        placement: 'placement',
        suppressAutomaticPlacement: 'suppress-automatic-placement'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxValidationTooltip,
    serializeShadowRoot,
});
const IxWorkflowStep = /*@__PURE__*/ createComponent({
    tagName: 'ix-workflow-step',
    properties: {
        vertical: 'vertical',
        disabled: 'disabled',
        status: 'status',
        clickable: 'clickable',
        selected: 'selected',
        position: 'position'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxWorkflowStep,
    serializeShadowRoot,
});
const IxWorkflowSteps = /*@__PURE__*/ createComponent({
    tagName: 'ix-workflow-steps',
    properties: {
        vertical: 'vertical',
        clickable: 'clickable',
        selectedIndex: 'selected-index'
    },
    hydrateModule: import('@siemens/ix/hydrate'),
    clientModule: clientComponents.IxWorkflowSteps,
    serializeShadowRoot,
});

export { IxActionCard, IxApplication, IxApplicationHeader, IxAvatar, IxBlind, IxBreadcrumb, IxBreadcrumbItem, IxButton, IxCard, IxCardAccordion, IxCardContent, IxCardList, IxCardTitle, IxCategoryFilter, IxCheckbox, IxCheckboxGroup, IxChip, IxCol, IxContent, IxContentHeader, IxCustomField, IxDateDropdown, IxDateInput, IxDatePicker, IxDatetimePicker, IxDivider, IxDrawer, IxDropdown, IxDropdownButton, IxDropdownHeader, IxDropdownItem, IxDropdownQuickActions, IxEmptyState, IxEventList, IxEventListItem, IxExpandingSearch, IxFieldLabel, IxFilterChip, IxFlipTile, IxFlipTileContent, IxGroup, IxGroupContextMenu, IxGroupItem, IxHelperText, IxIconButton, IxIconToggleButton, IxInput, IxInputGroup, IxKeyValue, IxKeyValueList, IxKpi, IxLayoutAuto, IxLayoutGrid, IxLinkButton, IxMenu, IxMenuAbout, IxMenuAboutItem, IxMenuAboutNews, IxMenuAvatar, IxMenuAvatarItem, IxMenuCategory, IxMenuItem, IxMenuSettings, IxMenuSettingsItem, IxMessageBar, IxModal, IxModalContent, IxModalFooter, IxModalHeader, IxNumberInput, IxPagination, IxPane, IxPaneLayout, IxPill, IxProgressIndicator, IxPushCard, IxRadio, IxRadioGroup, IxRow, IxSelect, IxSelectItem, IxSlider, IxSpinner, IxSplitButton, IxTabItem, IxTabs, IxTextarea, IxTile, IxTimeInput, IxTimePicker, IxToast, IxToastContainer, IxToggle, IxToggleButton, IxTooltip, IxTypography, IxUpload, IxValidationTooltip, IxWorkflowStep, IxWorkflowSteps, serializeShadowRoot };
