// @siemens/ix-react/dist/delegate.js
import { registerFrameworkDelegate } from '@siemens/ix';
import ReactDOMClient from 'react-dom/client';

/*
 * SPDX-FileCopyrightText: 2024 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
let viewInstance = 0;
function createViewInstance() {
    return `ix-react-view-${viewInstance++}`;
}
const mountedRootNodes = {};
async function fallbackRootDom(id, view) {
    return new Promise((resolve) => {
        const rootElement = document.createElement('DIV');
        rootElement.id = id;
        rootElement.style.display = 'contents';
        document.body.appendChild(rootElement);
        const root = ReactDOMClient.createRoot(rootElement);
        root.render(view);
        mountedRootNodes[id] = root;
        setTimeout(() => {
            const viewElement = rootElement.children[0];
            resolve(viewElement);
        });
    });
}
async function fallbackRemoveViewFromRootDom(view) {
    const parent = view.parentElement;
    const id = parent.id;
    if (id in mountedRootNodes) {
        mountedRootNodes[id].unmount();
        delete mountedRootNodes[id];
        parent.remove();
    }
}
class ReactFrameworkDelegate {
    constructor() {
        this.isUsingReactPortal = false;
        this.portalInitPromise = new Promise((resolve) => (this.resolvePortalInitPromise = resolve));
    }
    async attachView(view) {
        const id = createViewInstance();
        if (!this.isUsingReactPortal) {
            return fallbackRootDom(id, view);
        }
        await this.isPortalReady();
        if (this.attachViewToPortal) {
            const refElement = await this.attachViewToPortal(id, view);
            return refElement;
        }
        console.error('Portal could not be initialized');
    }
    async removeView(view) {
        if (!this.removeViewFromPortal) {
            return fallbackRemoveViewFromRootDom(view);
        }
        const parent = view.parentElement;
        const id = parent.getAttribute('data-portal-id');
        this.removeViewFromPortal(id);
    }
    portalReady() {
        this.resolvePortalInitPromise?.();
    }
    async isPortalReady() {
        return this.portalInitPromise;
    }
}
const reactFrameworkDelegate = new ReactFrameworkDelegate();
registerFrameworkDelegate(reactFrameworkDelegate);

export { ReactFrameworkDelegate, reactFrameworkDelegate };
