// @siemens/ix/dist/collection/components/split-button/split-button.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host, } from "@stencil/core";
import { iconContextMenu } from "@siemens/ix-icons/icons";
import { makeRef } from "../utils/make-ref";
export class SplitButton {
    constructor() {
        /**
         * Color variant of button
         */
        this.variant = 'primary';
        /**
         * Controls if the dropdown will be closed in response to a click event depending on the position of the event relative to the dropdown.
         */
        this.closeBehavior = 'both';
        /**
         * Disabled
         */
        this.disabled = false;
        /**
         * Disables only the main button while keeping the dropdown trigger enabled
         *
         *  @since 4.1.0
         */
        this.disableButton = false;
        /**
         * Disables only the dropdown trigger while keeping the main button enabled
         *
         * @since 4.1.0
         */
        this.disableDropdownButton = false;
        /**
         * Placement of the dropdown
         */
        this.placement = 'bottom-start';
        this.toggle = false;
        this.triggerElementRef = makeRef();
    }
    get isDisabledButton() {
        return this.disabled || this.disableButton;
    }
    get isDisabledIcon() {
        return this.disabled || this.disableDropdownButton;
    }
    render() {
        var _a;
        const hasOutline = this.variant.toLocaleLowerCase().includes('secondary');
        const baseAttributes = {
            variant: this.variant,
        };
        const buttonAttributes = Object.assign(Object.assign({}, baseAttributes), { disabled: this.isDisabledButton, class: {
                'left-button-border': !hasOutline,
            } });
        const dropdownButtonAttributes = Object.assign(Object.assign({}, baseAttributes), { disabled: this.isDisabledIcon });
        return (h(Host, { key: '18955e55bbab4f5899f27cf2b38892c60bece302' }, h("div", { key: '8cc32b2686c3e02ee1f2a43d4a5f310cd2930dad', class: { 'btn-group': true, 'middle-gap': !hasOutline } }, this.label ? (h("ix-button", Object.assign({}, buttonAttributes, { icon: this.icon, onClick: (e) => this.buttonClick.emit(e), "aria-label": this.ariaLabelButton }), this.label)) : (h("ix-icon-button", Object.assign({}, buttonAttributes, { icon: this.icon, onClick: (e) => this.buttonClick.emit(e), "aria-label": this.ariaLabelButton }))), h("ix-icon-button", Object.assign({ key: 'ae7a2875314967f3100d1abcef3528a9822b0503' }, dropdownButtonAttributes, { ref: this.triggerElementRef, class: 'anchor', icon: (_a = this.splitIcon) !== null && _a !== void 0 ? _a : iconContextMenu, "aria-label": this.ariaLabelSplitIconButton }))), h("ix-dropdown", { key: 'd81abc1f01f55f2bc85740898e7f59173e62e182', closeBehavior: this.closeBehavior, trigger: this.triggerElementRef.waitForCurrent() }, h("slot", { key: '19c2623131a8ab3799e5a9bb75cc356bf500fffd' }))));
    }
    static get is() { return "ix-split-button"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["split-button.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["split-button.css"]
        };
    }
    static get properties() {
        return {
            "variant": {
                "type": "string",
                "attribute": "variant",
                "mutable": false,
                "complexType": {
                    "original": "SplitButtonVariant",
                    "resolved": "\"danger-primary\" | \"danger-secondary\" | \"danger-tertiary\" | \"primary\" | \"secondary\" | \"subtle-primary\" | \"subtle-secondary\" | \"subtle-tertiary\" | \"tertiary\"",
                    "references": {
                        "SplitButtonVariant": {
                            "location": "import",
                            "path": "./split-button.types",
                            "id": "src/components/split-button/split-button.types.ts::SplitButtonVariant"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Color variant of button"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'primary'"
            },
            "closeBehavior": {
                "type": "any",
                "attribute": "close-behavior",
                "mutable": false,
                "complexType": {
                    "original": "CloseBehavior",
                    "resolved": "\"both\" | \"inside\" | \"outside\" | boolean",
                    "references": {
                        "CloseBehavior": {
                            "location": "import",
                            "path": "../dropdown/dropdown-controller",
                            "id": "src/components/dropdown/dropdown-controller.ts::CloseBehavior"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Controls if the dropdown will be closed in response to a click event depending on the position of the event relative to the dropdown."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'both'"
            },
            "label": {
                "type": "string",
                "attribute": "label",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Button label"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "ariaLabelButton": {
                "type": "string",
                "attribute": "aria-label-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.2.0"
                        }],
                    "text": "ARIA label for the button (use if no label and icon button)"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "icon": {
                "type": "string",
                "attribute": "icon",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Button icon"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "splitIcon": {
                "type": "string",
                "attribute": "split-icon",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Icon of the button on the right"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "ariaLabelSplitIconButton": {
                "type": "string",
                "attribute": "aria-label-split-icon-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.2.0"
                        }],
                    "text": "ARIA label for the split icon button"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "disabled": {
                "type": "boolean",
                "attribute": "disabled",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Disabled"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "disableButton": {
                "type": "boolean",
                "attribute": "disable-button",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.1.0"
                        }],
                    "text": "Disables only the main button while keeping the dropdown trigger enabled"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "disableDropdownButton": {
                "type": "boolean",
                "attribute": "disable-dropdown-button",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.1.0"
                        }],
                    "text": "Disables only the dropdown trigger while keeping the main button enabled"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "placement": {
                "type": "string",
                "attribute": "placement",
                "mutable": false,
                "complexType": {
                    "original": "AlignedPlacement",
                    "resolved": "\"bottom-end\" | \"bottom-start\" | \"left-end\" | \"left-start\" | \"right-end\" | \"right-start\" | \"top-end\" | \"top-start\"",
                    "references": {
                        "AlignedPlacement": {
                            "location": "import",
                            "path": "../dropdown/placement",
                            "id": "src/components/dropdown/placement.ts::AlignedPlacement"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Placement of the dropdown"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'bottom-start'"
            }
        };
    }
    static get states() {
        return {
            "toggle": {}
        };
    }
    static get events() {
        return [{
                "method": "buttonClick",
                "name": "buttonClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Button clicked"
                },
                "complexType": {
                    "original": "MouseEvent",
                    "resolved": "MouseEvent",
                    "references": {
                        "MouseEvent": {
                            "location": "global",
                            "id": "global::MouseEvent"
                        }
                    }
                }
            }];
    }
    static get elementRef() { return "hostElement"; }
}
