// @siemens/ix/dist/collection/components/modal/modal.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
        r = Reflect.decorate(decorators, target, key, desc);
    else
        for (var i = decorators.length - 1; i >= 0; i--)
            if (d = decorators[i])
                r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
/*
 * SPDX-FileCopyrightText: 2024 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host, } from "@stencil/core";
import { animate } from "animejs";
import { a11yBoolean, a11yHostAttributes } from "../utils/a11y";
import Animation from "../utils/animation";
import { OnListener } from "../utils/listener";
import { waitForElement } from "../utils/waitForElement";
export class Modal {
    constructor() {
        this.ariaAttributes = {};
        this.isMouseDownInsideDialog = false;
        /**
         * Modal size
         */
        this.size = '360';
        /**
         * Should the modal animation be disabled
         */
        this.disableAnimation = false;
        /**
         * Hide the backdrop behind the modal dialog
         */
        this.hideBackdrop = false;
        /**
         * Dismiss modal on backdrop click
         */
        this.closeOnBackdropClick = false;
        /**
         * Centered modal
         */
        this.centered = false;
        /**
         * If set to true the modal cannot be closed by pressing the Escape key
         */
        this.disableEscapeClose = false;
        this.modalVisible = false;
    }
    onKey(e) {
        if (e.key === 'Escape') {
            e.preventDefault();
        }
    }
    get dialog() {
        return this.hostElement.shadowRoot.querySelector('dialog');
    }
    slideInModal() {
        const duration = this.disableAnimation ? 0 : Animation.mediumTime;
        const translateY = this.centered ? ['-90%', '-50%'] : [0, 40];
        animate(this.dialog, {
            duration,
            opacity: [0, 1],
            translateY,
            translateX: ['-50%', '-50%'],
            easing: 'easeOutSine',
        });
    }
    slideOutModal(completeCallback) {
        const duration = this.disableAnimation ? 0 : Animation.mediumTime;
        const translateY = this.centered ? ['-50%', '-90%'] : [40, 0];
        animate(this.dialog, {
            duration,
            opacity: [1, 0],
            translateY,
            translateX: ['-50%', '-50%'],
            easing: 'easeInSine',
            complete: () => {
                if (completeCallback) {
                    completeCallback();
                }
            },
        });
    }
    onMouseDown(event) {
        this.isMouseDownInsideDialog = this.isPointInsideDialog(event.clientX, event.clientY);
    }
    onMouseUp(event) {
        const isMouseUpInsideDialog = this.isPointInsideDialog(event.clientX, event.clientY);
        if (this.closeOnBackdropClick &&
            !this.isMouseDownInsideDialog &&
            !isMouseUpInsideDialog) {
            this.dismissModal();
        }
    }
    isPointInsideDialog(x, y) {
        const rect = this.dialog.getBoundingClientRect();
        return (x >= rect.left && x <= rect.right && y >= rect.top && y <= rect.bottom);
    }
    /**
     * Show the dialog
     */
    async showModal() {
        try {
            const dialog = await waitForElement('dialog', this.hostElement.shadowRoot);
            this.modalVisible = true;
            dialog.showModal();
        }
        catch (e) {
            console.error('HTMLDialogElement not existing');
        }
    }
    /**
     * Dismiss the dialog
     */
    async dismissModal(reason) {
        if (!this.modalVisible) {
            return;
        }
        let allowDismiss = true;
        if (this.beforeDismiss !== undefined) {
            allowDismiss = await this.beforeDismiss(reason);
        }
        if (!allowDismiss) {
            return;
        }
        this.slideOutModal(() => {
            this.modalVisible = false;
            this.dialog.close(JSON.stringify({
                type: 'dismiss',
                reason,
            }, null, 2));
            this.dialogDismiss.emit(reason);
        });
    }
    /**
     * Close the dialog
     */
    async closeModal(reason) {
        if (!this.modalVisible) {
            return;
        }
        this.slideOutModal(() => {
            this.modalVisible = false;
            this.dialog.close(JSON.stringify({
                type: 'close',
                reason,
            }, null, 2));
            this.dialogClose.emit(reason);
        });
    }
    componentDidLoad() {
        this.slideInModal();
    }
    componentWillLoad() {
        this.ariaAttributes = a11yHostAttributes(this.hostElement);
    }
    render() {
        return (h(Host, { key: '0170729f63356223c39bcd6b0bb7124b2036f2c4', class: {
                visible: this.modalVisible,
                'no-backdrop': this.hideBackdrop,
                'align-center': this.centered,
            } }, h("div", { key: '436b9f2efce7279defa936c9ee79fa44d090e388', class: "dialog-backdrop" }, h("dialog", { key: 'e3bc4891433993a5dcbe186011fd91d70277ca3d', "aria-modal": a11yBoolean(true), "aria-describedby": this.ariaAttributes['aria-describedby'], "aria-labelledby": this.ariaAttributes['aria-labelledby'], class: {
                modal: true,
                [`modal-size-${this.size}`]: true,
            }, onClose: () => this.dismissModal(), onMouseDown: (event) => this.onMouseDown(event), onMouseUp: (event) => this.onMouseUp(event), onCancel: (e) => {
                e.preventDefault();
                this.dismissModal();
            } }, h("slot", { key: '5a3e86b9c04f8e1564be6026b0352940978f0666' })))));
    }
    static get is() { return "ix-modal"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["modal.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modal.css"]
        };
    }
    static get properties() {
        return {
            "size": {
                "type": "string",
                "attribute": "size",
                "mutable": false,
                "complexType": {
                    "original": "IxModalSize",
                    "resolved": "\"360\" | \"480\" | \"600\" | \"720\" | \"840\" | \"full-screen\" | \"full-width\"",
                    "references": {
                        "IxModalSize": {
                            "location": "import",
                            "path": "./modal.types",
                            "id": "src/components/modal/modal.types.ts::IxModalSize"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Modal size"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'360'"
            },
            "disableAnimation": {
                "type": "boolean",
                "attribute": "disable-animation",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Should the modal animation be disabled"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "hideBackdrop": {
                "type": "boolean",
                "attribute": "hide-backdrop",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Hide the backdrop behind the modal dialog"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "closeOnBackdropClick": {
                "type": "boolean",
                "attribute": "close-on-backdrop-click",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Dismiss modal on backdrop click"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "beforeDismiss": {
                "type": "unknown",
                "attribute": "before-dismiss",
                "mutable": false,
                "complexType": {
                    "original": "(reason?: unknown) => boolean | Promise<boolean>",
                    "resolved": "((reason?: unknown) => boolean | Promise<boolean>) | undefined",
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Is called before the modal is dismissed.\n\n- Return `true` to proceed in dismissing the modal\n- Return `false` to abort in dismissing the modal"
                },
                "getter": false,
                "setter": false
            },
            "centered": {
                "type": "boolean",
                "attribute": "centered",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Centered modal"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "disableEscapeClose": {
                "type": "boolean",
                "attribute": "disable-escape-close",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "If set to true the modal cannot be closed by pressing the Escape key"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
    static get states() {
        return {
            "modalVisible": {}
        };
    }
    static get events() {
        return [{
                "method": "dialogClose",
                "name": "dialogClose",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Dialog close"
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }, {
                "method": "dialogDismiss",
                "name": "dialogDismiss",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Dialog cancel"
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }];
    }
    static get methods() {
        return {
            "showModal": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "HTMLDialogElement": {
                            "location": "global",
                            "id": "global::HTMLDialogElement"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Show the dialog",
                    "tags": []
                }
            },
            "dismissModal": {
                "complexType": {
                    "signature": "<T = unknown>(reason?: T) => Promise<void>",
                    "parameters": [{
                            "name": "reason",
                            "type": "T | undefined",
                            "docs": ""
                        }],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "T": {
                            "location": "global",
                            "id": "global::T"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Dismiss the dialog",
                    "tags": []
                }
            },
            "closeModal": {
                "complexType": {
                    "signature": "<T = unknown>(reason: T) => Promise<void>",
                    "parameters": [{
                            "name": "reason",
                            "type": "T",
                            "docs": ""
                        }],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "T": {
                            "location": "global",
                            "id": "global::T"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Close the dialog",
                    "tags": []
                }
            }
        };
    }
    static get elementRef() { return "hostElement"; }
}
__decorate([
    OnListener('keydown', (self) => self.disableEscapeClose)
], Modal.prototype, "onKey", null);
