// @siemens/ix/dist/collection/components/avatar/avatar.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Fragment, h, Host, readTask, } from "@stencil/core";
import { BaseButton } from "../button/base-button";
import { a11yBoolean, a11yHostAttributes } from "../utils/a11y";
import { makeRef } from "../utils/make-ref";
import { closestElement, hasSlottedElements } from "../utils/shadow-dom";
function DefaultAvatar(props) {
    const { initials } = props;
    if (initials) {
        return (h("div", { class: 'avatar-initials' }, h("ix-typography", { format: "label-lg" }, initials)));
    }
    return (h("svg", { class: 'avatar-image', xmlns: "http://www.w3.org/2000/svg", width: "32", height: "32", viewBox: "0 0 32 32", "aria-label": props.a11yLabel }, h("g", { fill: "none", "fill-rule": "evenodd" }, h("path", { id: "avatar-path-background", d: "M16 0C7.163 0 0 7.163 0 16s7.163 16 16 16 16-7.163\n          16-16c0-4.243-1.686-8.313-4.686-11.314C24.314 1.686 20.244 0 16 0z" }), h("path", { id: "avatar-path-person", d: "M17.897 17.91c3.8-.018 7.358 1.875 9.485 5.046-2.417 3.999-6.734 6.434-11.382\n        6.42-4.648.014-8.965-2.421-11.382-6.42 2.127-3.171 5.685-5.064\n        9.485-5.045h3.794zM15.821 2.129c3.682 0 6.667 2.984 6.667 6.666 0 3.682-2.985\n        6.667-6.667 6.667s-6.667-2.985-6.667-6.667 2.985-6.666 6.667-6.666z" }))));
}
function AvatarImage(props) {
    return (h("div", { class: "avatar" }, props.image ? (h("img", { src: props.image, class: "avatar-image", "aria-label": props.a11yLabel })) : (h(DefaultAvatar, { initials: props.initials, a11yLabel: props.a11yLabel }))));
}
function UserInfo(props) {
    return (h(Fragment, null, h("div", { class: "user-info", onClick: (event) => event.preventDefault() }, h(AvatarImage, { image: props.image, initials: props.initials, a11yLabel: props.a11yLabel }), h("div", { class: "user" }, h("div", { class: "username" }, props.userName), props.extra && (h("ix-typography", { class: "extra", "text-color": 'soft' }, props.extra))))));
}
export class Avatar {
    constructor() {
        this.isClosestApplicationHeader = false;
        this.hasSlottedElements = false;
        this.tooltipRef = makeRef();
    }
    componentWillLoad() {
        const closest = closestElement('ix-application-header', this.hostElement);
        this.isClosestApplicationHeader = closest !== null;
    }
    async slottedChanged() {
        this.hasSlottedElements = hasSlottedElements(this.slotElement);
    }
    resolveAvatarTrigger() {
        return new Promise((resolve, reject) => {
            readTask(() => {
                const button = this.hostElement.shadowRoot.querySelector('button');
                if (button) {
                    resolve(button);
                }
                else {
                    reject(new Error('ix-avatar - trigger element not found'));
                }
            });
        });
    }
    onDropdownClick(event) {
        if (event.target === this.dropdownElement) {
            event.preventDefault();
        }
    }
    render() {
        const a11y = a11yHostAttributes(this.hostElement);
        const a11yLabel = a11y['aria-label'];
        const tooltipText = this.tooltipText || this.username;
        const ariaHidden = tooltipText === this.username;
        const Avatar = (h(Fragment, null, h(AvatarImage, { image: this.image, initials: this.initials, a11yLabel: a11yLabel !== null && a11yLabel !== void 0 ? a11yLabel : this.a11yLabel }), !!tooltipText && (h("ix-tooltip", { ref: this.tooltipRef, for: this.hostElement, "aria-hidden": a11yBoolean(ariaHidden), "aria-label": this.ariaLabelTooltip }, tooltipText))));
        if (this.isClosestApplicationHeader) {
            return (h(Host, { slot: "ix-application-header-avatar", class: 'avatar-button' }, h(BaseButton, { disabled: false, iconOval: false, icon: undefined, iconOnly: false, loading: false, selected: false, type: "button", variant: "tertiary" }, Avatar), h("ix-dropdown", { ref: (ref) => (this.dropdownElement = ref), trigger: this.resolveAvatarTrigger(), class: "avatar-dropdown", onClick: (e) => this.onDropdownClick(e), onShowChanged: (event) => {
                    if (event.detail && this.tooltipRef.current) {
                        this.tooltipRef.current.hideTooltip(0);
                    }
                } }, this.username && (h(Fragment, null, h(UserInfo, { extra: this.extra, image: this.image, initials: this.initials, userName: this.username, a11yLabel: a11yLabel !== null && a11yLabel !== void 0 ? a11yLabel : this.a11yLabel }), this.hasSlottedElements && (h("ix-divider", { onClick: (e) => e.preventDefault() })))), h("slot", { onSlotchange: () => this.slottedChanged(), ref: (ref) => (this.slotElement = ref) }))));
        }
        return h(Host, null, Avatar);
    }
    static get is() { return "ix-avatar"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["avatar.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["avatar.css"]
        };
    }
    static get properties() {
        return {
            "a11yLabel": {
                "type": "string",
                "attribute": "a11y-label",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "deprecated",
                            "text": "Set the native `aria-label` on the ix-avatar host element. Will be removed in 5.0.0"
                        }],
                    "text": "Accessibility label for the image\nWill be set as aria-label on the nested HTML img element"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "image": {
                "type": "string",
                "attribute": "image",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Display an avatar image"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "initials": {
                "type": "string",
                "attribute": "initials",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Display the initials of the user. Will be overwritten by image"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "username": {
                "type": "string",
                "attribute": "username",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "If set an info card displaying the username will be placed inside the dropdown.\nNote: Only working if avatar is part of the ix-application-header"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "extra": {
                "type": "string",
                "attribute": "extra",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Optional description text that will be displayed underneath the username.\nNote: Only working if avatar is part of the ix-application-header"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "tooltipText": {
                "type": "string",
                "attribute": "tooltip-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "Text to display in a tooltip when hovering over the avatar"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "ariaLabelTooltip": {
                "type": "string",
                "attribute": "aria-label-tooltip",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "aria-label for the tooltip"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            }
        };
    }
    static get states() {
        return {
            "isClosestApplicationHeader": {},
            "hasSlottedElements": {}
        };
    }
    static get elementRef() { return "hostElement"; }
}
