// @siemens/ix/dist/collection/components/progress-indicator/progress-indicator.js
/*
 * SPDX-FileCopyrightText: 2025 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Fragment, Host, h } from "@stencil/core";
import { LinearBar } from "./linear";
import { iconCirclePause, iconError, iconInfo, iconSuccess, iconWarning, } from "@siemens/ix-icons/icons";
import { CircularProgress } from "./circular";
/**
 * @since 3.2.0
 */
export class ProgressIndicator {
    constructor() {
        /**
         * The type of progress indicator to use.
         */
        this.type = 'linear';
        /**
         * The size of the progress indicator.
         */
        this.size = 'md';
        /**
         * The value of the progress indicator.
         */
        this.value = 0;
        /**
         * The minimum value of the progress indicator.
         */
        this.min = 0;
        /**
         * The maximum value of the progress indicator.
         */
        this.max = 100;
        /**
         * The state of the progress indicator.
         * This is used to indicate the current state of the progress indicator.
         */
        this.status = 'default';
        /**
         * The text alignment for the helper text.
         * Can be 'left', 'center', or 'right'.
         */
        this.textAlignment = 'left';
        /**
         * Show the helper text as a tooltip
         */
        this.showTextAsTooltip = false;
    }
    getHelperText() {
        let icon = null;
        switch (this.status) {
            case 'error':
                icon = iconError;
                break;
            case 'info':
                icon = iconInfo;
                break;
            case 'warning':
                icon = iconWarning;
                break;
            case 'success':
                icon = iconSuccess;
                break;
            case 'paused':
                icon = iconCirclePause;
                break;
            default:
                icon = null;
        }
        if (!this.helperText) {
            return h("slot", { name: "helper-text" });
        }
        return (h("div", { class: {
                'helper-text': true,
                [this.status]: true,
            } }, icon && h("ix-icon", { name: icon, size: "16" }), h("div", { class: {
                text: true,
                'align-left': this.textAlignment === 'left',
                'align-center': this.textAlignment === 'center',
                'align-right': this.textAlignment === 'right',
            } }, this.helperText), h("slot", { name: "helper-text" })));
    }
    render() {
        const normalizedValue = ((this.value - this.min) / (this.max - this.min)) * 100;
        const clampedValue = Math.max(0, Math.min(normalizedValue, 100));
        return (h(Host, { key: '55cecc574b40c14bd8a4ba7a4d1f0e1d8d73f07b', class: {
                linear: this.type === 'linear',
                circular: this.type === 'circular',
            }, tabIndex: -1 }, h("div", { key: '3db1f913e4c4793106d7865b52ef4dcb07889188', class: {
                'progress-indicator': true,
                [this.size]: true,
                [this.status]: true,
                ['text-center']: this.textAlignment === 'center',
                ['text-left']: this.textAlignment === 'left',
                ['text-right']: this.textAlignment === 'right',
            } }, this.label && (h("ix-typography", { key: 'd8cd960dcfee8bfc20102d2b071661a186dcc854', format: "label", textColor: this.status === 'error' ? 'alarm' : 'soft', class: 'label' }, this.label)), h("div", { key: 'd149e99776e95f8051cc02e70d16d3c5103d98ba', class: "progress-container" }, this.type === 'linear' ? (h(Fragment, null, h(LinearBar, { value: clampedValue, status: this.status }), h("div", { class: "linear-slot" }, h("slot", null)))) : (h(CircularProgress, { status: this.status, alignment: this.textAlignment, value: clampedValue, size: this.size }, h("slot", null)))), this.showTextAsTooltip === true && this.helperText ? (h("ix-tooltip", { for: this.hostElement, showDelay: 500, placement: "bottom" }, this.getHelperText())) : (this.getHelperText()))));
    }
    static get is() { return "ix-progress-indicator"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["progress-indicator.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["progress-indicator.css"]
        };
    }
    static get properties() {
        return {
            "type": {
                "type": "string",
                "attribute": "type",
                "mutable": false,
                "complexType": {
                    "original": "'linear' | 'circular'",
                    "resolved": "\"circular\" | \"linear\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The type of progress indicator to use."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'linear'"
            },
            "size": {
                "type": "string",
                "attribute": "size",
                "mutable": false,
                "complexType": {
                    "original": "ProgressIndicatorSize",
                    "resolved": "\"lg\" | \"md\" | \"sm\" | \"xl\" | \"xs\"",
                    "references": {
                        "ProgressIndicatorSize": {
                            "location": "import",
                            "path": "./progress-indicator.types",
                            "id": "src/components/progress-indicator/progress-indicator.types.ts::ProgressIndicatorSize"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The size of the progress indicator."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'md'"
            },
            "value": {
                "type": "number",
                "attribute": "value",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The value of the progress indicator."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "0"
            },
            "min": {
                "type": "number",
                "attribute": "min",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The minimum value of the progress indicator."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "0"
            },
            "max": {
                "type": "number",
                "attribute": "max",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The maximum value of the progress indicator."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "100"
            },
            "status": {
                "type": "string",
                "attribute": "status",
                "mutable": false,
                "complexType": {
                    "original": "ProgressIndicatorStatus",
                    "resolved": "\"default\" | \"error\" | \"info\" | \"paused\" | \"success\" | \"warning\"",
                    "references": {
                        "ProgressIndicatorStatus": {
                            "location": "import",
                            "path": "./progress-indicator.types",
                            "id": "src/components/progress-indicator/progress-indicator.types.ts::ProgressIndicatorStatus"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The state of the progress indicator.\nThis is used to indicate the current state of the progress indicator."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'default'"
            },
            "label": {
                "type": "string",
                "attribute": "label",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The label for the progress indicator."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "helperText": {
                "type": "string",
                "attribute": "helper-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The helper text for the progress indicator."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "textAlignment": {
                "type": "string",
                "attribute": "text-alignment",
                "mutable": false,
                "complexType": {
                    "original": "'left' | 'center' | 'right'",
                    "resolved": "\"center\" | \"left\" | \"right\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The text alignment for the helper text.\nCan be 'left', 'center', or 'right'."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'left'"
            },
            "showTextAsTooltip": {
                "type": "boolean",
                "attribute": "show-text-as-tooltip",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show the helper text as a tooltip"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
    static get elementRef() { return "hostElement"; }
}
