// @siemens/ix/dist/collection/components/utils/input/picker-input.util.js
/*
 * SPDX-FileCopyrightText: 2025 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { dropdownController } from "../../dropdown/dropdown-controller";
export async function openDropdown(dropdownElementRef) {
    const dropdownElement = await dropdownElementRef.waitForCurrent();
    const id = dropdownElement.getAttribute('data-ix-dropdown');
    dropdownController.dismissAll();
    if (!id) {
        return;
    }
    const dropdown = dropdownController.getDropdownById(id);
    if (!dropdown) {
        return;
    }
    dropdownController.present(dropdown);
}
export async function closeDropdown(dropdownElementRef) {
    const dropdownElement = await dropdownElementRef.waitForCurrent();
    const id = dropdownElement.getAttribute('data-ix-dropdown');
    if (!id) {
        return;
    }
    const dropdown = dropdownController.getDropdownById(id);
    if (!dropdown) {
        return;
    }
    dropdownController.dismiss(dropdown);
}
export function handleIconClick(event, show, openDropdownFn, inputElementRef) {
    if (!show) {
        event.stopPropagation();
        event.preventDefault();
        openDropdownFn();
    }
    if (inputElementRef.current) {
        inputElementRef.current.focus();
    }
}
export function createValidityState(isInputInvalid, required, value) {
    return {
        badInput: false,
        customError: false,
        patternMismatch: isInputInvalid,
        rangeOverflow: false,
        rangeUnderflow: false,
        stepMismatch: false,
        tooLong: false,
        tooShort: false,
        typeMismatch: false,
        valid: !isInputInvalid,
        valueMissing: !!required && !value,
    };
}
