// @siemens/ix/dist/collection/tests/utils/test/page.js
/*
 * SPDX-FileCopyrightText: 2024 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { test as testBase, expect, } from "@playwright/test";
async function extendPageFixture(page, testInfo) {
    var _a, _b, _c, _d;
    const originalGoto = page.goto.bind(page);
    const originalScreenshot = page.screenshot.bind(page);
    const dataIxTheme = (_b = (_a = testInfo.project.metadata) === null || _a === void 0 ? void 0 : _a['data-ix-theme']) !== null && _b !== void 0 ? _b : 'classic';
    const dataIxColorSchema = (_d = (_c = testInfo.project.metadata) === null || _c === void 0 ? void 0 : _c['data-ix-color-schema']) !== null && _d !== void 0 ? _d : 'dark';
    testInfo.annotations.push({
        type: `${dataIxTheme} ${dataIxColorSchema}`,
    });
    page.goto = async (url, options) => {
        if (testInfo.componentTest === true) {
            return originalGoto(url, options);
        }
        const response = await originalGoto(`http://127.0.0.1:8080/src/tests/${url}?data-ix-theme=${dataIxTheme}&data-ix-color-schema=${dataIxColorSchema}`, options);
        await page.waitForTimeout(1000);
        return response;
    };
    page.screenshot = async (options) => {
        await page.waitForTimeout(150);
        return originalScreenshot(options);
    };
    return page;
}
async function mountComponent(page, selector, config) {
    return page.evaluateHandle(async ({ componentSelector, config }) => {
        if (config === null || config === void 0 ? void 0 : config.headTags) {
            config.headTags.forEach((tag) => {
                const head = document.querySelector('head');
                if (!head) {
                    throw new Error('No head tag found in the document.');
                }
                head.innerHTML += tag;
            });
        }
        if (config === null || config === void 0 ? void 0 : config.icons) {
            const moduleImport = await import(
            //@ts-expect-error - Only existing on runtime
            '/www/node_modules/@siemens/ix-icons/dist/index.js');
            const addIcons = moduleImport.addIcons;
            addIcons(config.icons);
        }
        const loadScript = document.createElement('script');
        loadScript.src = '/scripts/e2e/load-e2e-runtime.js';
        document.body.appendChild(loadScript);
        await new Promise((resolve) => {
            loadScript.onload = async () => {
                resolve();
            };
        });
        await window.customElements.whenDefined('ix-button');
        const mount = document.querySelector('#mount');
        if (!mount) {
            throw new Error('No mount point found in the document.');
        }
        mount.innerHTML = componentSelector;
        return mount.children.item(0);
    }, { componentSelector: selector, config });
}
export const regressionTest = testBase.extend({
    page: async ({ page }, use, testInfo) => {
        page = await extendPageFixture(page, testInfo);
        await page.route('*/**/svg/*.svg', async (route, request) => {
            if (!process.env.CI) {
                const [, svg] = request.url().split('/svg/');
                console.warn(testInfo.file, testInfo.title, 'SVGs fetched by static path', svg);
                expect(false, 'SVGs fetched by static path').toBe(true);
            }
            return route.continue();
        });
        await use(page);
    },
    createElement: async ({ page }, use) => {
        use((selector, appendTo) => page.evaluateHandle(async ({ selector, appendTo }) => {
            const elm = document.createElement(selector);
            if (appendTo) {
                appendTo.appendChild(elm);
            }
            return elm;
        }, {
            selector,
            appendTo,
        }));
    },
    mount: async ({ page }, use, testInfo) => {
        var _a, _b, _c, _d;
        testInfo.componentTest = true;
        const dataIxTheme = (_b = (_a = testInfo.project.metadata) === null || _a === void 0 ? void 0 : _a['data-ix-theme']) !== null && _b !== void 0 ? _b : 'classic';
        const dataIxColorSchema = (_d = (_c = testInfo.project.metadata) === null || _c === void 0 ? void 0 : _c['data-ix-color-schema']) !== null && _d !== void 0 ? _d : 'dark';
        testInfo.annotations.push({
            type: `${dataIxTheme} ${dataIxColorSchema}`,
        });
        await page.goto(`http://127.0.0.1:8080/src/tests/utils/ct/index.html?data-ix-theme=${dataIxTheme}&data-ix-color-schema=${dataIxColorSchema}`);
        await use((selector, config) => mountComponent(page, selector, config));
    },
});
export const test = regressionTest;
