// @siemens/ix/dist/collection/components/event-list/event-list.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host } from "@stencil/core";
import { createMutationObserver } from "../utils/mutation-observer";
import { convertToRemString } from "../utils/rwd.util";
import { animate } from "animejs";
export class EventList {
    constructor() {
        this.mutationObserver = createMutationObserver(this.onMutation.bind(this));
        /**
         * Determines the height of list items.
         * This can either be one of two predefined sizes ('S' or 'L') or an absolute pixel value.
         * In case a number is supplied it will get converted to rem internally.
         * Defaults to 'S'.
         */
        this.itemHeight = 'S';
        /**
         * Make event-list items more compact
         */
        this.compact = false;
        /**
         * Animate state change transitions. Defaults to 'false'.
         */
        this.animated = false;
        /**
         * Display a chevron icon in list items. Defaults to 'false'
         */
        this.chevron = false;
    }
    watchChevron(chevron) {
        this.handleChevron(chevron);
    }
    componentDidLoad() {
        if (this.animated) {
            this.triggerFadeIn();
        }
        if (!Number.isNaN(Number(this.itemHeight))) {
            const height = convertToRemString(this.itemHeight);
            this.hostElement
                .querySelectorAll('ix-event-list-item')
                .forEach((item) => {
                this.setCustomHeight(item, height);
            });
        }
        this.handleChevron(this.chevron);
        this.mutationObserver.observe(this.hostElement, {
            childList: true,
            subtree: true,
        });
    }
    onMutation(mutationRecords) {
        this.triggerFadeOut().then(() => {
            if (!Number.isNaN(Number(this.itemHeight))) {
                const height = convertToRemString(this.itemHeight);
                const eventListItems = this.findEventListItems(mutationRecords);
                eventListItems.forEach((item) => this.setCustomHeight(item, height));
            }
            this.handleChevron(this.chevron);
            this.triggerFadeIn();
        });
    }
    findEventListItems(mutationRecords) {
        const eventListItems = [];
        mutationRecords.forEach((mutation) => {
            if (mutation.type !== 'childList') {
                return;
            }
            mutation.addedNodes.forEach((node) => {
                const element = node;
                if (element.tagName === 'IX-EVENT-LIST-ITEM') {
                    eventListItems.push(element);
                }
            });
        });
        return eventListItems;
    }
    setCustomHeight(item, height) {
        item.style.setProperty('--event-list-item-height', height);
    }
    triggerFadeOut() {
        return new Promise((resolve) => {
            if (!this.animated) {
                resolve();
            }
            const listElement = this.hostElement.shadowRoot.querySelector('ul');
            animate(listElement, {
                opacity: [{ opacity: 1, easing: 'easeInSine' }, { opacity: 0 }],
                duration: EventList.fadeOutDuration,
                onComplete: () => {
                    resolve();
                },
            });
        });
    }
    triggerFadeIn() {
        if (!this.animated) {
            return;
        }
        const listItems = this.hostElement.querySelectorAll('ix-event-list-item');
        listItems.forEach((e, i) => {
            const delay = i * 80;
            const offset = delay / (delay + EventList.fadeInDuration);
            animate(e, {
                offset: offset,
                duration: EventList.fadeInDuration + delay,
                opacity: [0, 1],
                easing: 'easeInOutSine',
                delay: delay,
                autoplay: true,
            });
        });
    }
    handleChevron(chevron) {
        const listItems = this.hostElement.querySelectorAll('ix-event-list-item');
        listItems.forEach((e) => {
            if (chevron) {
                e.setAttribute('chevron', 'true');
            }
            else if (chevron !== undefined) {
                e.removeAttribute('chevron');
            }
        });
    }
    render() {
        return (h(Host, { key: '71648df2cdbbb1951fa970739cdf1a6f4bd1c40f', class: {
                'item-size-s': this.itemHeight === 'S',
                'item-size-l': this.itemHeight === 'L',
                compact: this.compact,
            } }, h("div", { key: 'ce46bd8d70b502d2dd19ffee75d2f3ae45e95d5a', role: "list" }, h("slot", { key: '016b05cd0e580af2e89232c3f4a0b5ebad8bda62' }))));
    }
    static get is() { return "ix-event-list"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["event-list.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["event-list.css"]
        };
    }
    static get properties() {
        return {
            "itemHeight": {
                "type": "any",
                "attribute": "item-height",
                "mutable": false,
                "complexType": {
                    "original": "'S' | 'L' | number",
                    "resolved": "\"L\" | \"S\" | number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Determines the height of list items.\nThis can either be one of two predefined sizes ('S' or 'L') or an absolute pixel value.\nIn case a number is supplied it will get converted to rem internally.\nDefaults to 'S'."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'S'"
            },
            "compact": {
                "type": "boolean",
                "attribute": "compact",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Make event-list items more compact"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "animated": {
                "type": "boolean",
                "attribute": "animated",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Animate state change transitions. Defaults to 'false'."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "chevron": {
                "type": "boolean",
                "attribute": "chevron",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Display a chevron icon in list items. Defaults to 'false'"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
    static get elementRef() { return "hostElement"; }
    static get watchers() {
        return [{
                "propName": "chevron",
                "methodName": "watchChevron"
            }];
    }
}
EventList.fadeOutDuration = 50;
EventList.fadeInDuration = 150;
