// @siemens/ix/dist/collection/components/tree-item/tree-item.js
/*
 * SPDX-FileCopyrightText: 2024 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host } from "@stencil/core";
import { iconChevronRightSmall } from "@siemens/ix-icons/icons";
export class TreeItem {
    constructor() {
        /**
         * Has tree item children
         */
        this.hasChildren = false;
    }
    render() {
        var _a, _b;
        return (h(Host, { key: '52817f9eed914ee6b01b67848af3f7073f21d65c', class: {
                selected: !!((_a = this.context) === null || _a === void 0 ? void 0 : _a.isSelected),
            } }, h("div", { key: '95b40939dac08653e875985a160e5aa45ebc353c', class: "icon-toggle-container" }, this.hasChildren ? (h("ix-icon", { name: iconChevronRightSmall, size: "24", class: {
                ['icon-toggle-down']: !!((_b = this.context) === null || _b === void 0 ? void 0 : _b.isExpanded),
            }, color: "color-std-text", onClick: (e) => {
                e.preventDefault();
                e.stopPropagation();
                this.toggle.emit();
            }, "aria-label": this.ariaLabelChevronIcon })) : null), h("div", { key: '656eda787f5fbf263ced33629d3ebde110c3f917', class: "tree-node-container", onClick: () => {
                this.itemClick.emit();
            } }, h("div", { key: 'eba55e54bf655fc0c3d0aacbfc3cb7e49fcb13f6', class: "tree-node-text" }, this.text), h("slot", { key: '4747766e632c2bdf7865b25b8d2abf82405c57c8' }))));
    }
    static get is() { return "ix-tree-item"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tree-item.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tree-item.css"]
        };
    }
    static get properties() {
        return {
            "text": {
                "type": "string",
                "attribute": "text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Text"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "hasChildren": {
                "type": "boolean",
                "attribute": "has-children",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Has tree item children"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "context": {
                "type": "unknown",
                "attribute": "context",
                "mutable": false,
                "complexType": {
                    "original": "TreeItemContext",
                    "resolved": "TreeItemContext | undefined",
                    "references": {
                        "TreeItemContext": {
                            "location": "import",
                            "path": "../tree/tree-model",
                            "id": "src/components/tree/tree-model.ts::TreeItemContext"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Context"
                },
                "getter": false,
                "setter": false
            },
            "ariaLabelChevronIcon": {
                "type": "string",
                "attribute": "aria-label-chevron-icon",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "ARIA label for the chevron icon"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            }
        };
    }
    static get events() {
        return [{
                "method": "toggle",
                "name": "toggle",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Expand/Collapsed toggled"
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }, {
                "method": "itemClick",
                "name": "itemClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Click on item not on the expand/collapse icon"
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }];
    }
}
