// @siemens/ix/dist/collection/components/date-time-card/date-time-card.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host } from "@stencil/core";
/**
 * @internal
 */
export class DateTimeCard {
    constructor() {
        /** Timepicker specific styling */
        this.timePickerAppearance = false;
        /**
         * Hide header
         */
        this.hideHeader = false;
        /**
         * Display footer
         */
        this.hasFooter = false;
        /**
         * Set corners style
         */
        this.corners = 'rounded';
    }
    cardClasses() {
        return {
            card: true,
            standaloneAppearance: this.embedded === false,
            rounded: this.corners === 'rounded',
            left: this.corners === 'left',
            right: this.corners === 'right',
            straight: this.corners === 'straight',
        };
    }
    render() {
        return (h(Host, { key: 'e4b044b0076f79934f3e561dab684a9aa53ab8c2' }, h("div", { key: '91f00686e9000d71745bca5bb958a401fe265ed2', class: this.cardClasses() }, !this.hideHeader && (h("div", { key: '974c90cd1522069f3e10831b6cec9d80eefc7630', class: "header-container" }, h("div", { key: '64d22e413cf579b3831580040d0921f73e37cbed', class: "header" }, h("slot", { key: 'dadc035aa3e2aad17d16a0b7edb9c1aa83f5b64a', name: "header" })), h("div", { key: '14c6cffd8a0086835c42dca7e5a12a9c569ebc2f', class: "separator" }))), h("div", { key: 'e3150b66b51fd87e5cb62c620d9e46d21a3b9aed', class: {
                content: true,
                'content--time-picker': this.timePickerAppearance,
            } }, h("slot", { key: '6fe2bfb5676a992be3caa5fd134dcb5f595f3ae3' })), this.hasFooter && (h("div", { key: '2c2a1a534b8e71a472e3a7e0bca7e5187322a107', class: "footer-container" }, h("div", { key: '67713b28bbfda3298d8778db37f8ca84d72270f0', class: "separator" }), h("div", { key: '247a0ac8c8fe7ddfa7b2d05c0bd9a1671229469c', class: "footer" }, h("slot", { key: 'b1f5a9ace8a8c69eff1a954850cca8174619dcc8', name: "footer" })))))));
    }
    static get is() { return "ix-date-time-card"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["date-time-card.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["date-time-card.css"]
        };
    }
    static get properties() {
        return {
            "embedded": {
                "type": "boolean",
                "attribute": "embedded",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }],
                    "text": ""
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "timePickerAppearance": {
                "type": "boolean",
                "attribute": "time-picker-appearance",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Timepicker specific styling"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "hideHeader": {
                "type": "boolean",
                "attribute": "hide-header",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Hide header"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "hasFooter": {
                "type": "boolean",
                "attribute": "has-footer",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Display footer"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "corners": {
                "type": "string",
                "attribute": "corners",
                "mutable": false,
                "complexType": {
                    "original": "DateTimeCardCorners",
                    "resolved": "\"left\" | \"right\" | \"rounded\" | \"straight\"",
                    "references": {
                        "DateTimeCardCorners": {
                            "location": "import",
                            "path": "./date-time-card.types",
                            "id": "src/components/date-time-card/date-time-card.types.ts::DateTimeCardCorners"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Set corners style"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'rounded'"
            }
        };
    }
}
