// @siemens/ix/dist/collection/components/menu-about-news/menu-about-news.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { iconClose, iconShout } from "@siemens/ix-icons/icons";
import { h, Host } from "@stencil/core";
export class MenuAboutNews {
    constructor() {
        /**
         * Show about news
         */
        this.show = false;
        /**
         * i18n label for 'Show more' button
         */
        this.i18nShowMore = 'Show more';
        /** @internal */
        this.expanded = false;
    }
    render() {
        return (h(Host, { key: '61a51cc04b238505c64285babebaaddf1201a9f7', class: {
                expanded: this.expanded,
                show: !this.show,
            } }, h("div", { key: 'c63c6cab269ef75f01d0396333c5c44ee271d354', class: "banner-container" }, h("ix-icon", { key: '2a3b5ced1c9f9fce0f742a34d9862e115807f6fb', color: "color-inv-contrast-text", name: iconShout, size: "32" }), h("svg", { key: '2433dd408f46a4d0f47e5cf075980b6a2ea2ece7', viewBox: "0 0 48 56", xmlns: "http://www.w3.org/2000/svg" }, h("polygon", { key: '386186a0847a504ca07052e1a49ee0c9f6314377', points: "0 0 48 0 48 56 24 48 0 56" }))), h("div", { key: 'ec280ef23c279507e186fa0ba0743e64e49d38c8', class: "cui-popover-news-header" }, h("ix-typography", { key: '784c5718192c7641e43004439cedcdc0f6285380', format: "label", bold: true }, this.label)), h("ix-icon-button", { key: '0c5a987a93067b0641ad0d7cd99e747aaa884d40', size: "24", icon: iconClose, iconColor: "color-soft-text", variant: "tertiary", onClick: () => {
                this.show = false;
                this.closePopover.emit();
            } }), h("div", { key: '780df8bab613bad3b80a3528947fdc0bab41187c', class: "slot-container" }, h("slot", { key: '1992c8d99353665ee6b88b4d21c34869284029d6' })), this.aboutItemLabel ? (h("div", { class: "cui-popover-news-footer" }, h("ix-button", { variant: "primary", onClick: (e) => {
                this.show = false;
                this.showMore.emit(e);
            } }, this.i18nShowMore))) : null, h("div", { key: 'b877e4087ebe56fca5c82a7acada5ba7c7ef8abb', id: "arrow" })));
    }
    static get is() { return "ix-menu-about-news"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["menu-about-news.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["menu-about-news.css"]
        };
    }
    static get properties() {
        return {
            "show": {
                "type": "boolean",
                "attribute": "show",
                "mutable": true,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show about news"
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "defaultValue": "false"
            },
            "label": {
                "type": "string",
                "attribute": "label",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Title of the about news"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "i18nShowMore": {
                "type": "string",
                "attribute": "i18n-show-more",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "i18n label for 'Show more' button"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'Show more'"
            },
            "aboutItemLabel": {
                "type": "string",
                "attribute": "about-item-label",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Subtitle of the about news"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "expanded": {
                "type": "boolean",
                "attribute": "expanded",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }],
                    "text": ""
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
    static get events() {
        return [{
                "method": "showMore",
                "name": "showMore",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Show More button is pressed"
                },
                "complexType": {
                    "original": "MouseEvent",
                    "resolved": "MouseEvent",
                    "references": {
                        "MouseEvent": {
                            "location": "global",
                            "id": "global::MouseEvent"
                        }
                    }
                }
            }, {
                "method": "closePopover",
                "name": "closePopover",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Popover closed"
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }];
    }
}
