// @siemens/ix/dist/collection/components/radio-group/radio-group.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
        r = Reflect.decorate(decorators, target, key, desc);
    else
        for (var i = decorators.length - 1; i >= 0; i--)
            if (d = decorators[i])
                r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Host, h, } from "@stencil/core";
import { HookValidationLifecycle, } from "../utils/input";
import { makeRef } from "../utils/make-ref";
/**
 * @form-ready
 */
export class RadiobuttonGroup {
    constructor() {
        /**
         * Alignment of the radio buttons in the group
         */
        this.direction = 'column';
        /**
         * Required state of the checkbox component
         *
         * @internal
         */
        this.required = false;
        this.isInvalid = false;
        this.isValid = false;
        this.isInfo = false;
        this.isWarning = false;
        this.touched = false;
        this.groupRef = makeRef();
        this.observer = new MutationObserver(() => {
            this.ensureOnlyLastRadioChecked();
            this.hasNestedRequiredRadio();
        });
    }
    get radiobuttonElements() {
        return Array.from(this.hostElement.querySelectorAll('ix-radio'));
    }
    connectedCallback() {
        this.observer.observe(this.hostElement, {
            childList: true,
            subtree: true,
            attributes: true,
            attributeFilter: ['checked', 'required'],
        });
    }
    componentWillLoad() {
        this.selectInitialValue();
        this.ensureOnlyLastRadioChecked();
        this.hasNestedRequiredRadio();
    }
    disconnectedCallback() {
        if (this.observer) {
            this.observer.disconnect();
        }
    }
    selectInitialValue() {
        if (!this.value) {
            return;
        }
        this.radiobuttonElements.forEach((radiobutton) => {
            radiobutton.checked = radiobutton.value === this.value;
        });
    }
    ensureOnlyLastRadioChecked() {
        const checkedRadios = this.radiobuttonElements.filter((radio) => radio.checked);
        checkedRadios.forEach((radio, index) => {
            if (index === checkedRadios.length - 1) {
                return;
            }
            radio.checked = false;
        });
        const hasCheckedRadio = this.isSomeRadioChecked();
        for (const radio of this.radiobuttonElements) {
            radio.tabIndex = radio.checked ? 0 : -1;
        }
        if (!hasCheckedRadio && this.radiobuttonElements.length > 0) {
            this.radiobuttonElements[0].tabIndex = 0;
        }
    }
    hasNestedRequiredRadio() {
        this.required = this.radiobuttonElements.some((radiobutton) => radiobutton.required);
    }
    isSomeRadioChecked() {
        return this.radiobuttonElements.some((radio) => radio.checked);
    }
    onValueChangeHandler(newValue) {
        this.touched = true;
        this.radiobuttonElements.forEach((radiobutton) => {
            radiobutton.checked = radiobutton.value === newValue;
        });
    }
    onCheckedChangeHandler(event) {
        this.radiobuttonElements.forEach((radiobutton) => {
            if (radiobutton !== event.target) {
                radiobutton.checked = false;
                return;
            }
            radiobutton.checked = true;
            this.valueChange.emit(radiobutton.value);
        });
    }
    onClassField({ isInvalid, isInfo, isValid, isWarning, isInvalidByRequired, }) {
        this.isInvalid = isInvalid || isInvalidByRequired;
        this.isInfo = isInfo;
        this.isValid = isValid;
        this.isWarning = isWarning;
    }
    /** @internal */
    hasValidValue() {
        return Promise.resolve(!!Array.from(this.hostElement.querySelectorAll('ix-radio')).find((radio) => radio.checked));
    }
    /** @internal */
    isTouched() {
        return Promise.resolve(this.touched);
    }
    /** @internal */
    async setCheckedToNextItem(currentRadio, forward = true) {
        const { radiobuttonElements } = this;
        const { length } = radiobuttonElements;
        if (length <= 1) {
            return;
        }
        const index = radiobuttonElements.indexOf(currentRadio);
        const step = forward ? 1 : -1;
        let nextIndex = (index + step + length) % length;
        while (radiobuttonElements[nextIndex].disabled) {
            if (nextIndex === index) {
                return;
            }
            nextIndex = (nextIndex + step + length) % length;
        }
        const nextRadio = radiobuttonElements[nextIndex];
        nextRadio.setCheckedState(true);
        nextRadio.focus();
    }
    render() {
        return (h(Host, { key: 'db3158ccabcd99d1b5185607b848c424a4c4c970', onIxBlur: () => (this.touched = true), ref: this.groupRef, role: "radiogroup" }, h("ix-field-wrapper", { key: '5668fbc12f4d264869dfe03dc2bbd3d380da6ae1', label: this.label, helperText: this.helperText, invalidText: this.invalidText, infoText: this.infoText, warningText: this.warningText, validText: this.validText, showTextAsTooltip: this.showTextAsTooltip, isValid: this.isValid, isInfo: this.isInfo, isWarning: this.isWarning, isInvalid: this.isInvalid, controlRef: this.groupRef }, h("div", { key: 'a5ec59c15eb8353d66bc6f0668dd7148dbd447a6', class: {
                'checkbox-container': true,
                'row-layout': this.direction === 'row',
            } }, h("slot", { key: '325b19034d55cc20cd556a24585d05ce32f22bb6' })))));
    }
    static get is() { return "ix-radio-group"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["radio-group.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["radio-group.css"]
        };
    }
    static get properties() {
        return {
            "helperText": {
                "type": "string",
                "attribute": "helper-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Show text below the field component"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "label": {
                "type": "string",
                "attribute": "label",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Label for the field component"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "value": {
                "type": "string",
                "attribute": "value",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Value of the radiobutton group component"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "invalidText": {
                "type": "string",
                "attribute": "invalid-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Error text for the field component"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "infoText": {
                "type": "string",
                "attribute": "info-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Info text for the field component"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "warningText": {
                "type": "string",
                "attribute": "warning-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Warning text for the field component"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "validText": {
                "type": "string",
                "attribute": "valid-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Valid text for the field component"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "showTextAsTooltip": {
                "type": "boolean",
                "attribute": "show-text-as-tooltip",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Show helper, info, warning, error and valid text as tooltip"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "direction": {
                "type": "string",
                "attribute": "direction",
                "mutable": false,
                "complexType": {
                    "original": "'column' | 'row'",
                    "resolved": "\"column\" | \"row\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Alignment of the radio buttons in the group"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'column'"
            },
            "required": {
                "type": "boolean",
                "attribute": "required",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }],
                    "text": "Required state of the checkbox component"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
    static get states() {
        return {
            "isInvalid": {},
            "isValid": {},
            "isInfo": {},
            "isWarning": {}
        };
    }
    static get events() {
        return [{
                "method": "valueChange",
                "name": "valueChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the value of the radiobutton group changes"
                },
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                }
            }];
    }
    static get methods() {
        return {
            "hasValidValue": {
                "complexType": {
                    "signature": "() => Promise<boolean>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<boolean>"
                },
                "docs": {
                    "text": "",
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }]
                }
            },
            "isTouched": {
                "complexType": {
                    "signature": "() => Promise<boolean>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<boolean>"
                },
                "docs": {
                    "text": "",
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }]
                }
            },
            "setCheckedToNextItem": {
                "complexType": {
                    "signature": "(currentRadio: HTMLIxRadioElement, forward?: boolean) => Promise<void>",
                    "parameters": [{
                            "name": "currentRadio",
                            "type": "HTMLIxRadioElement",
                            "docs": ""
                        }, {
                            "name": "forward",
                            "type": "boolean",
                            "docs": ""
                        }],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "HTMLIxRadioElement": {
                            "location": "global",
                            "id": "global::HTMLIxRadioElement"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "",
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }]
                }
            }
        };
    }
    static get elementRef() { return "hostElement"; }
    static get watchers() {
        return [{
                "propName": "value",
                "methodName": "onValueChangeHandler"
            }];
    }
    static get listeners() {
        return [{
                "name": "checkedChange",
                "method": "onCheckedChangeHandler",
                "target": undefined,
                "capture": false,
                "passive": false
            }];
    }
}
__decorate([
    HookValidationLifecycle({
        includeChildren: true,
    })
], RadiobuttonGroup.prototype, "onClassField", null);
