// @siemens/ix/dist/collection/components/key-value-list/key-value-list.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host } from "@stencil/core";
export class KeyValueList {
    constructor() {
        /**
         * Optional striped key value list style
         */
        this.striped = false;
    }
    render() {
        return (h(Host, { key: 'fcc3ec3470c25cac50b88a1006b30be059f778f8', class: { keyValueList: true, 'keyValueList--striped': this.striped } }, h("slot", { key: '05645a91f92d7db584bd8235f73979249870e734' })));
    }
    static get is() { return "ix-key-value-list"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["key-value-list.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["key-value-list.css"]
        };
    }
    static get properties() {
        return {
            "striped": {
                "type": "boolean",
                "attribute": "striped",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Optional striped key value list style"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
}
