// @siemens/ix/dist/collection/components/button/button.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host, } from "@stencil/core";
import { BaseButton } from "./base-button";
export class Button {
    constructor() {
        /**
         * Button variant
         */
        this.variant = 'primary';
        /**
         * Disable the button
         */
        this.disabled = false;
        /**
         * Type of the button
         */
        this.type = 'button';
        /**
         * Loading button
         */
        this.loading = false;
        /** @internal */
        this.alignment = 'center';
        /** @internal */
        this.iconSize = '24';
        /**
         * Specifies where to open the linked document when href is provided.
         *
         * @since 4.0.0
         */
        this.target = '_self';
    }
    handleClick(event) {
        if (this.disabled || this.loading) {
            event.stopPropagation();
            event.preventDefault();
        }
    }
    componentDidLoad() {
        if (this.type !== 'submit') {
            return;
        }
        const submitButton = document.createElement('button');
        submitButton.style.display = 'none';
        submitButton.type = 'submit';
        submitButton.tabIndex = -1;
        if (this.form) {
            submitButton.setAttribute('form', this.form);
        }
        this.hostElement.appendChild(submitButton);
        this.submitButtonElement = submitButton;
    }
    handleFormChange(newValue) {
        if (this.submitButtonElement) {
            if (newValue) {
                this.submitButtonElement.setAttribute('form', newValue);
            }
            else {
                this.submitButtonElement.removeAttribute('form');
            }
        }
    }
    componentDidRender() {
        if (this.submitButtonElement &&
            !this.hostElement.contains(this.submitButtonElement)) {
            this.hostElement.appendChild(this.submitButtonElement);
        }
    }
    dispatchFormEvents() {
        if (this.type === 'submit' &&
            this.submitButtonElement &&
            !this.disabled &&
            !this.loading) {
            this.submitButtonElement.click();
        }
    }
    setFocus() {
        var _a;
        (_a = this.hostElement
            .shadowRoot.querySelector(this.href ? 'a' : 'button')) === null || _a === void 0 ? void 0 : _a.focus();
    }
    render() {
        const baseButtonProps = {
            variant: this.variant,
            iconOnly: false,
            iconOval: false,
            selected: false,
            disabled: this.disabled || this.loading,
            icon: this.icon,
            iconRight: this.iconRight,
            iconSize: this.iconSize,
            loading: this.loading,
            onClick: () => this.dispatchFormEvents(),
            type: this.type,
            alignment: this.alignment,
            tabIndex: 0,
            ariaAttributes: {
                'aria-label': this.ariaLabelButton,
            },
            href: this.href,
            target: this.target,
            rel: this.rel,
        };
        return (h(Host, { key: 'd59ad1ad7693982ea33b27383a41edfca9a59430', class: {
                disabled: this.disabled || this.loading,
            } }, h(BaseButton, Object.assign({ key: 'd27e0296718f433c7df2896bd73b5991ef302c9a' }, baseButtonProps), h("slot", { key: 'c246c59dfd3406d389b9059020422f7a9779ab67' }))));
    }
    static get is() { return "ix-button"; }
    static get encapsulation() { return "shadow"; }
    static get delegatesFocus() { return true; }
    static get originalStyleUrls() {
        return {
            "$": ["./button.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["button.css"]
        };
    }
    static get properties() {
        return {
            "ariaLabelButton": {
                "type": "string",
                "attribute": "aria-label-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.2.0"
                        }],
                    "text": "ARIA label for the button\nWill be set as aria-label on the nested HTML button element"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "variant": {
                "type": "string",
                "attribute": "variant",
                "mutable": false,
                "complexType": {
                    "original": "ButtonVariant",
                    "resolved": "\"danger-primary\" | \"danger-secondary\" | \"danger-tertiary\" | \"primary\" | \"secondary\" | \"subtle-primary\" | \"subtle-secondary\" | \"subtle-tertiary\" | \"tertiary\"",
                    "references": {
                        "ButtonVariant": {
                            "location": "local",
                            "path": "/home/runner/work/ix/ix/packages/core/src/components/button/button.tsx",
                            "id": "src/components/button/button.tsx::ButtonVariant"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Button variant"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'primary'"
            },
            "disabled": {
                "type": "boolean",
                "attribute": "disabled",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Disable the button"
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "defaultValue": "false"
            },
            "type": {
                "type": "string",
                "attribute": "type",
                "mutable": false,
                "complexType": {
                    "original": "'button' | 'submit'",
                    "resolved": "\"button\" | \"submit\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Type of the button"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'button'"
            },
            "loading": {
                "type": "boolean",
                "attribute": "loading",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Loading button"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "form": {
                "type": "string",
                "attribute": "form",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.1.0"
                        }],
                    "text": "Provide a form element ID to automatically submit the from if the button is pressed. Only works in combination with type=\"submit\"."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "icon": {
                "type": "string",
                "attribute": "icon",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Icon name"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "iconRight": {
                "type": "string",
                "attribute": "icon-right",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "Icon name for the right side of the button"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "alignment": {
                "type": "string",
                "attribute": "alignment",
                "mutable": false,
                "complexType": {
                    "original": "'center' | 'start'",
                    "resolved": "\"center\" | \"start\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }],
                    "text": ""
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'center'"
            },
            "iconSize": {
                "type": "string",
                "attribute": "icon-size",
                "mutable": false,
                "complexType": {
                    "original": "'12' | '16' | '24'",
                    "resolved": "\"12\" | \"16\" | \"24\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }],
                    "text": ""
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'24'"
            },
            "href": {
                "type": "string",
                "attribute": "href",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "URL for the button link. When provided, the button will render as an anchor tag."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "target": {
                "type": "string",
                "attribute": "target",
                "mutable": false,
                "complexType": {
                    "original": "AnchorTarget",
                    "resolved": "\"_blank\" | \"_parent\" | \"_self\" | \"_top\" | undefined",
                    "references": {
                        "AnchorTarget": {
                            "location": "import",
                            "path": "./button.interface",
                            "id": "src/components/button/button.interface.ts::AnchorTarget"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "Specifies where to open the linked document when href is provided."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'_self'"
            },
            "rel": {
                "type": "string",
                "attribute": "rel",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "Specifies the relationship between the current document and the linked document when href is provided."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            }
        };
    }
    static get elementRef() { return "hostElement"; }
    static get watchers() {
        return [{
                "propName": "form",
                "methodName": "handleFormChange"
            }];
    }
    static get listeners() {
        return [{
                "name": "click",
                "method": "handleClick",
                "target": undefined,
                "capture": true,
                "passive": false
            }];
    }
}
