// @siemens/ix/dist/collection/components/breadcrumb-item/breadcrumb-item.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Fragment, h, Host, } from "@stencil/core";
import { animate } from "animejs";
import { BaseButton } from "../button/base-button";
import { a11yHostAttributes } from "../utils/a11y";
import { iconChevronRightSmall } from "@siemens/ix-icons/icons";
import Animation from "../utils/animation";
export class BreadcrumbItem {
    constructor() {
        /**
         * Specifies where to open the linked document when href is provided.
         *
         * @since 4.0.0
         */
        this.target = '_self';
        /**@internal */
        this.subtle = false;
        /**@internal */
        this.invisible = false;
        /**@internal */
        this.hideChevron = false;
        /** @internal */
        this.isDropdownTrigger = false;
    }
    componentDidLoad() {
        this.animationFadeIn();
    }
    componentWillLoad() {
        this.a11y = a11yHostAttributes(this.hostElement, [
            'aria-describedby',
            'aria-controls',
            'aria-expanded',
        ]);
    }
    animationFadeIn() {
        animate(this.hostElement, {
            duration: Animation.defaultTime,
            opacity: [0, 1],
            translateX: ['-100%', '0%'],
            easing: 'linear',
        });
    }
    render() {
        const props = {
            variant: this.subtle ? 'subtle-primary' : 'tertiary',
            iconOnly: false,
            iconOval: false,
            disabled: false,
            icon: this.icon,
            iconSize: '16',
            loading: false,
            selected: false,
            type: 'button',
            tabIndex: this.hostElement.tabIndex !== -1 ? this.hostElement.tabIndex : 0,
            extraClasses: {
                'dropdown-trigger': this.isDropdownTrigger,
            },
            ariaAttributes: Object.assign(Object.assign({}, this.a11y), { 'aria-label': this.ariaLabelButton }),
            href: this.href,
            target: this.target,
            rel: this.rel,
        };
        if (this.invisible) {
            return h(Host, { class: 'invisible' });
        }
        return (h(Host, { class: {
                'hide-chevron': this.hideChevron,
            }, onClick: () => this.itemClick.emit(this.label) }, h("li", null, h(BaseButton, Object.assign({}, props, { afterContent: h(Fragment, null, !this.hideChevron && (h("ix-icon", { name: iconChevronRightSmall, size: "16", class: 'chevron' }))) }), this.label, h("slot", null)))));
    }
    static get is() { return "ix-breadcrumb-item"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["breadcrumb-item.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["breadcrumb-item.css"]
        };
    }
    static get properties() {
        return {
            "ariaLabelButton": {
                "type": "string",
                "attribute": "aria-label-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.2.0"
                        }],
                    "text": "ARIA label for the button\nWill be set as aria-label for the nested HTML button element"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "label": {
                "type": "string",
                "attribute": "label",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Breadcrumb label"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "icon": {
                "type": "string",
                "attribute": "icon",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Icon to be displayed next ot the label"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "href": {
                "type": "string",
                "attribute": "href",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "URL for the button link. When provided, the button will render as an anchor tag."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "target": {
                "type": "string",
                "attribute": "target",
                "mutable": false,
                "complexType": {
                    "original": "AnchorTarget",
                    "resolved": "\"_blank\" | \"_parent\" | \"_self\" | \"_top\" | undefined",
                    "references": {
                        "AnchorTarget": {
                            "location": "import",
                            "path": "../button/button.interface",
                            "id": "src/components/button/button.interface.ts::AnchorTarget"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "Specifies where to open the linked document when href is provided."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'_self'"
            },
            "rel": {
                "type": "string",
                "attribute": "rel",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "Specifies the relationship between the current document and the linked document when href is provided."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "subtle": {
                "type": "boolean",
                "attribute": "subtle",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }],
                    "text": ""
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "invisible": {
                "type": "boolean",
                "attribute": "invisible",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }],
                    "text": ""
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "hideChevron": {
                "type": "boolean",
                "attribute": "hide-chevron",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }],
                    "text": ""
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "isDropdownTrigger": {
                "type": "boolean",
                "attribute": "is-dropdown-trigger",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }],
                    "text": ""
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
    static get states() {
        return {
            "a11y": {}
        };
    }
    static get events() {
        return [{
                "method": "itemClick",
                "name": "itemClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }],
                    "text": ""
                },
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "hostElement"; }
}
