// @siemens/ix/dist/collection/components/chip/chip.js
/*
 * SPDX-FileCopyrightText: 2024 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { iconCloseSmall } from "@siemens/ix-icons/icons";
import { h, Host, } from "@stencil/core";
import { makeRef } from "../utils/make-ref";
export class Chip {
    constructor() {
        /**
         * Chip variant
         */
        this.variant = 'primary';
        /**
         * Determines if the chip is interactive. If true no user input (e.g. mouse states, keyboard navigation)
         * will be possible and also the close button will not be present.
         */
        this.inactive = false;
        /**
         * Show close icon
         */
        this.closable = false;
        /**
         * Show chip with outline style
         */
        this.outline = false;
        /**
         * Display a tooltip. By default, no tooltip will be displayed.
         * Add the attribute to display the text content of the component as a tooltip or use a string to display a custom text.
         *
         * @since 3.0.0
         */
        this.tooltipText = false;
        /**
         * Center the content of the chip.
         * Set to false to disable centering.
         * @since 3.2.0
         */
        this.centerContent = false;
        this.containerElementRef = makeRef();
    }
    getCloseButton() {
        return (h("div", { class: "close-button-container" }, h("ix-icon-button", { type: "button", variant: "subtle-tertiary", icon: iconCloseSmall, class: "close-button", oval: true, size: "16", style: this.variant === 'custom' ? { color: this.chipColor } : {}, onClick: (event) => {
                this.closeChip.emit(event);
                event.stopPropagation();
            }, "aria-label": this.ariaLabelCloseButton })));
    }
    getTooltip() {
        var _a;
        if (!this.tooltipText && !this.hostElement.hasAttribute('tooltip-text')) {
            return null;
        }
        const text = typeof this.tooltipText === 'string' && this.tooltipText.trim()
            ? this.tooltipText
            : (_a = this.hostElement.textContent) === null || _a === void 0 ? void 0 : _a.trim();
        return (h("ix-tooltip", { for: this.containerElementRef.waitForCurrent() }, text));
    }
    render() {
        let customStyle = {};
        if (this.variant === 'custom') {
            customStyle = {
                color: this.chipColor,
                [this.outline ? 'borderColor' : 'backgroundColor']: this.background,
            };
        }
        return (h(Host, { key: '89e8d6fc156f9b2b761000d213655b36f3b16bf2', tabIndex: "-1", class: {
                inactive: this.inactive,
            }, style: this.variant === 'custom'
                ? {
                    '--ix-icon-button-color': this.chipColor,
                }
                : {} }, h("div", { key: 'a9b92e446d5c668d449996a93627879b428478e5', ref: this.containerElementRef, style: Object.assign({}, customStyle), class: {
                container: true,
                outline: this.outline,
                inactive: this.inactive,
                alarm: this.variant === 'alarm',
                critical: this.variant === 'critical',
                info: this.variant === 'info',
                neutral: this.variant === 'neutral',
                primary: this.variant === 'primary',
                success: this.variant === 'success',
                warning: this.variant === 'warning',
                custom: this.variant === 'custom',
                closable: this.closable,
                icon: !!this.icon,
                centerContent: this.centerContent,
            } }, h("div", { key: 'f3bf2fa31dfbaec0d014dbfa972ad822e2514e91', class: "content-wrapper" }, this.icon && (h("ix-icon", { key: '066a4f272fa760fafd6c69cd6b0f05bf459a3604', class: {
                'with-icon': true,
            }, name: this.icon, size: '24', style: this.variant === 'custom'
                ? { color: this.outline ? this.background : this.chipColor }
                : undefined })), h("span", { key: 'e08beb2e151300bdcdfe62c0b6a832935a7d2f68', class: "slot-container" }, h("slot", { key: 'b087848f9f85fc01af5cc3a03d4ae885e0f01335' }))), this.inactive === false && this.closable && this.getCloseButton()), this.getTooltip()));
    }
    static get is() { return "ix-chip"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["chip.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["chip.css"]
        };
    }
    static get properties() {
        return {
            "variant": {
                "type": "string",
                "attribute": "variant",
                "mutable": false,
                "complexType": {
                    "original": "| 'primary'\n    | 'alarm'\n    | 'critical'\n    | 'warning'\n    | 'info'\n    | 'neutral'\n    | 'success'\n    | 'custom'",
                    "resolved": "\"alarm\" | \"critical\" | \"custom\" | \"info\" | \"neutral\" | \"primary\" | \"success\" | \"warning\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Chip variant"
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "defaultValue": "'primary'"
            },
            "inactive": {
                "type": "boolean",
                "attribute": "inactive",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Determines if the chip is interactive. If true no user input (e.g. mouse states, keyboard navigation)\nwill be possible and also the close button will not be present."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "closable": {
                "type": "boolean",
                "attribute": "closable",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show close icon"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "icon": {
                "type": "string",
                "attribute": "icon",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Show icon"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "background": {
                "type": "string",
                "attribute": "background",
                "mutable": false,
                "complexType": {
                    "original": "string | undefined",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Custom background color.\nOnly has an effect on chips with `variant='custom'`"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "chipColor": {
                "type": "string",
                "attribute": "chip-color",
                "mutable": false,
                "complexType": {
                    "original": "string | undefined",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Custom font and icon color.\nOnly has an effect on chips with `variant='custom'`"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "outline": {
                "type": "boolean",
                "attribute": "outline",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show chip with outline style"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "tooltipText": {
                "type": "any",
                "attribute": "tooltip-text",
                "mutable": false,
                "complexType": {
                    "original": "string | boolean",
                    "resolved": "boolean | string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.0.0"
                        }],
                    "text": "Display a tooltip. By default, no tooltip will be displayed.\nAdd the attribute to display the text content of the component as a tooltip or use a string to display a custom text."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "centerContent": {
                "type": "boolean",
                "attribute": "center-content",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.2.0"
                        }],
                    "text": "Center the content of the chip.\nSet to false to disable centering."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "ariaLabelCloseButton": {
                "type": "string",
                "attribute": "aria-label-close-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "ARIA label for the close button\nWill be set as aria-label on the nested HTML button element"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            }
        };
    }
    static get events() {
        return [{
                "method": "closeChip",
                "name": "closeChip",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Fire event if close button is clicked"
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }];
    }
    static get elementRef() { return "hostElement"; }
}
