// @siemens/ix/dist/collection/components/input/number-input.js
/*
 * SPDX-FileCopyrightText: 2025 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
        r = Reflect.decorate(decorators, target, key, desc);
    else
        for (var i = decorators.length - 1; i >= 0; i--)
            if (d = decorators[i])
                r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { iconMinus, iconPlus } from "@siemens/ix-icons/icons";
import { h, Host, } from "@stencil/core";
import { HookValidationLifecycle, } from "../utils/input";
import { makeRef } from "../utils/make-ref";
import { InputElement, SlotEnd, SlotStart } from "./input.fc";
import { addDisposableChangesAndVisibilityObservers, adjustPaddingForStartAndEnd, checkAllowedKeys, checkInternalValidity, mapValidationResult, onInputBlur, } from "./input.util";
let numberInputIds = 0;
const INVALID_NUMBER_INPUT_REGEX = /[^\dEe+\-.,]/;
/**
 * @form-ready
 * @slot start - Element will be displayed at the start of the input
 * @slot end - Element will be displayed at the end of the input
 */
export class NumberInput {
    constructor() {
        /**
         * The value of the input field. Supports numeric values, scientific notation (1E6, 1E-6), or undefined for empty.
         */
        this.value = 0;
        /**
         * Indicates if the field is required. When required, empty values (undefined) are not accepted.
         */
        this.required = false;
        /**
         * Disables the input field
         */
        this.disabled = false;
        /**
         * Indicates if the field is read-only
         */
        this.readonly = false;
        /**
         * Step value to increment or decrement the input value. Default step value is 1.
         *
         */
        this.step = 1;
        /**
         * If false, pressing Enter will submit the form (if inside a form).
         * Set to true to suppress submit on Enter.
         */
        this.suppressSubmitOnEnter = false;
        /**
         * Text alignment within the number input. 'start' aligns the text to the start of the input, 'end' aligns the text to the end of the input.
         */
        this.textAlignment = 'end';
        /**
         * If true, the valueChange event will return null instead of 0 for an empty input state.
         * This property will be removed in 5.0.0 and this behaviour will be default.
         *
         * @since 4.1.0
         */
        this.allowEmptyValueChange = false;
        this.isInvalid = false;
        this.isValid = false;
        this.isInfo = false;
        this.isWarning = false;
        this.isInvalidByRequired = false;
        this.inputRef = makeRef();
        this.slotEndRef = makeRef();
        this.slotStartRef = makeRef();
        this.numberInputId = `number-input-${numberInputIds++}`;
        this.touched = false;
        this.handleInputChange = (inputValue) => {
            const parsedValue = this.convertNumberStringToFloat(inputValue);
            const isScientificNotation = this.isScientificNotation(inputValue.trim());
            if (isScientificNotation) {
                this.formInternals.setFormValue(inputValue);
            }
            this.handleValueChangeEvent(parsedValue);
        };
        this.handleBlur = () => {
            if (!this.inputRef.current)
                return;
            const inputValue = this.inputRef.current.value;
            const parsedValue = this.convertNumberStringToFloat(inputValue);
            if (parsedValue !== undefined) {
                this.inputRef.current.value = this.formatValue(parsedValue);
            }
            this.updateFormInternalValue(parsedValue);
            onInputBlur(this, this.inputRef.current);
            this.touched = true;
        };
        this.handleKeyDown = (event) => {
            if (this.disabled || this.readonly) {
                return;
            }
            switch (event.key) {
                case 'ArrowUp':
                    event.preventDefault();
                    this.handleStepOperation('up');
                    break;
                case 'ArrowDown':
                    event.preventDefault();
                    this.handleStepOperation('down');
                    break;
            }
        };
        this.handleBeforeInput = (e) => {
            var _a, _b;
            if (this.disabled || this.readonly)
                return;
            if (e.inputType === 'insertText') {
                const character = e.data;
                if (character && INVALID_NUMBER_INPUT_REGEX.test(character)) {
                    e.preventDefault();
                }
            }
            if (e.inputType === 'insertFromPaste') {
                const dt = e.dataTransfer || e.clipboardData;
                const text = (_b = (_a = dt === null || dt === void 0 ? void 0 : dt.getData) === null || _a === void 0 ? void 0 : _a.call(dt, 'text')) !== null && _b !== void 0 ? _b : '';
                if (INVALID_NUMBER_INPUT_REGEX.test(text)) {
                    e.preventDefault();
                }
            }
        };
        this.handlePaste = (e) => {
            var _a, _b;
            // Fallback for browsers that don’t fire beforeinput for paste
            const text = (_b = (_a = e.clipboardData) === null || _a === void 0 ? void 0 : _a.getData('text')) !== null && _b !== void 0 ? _b : '';
            if (INVALID_NUMBER_INPUT_REGEX.test(text)) {
                e.preventDefault();
            }
        };
    }
    onValueChange(newValue) {
        this.updateFormInternalValue(newValue);
    }
    updateClassMappings(result) {
        mapValidationResult(this, result);
    }
    componentWillLoad() {
        this.updateFormInternalValue(this.value);
    }
    connectedCallback() {
        this.disposableChangesAndVisibilityObservers =
            addDisposableChangesAndVisibilityObservers(this.hostElement, this.updatePaddings.bind(this));
    }
    disconnectedCallback() {
        var _a;
        (_a = this.disposableChangesAndVisibilityObservers) === null || _a === void 0 ? void 0 : _a.call(this);
    }
    updatePaddings() {
        adjustPaddingForStartAndEnd(this.slotStartRef.current, this.slotEndRef.current, this.inputRef.current);
    }
    convertNumberStringToFloat(input) {
        if (!input || input.trim() === '') {
            return undefined;
        }
        const parsed = Number.parseFloat(input);
        return Number.isNaN(parsed) ? undefined : parsed;
    }
    isScientificNotation(input) {
        const parsed = Number.parseFloat(input);
        return (!Number.isNaN(parsed) && Number.isFinite(parsed) && /[eE]/.test(input));
    }
    formatValue(value) {
        if (value === undefined || value === null) {
            return '';
        }
        return value.toString();
    }
    handleValueChangeEvent(value) {
        this.valueChange.emit(this.allowEmptyValueChange ? value : (value !== null && value !== void 0 ? value : 0));
    }
    updateFormInternalValue(value) {
        const formValue = value !== undefined && value !== null ? value.toString() : '';
        this.formInternals.setFormValue(formValue);
        this.value = value;
    }
    handleStepOperation(operation) {
        var _a, _b;
        if (!this.inputRef.current) {
            return;
        }
        const currentValue = (_a = this.value) !== null && _a !== void 0 ? _a : 0;
        const stepValue = typeof this.step === 'string'
            ? Number.parseFloat(this.step)
            : ((_b = this.step) !== null && _b !== void 0 ? _b : 1);
        let newValue;
        if (operation === 'up') {
            newValue = currentValue + stepValue;
        }
        else {
            newValue = currentValue - stepValue;
        }
        if (this.min !== undefined) {
            const minValue = typeof this.min === 'string' ? Number.parseFloat(this.min) : this.min;
            newValue = Math.max(newValue, minValue);
        }
        if (this.max !== undefined) {
            const maxValue = typeof this.max === 'string' ? Number.parseFloat(this.max) : this.max;
            newValue = Math.min(newValue, maxValue);
        }
        this.inputRef.current.value = newValue.toString();
        this.updateFormInternalValue(newValue);
        checkInternalValidity(this, this.inputRef.current);
        this.handleValueChangeEvent(newValue);
    }
    /** @internal */
    async getAssociatedFormElement() {
        return this.formInternals.form;
    }
    /** @internal */
    async hasValidValue() {
        const nativeInput = await this.getNativeInputElement();
        if (nativeInput.value === '') {
            return !this.required;
        }
        const parsedValue = this.convertNumberStringToFloat(nativeInput.value);
        return parsedValue !== undefined;
    }
    /**
     * Returns the native input element used under the hood
     */
    getNativeInputElement() {
        return this.inputRef.waitForCurrent();
    }
    /**
     * Focuses the input field
     */
    async focusInput() {
        return (await this.getNativeInputElement()).focus();
    }
    /**
     * Returns true if the input field has been touched
     * @internal
     */
    isTouched() {
        return Promise.resolve(this.touched);
    }
    render() {
        var _a;
        const showStepperButtons = this.showStepperButtons && (this.disabled || this.readonly) === false;
        return (h(Host, { key: '4b541dbe6fe2b32b5158eed8f398de300ea71bb4', class: {
                disabled: this.disabled,
                readonly: this.readonly,
            } }, h("ix-field-wrapper", { key: '7ef2bb50ba7c9b3b24612752155a2072329e4b61', id: this.numberInputId, required: this.required, label: this.label, helperText: this.helperText, invalidText: this.invalidText, infoText: this.infoText, warningText: this.warningText, validText: this.validText, showTextAsTooltip: this.showTextAsTooltip, isInvalid: this.isInvalid, isValid: this.isValid, isInfo: this.isInfo, isWarning: this.isWarning, controlRef: this.inputRef }, h("div", { key: '7e71d63d2a9ecbb16550925e68f3dc8dd76c3968', class: {
                'input-wrapper': true,
                'show-stepper-buttons': !!this.showStepperButtons,
            } }, h(SlotStart, { key: '8d936c9126f94c42db87b2d95d34ab178eadd31d', slotStartRef: this.slotStartRef, onSlotChange: () => this.updatePaddings() }), h(InputElement, { key: 'c56f510ea8a519204d3d36136d4b0a5a17246370', id: this.numberInputId, readonly: this.readonly, disabled: this.disabled, step: this.step, min: this.min, max: this.max, pattern: this.pattern, type: 'number', isInvalid: this.isInvalid, required: this.required, value: this.formatValue(this.value), placeholder: this.placeholder, inputRef: this.inputRef, onKeyPress: (event) => checkAllowedKeys(this, event), onKeyDown: (event) => this.handleKeyDown(event), onBeforeInput: (event) => this.handleBeforeInput(event), onPaste: (event) => this.handlePaste(event), valueChange: this.handleInputChange, updateFormInternalValue: (value) => {
                const isScientificNotation = this.isScientificNotation(value.trim());
                if (!isScientificNotation) {
                    const parsedValue = this.convertNumberStringToFloat(value);
                    this.updateFormInternalValue(parsedValue);
                }
            }, onBlur: this.handleBlur, form: (_a = this.formInternals.form) !== null && _a !== void 0 ? _a : undefined, suppressSubmitOnEnter: this.suppressSubmitOnEnter, textAlignment: this.textAlignment }), h(SlotEnd, { key: '2d7e82c421d51985c9d24c69b6f641279e3e84aa', slotEndRef: this.slotEndRef, onSlotChange: () => this.updatePaddings() }, h("div", { key: '5b1e6f03bf456305386537428b23b3fc2706e411', class: {
                'number-stepper-container': true,
                'container-hidden': !showStepperButtons,
            } }, h("ix-icon-button", { key: '44a5f583fb174c556a063f384174883ec307f9f4', variant: "subtle-tertiary", icon: iconMinus, size: "16", class: "number-stepper-button step-minus", "aria-label": "decrement number", onClick: () => this.handleStepOperation('down') }), h("ix-icon-button", { key: 'c007fe5451c3f124e61214160c7ebc8cc1150b25', variant: "subtle-tertiary", icon: iconPlus, size: "16", class: "number-stepper-button step-plus", "aria-label": "increment number", onClick: () => this.handleStepOperation('up') })))))));
    }
    static get is() { return "ix-number-input"; }
    static get encapsulation() { return "shadow"; }
    static get formAssociated() { return true; }
    static get originalStyleUrls() {
        return {
            "$": ["number-input.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["number-input.css"]
        };
    }
    static get properties() {
        return {
            "name": {
                "type": "string",
                "attribute": "name",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "name of the input element"
                },
                "getter": false,
                "setter": false,
                "reflect": true
            },
            "placeholder": {
                "type": "string",
                "attribute": "placeholder",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "placeholder of the input element"
                },
                "getter": false,
                "setter": false,
                "reflect": true
            },
            "value": {
                "type": "number",
                "attribute": "value",
                "mutable": true,
                "complexType": {
                    "original": "number",
                    "resolved": "number | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The value of the input field. Supports numeric values, scientific notation (1E6, 1E-6), or undefined for empty."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "defaultValue": "0"
            },
            "required": {
                "type": "boolean",
                "attribute": "required",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Indicates if the field is required. When required, empty values (undefined) are not accepted."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "defaultValue": "false"
            },
            "disabled": {
                "type": "boolean",
                "attribute": "disabled",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Disables the input field"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "readonly": {
                "type": "boolean",
                "attribute": "readonly",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Indicates if the field is read-only"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "helperText": {
                "type": "string",
                "attribute": "helper-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The helper text for the input field"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "infoText": {
                "type": "string",
                "attribute": "info-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The info text for the input field"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "showTextAsTooltip": {
                "type": "boolean",
                "attribute": "show-text-as-tooltip",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Indicates if the text should be shown as a tooltip"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "validText": {
                "type": "string",
                "attribute": "valid-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The valid text for the input field"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "warningText": {
                "type": "string",
                "attribute": "warning-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The warning text for the input field"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "label": {
                "type": "string",
                "attribute": "label",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The label for the input field"
                },
                "getter": false,
                "setter": false,
                "reflect": true
            },
            "invalidText": {
                "type": "string",
                "attribute": "invalid-text",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The error text for the input field"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "pattern": {
                "type": "string",
                "attribute": "pattern",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The pattern for the input field"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "min": {
                "type": "any",
                "attribute": "min",
                "mutable": false,
                "complexType": {
                    "original": "string | number",
                    "resolved": "number | string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The minimum value for the input field"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "max": {
                "type": "any",
                "attribute": "max",
                "mutable": false,
                "complexType": {
                    "original": "string | number",
                    "resolved": "number | string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The maximum value for the input field"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "allowedCharactersPattern": {
                "type": "string",
                "attribute": "allowed-characters-pattern",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The allowed characters pattern for the input field"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "showStepperButtons": {
                "type": "boolean",
                "attribute": "show-stepper-buttons",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Indicates if the stepper buttons should be shown"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "step": {
                "type": "any",
                "attribute": "step",
                "mutable": false,
                "complexType": {
                    "original": "string | number",
                    "resolved": "number | string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Step value to increment or decrement the input value. Default step value is 1."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "1"
            },
            "suppressSubmitOnEnter": {
                "type": "boolean",
                "attribute": "suppress-submit-on-enter",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "If false, pressing Enter will submit the form (if inside a form).\nSet to true to suppress submit on Enter."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "defaultValue": "false"
            },
            "textAlignment": {
                "type": "string",
                "attribute": "text-alignment",
                "mutable": false,
                "complexType": {
                    "original": "'start' | 'end'",
                    "resolved": "\"end\" | \"start\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Text alignment within the number input. 'start' aligns the text to the start of the input, 'end' aligns the text to the end of the input."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'end'"
            },
            "allowEmptyValueChange": {
                "type": "boolean",
                "attribute": "allow-empty-value-change",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.1.0"
                        }],
                    "text": "If true, the valueChange event will return null instead of 0 for an empty input state.\nThis property will be removed in 5.0.0 and this behaviour will be default."
                },
                "getter": false,
                "setter": false,
                "reflect": true,
                "defaultValue": "false"
            }
        };
    }
    static get states() {
        return {
            "isInvalid": {},
            "isValid": {},
            "isInfo": {},
            "isWarning": {},
            "isInvalidByRequired": {}
        };
    }
    static get events() {
        return [{
                "method": "valueChange",
                "name": "valueChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the value of the input field changes"
                },
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                }
            }, {
                "method": "validityStateChange",
                "name": "validityStateChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the validity state of the input field changes"
                },
                "complexType": {
                    "original": "ValidityState",
                    "resolved": "ValidityState",
                    "references": {
                        "ValidityState": {
                            "location": "global",
                            "id": "global::ValidityState"
                        }
                    }
                }
            }, {
                "method": "ixBlur",
                "name": "ixBlur",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Event emitted when the input field loses focus"
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }];
    }
    static get methods() {
        return {
            "getAssociatedFormElement": {
                "complexType": {
                    "signature": "() => Promise<HTMLFormElement | null>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "HTMLFormElement": {
                            "location": "global",
                            "id": "global::HTMLFormElement"
                        }
                    },
                    "return": "Promise<HTMLFormElement | null>"
                },
                "docs": {
                    "text": "",
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }]
                }
            },
            "hasValidValue": {
                "complexType": {
                    "signature": "() => Promise<boolean>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<boolean>"
                },
                "docs": {
                    "text": "",
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }]
                }
            },
            "getNativeInputElement": {
                "complexType": {
                    "signature": "() => Promise<HTMLInputElement>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        },
                        "HTMLInputElement": {
                            "location": "global",
                            "id": "global::HTMLInputElement"
                        }
                    },
                    "return": "Promise<HTMLInputElement>"
                },
                "docs": {
                    "text": "Returns the native input element used under the hood",
                    "tags": []
                }
            },
            "focusInput": {
                "complexType": {
                    "signature": "() => Promise<void>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<void>"
                },
                "docs": {
                    "text": "Focuses the input field",
                    "tags": []
                }
            },
            "isTouched": {
                "complexType": {
                    "signature": "() => Promise<boolean>",
                    "parameters": [],
                    "references": {
                        "Promise": {
                            "location": "global",
                            "id": "global::Promise"
                        }
                    },
                    "return": "Promise<boolean>"
                },
                "docs": {
                    "text": "Returns true if the input field has been touched",
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }]
                }
            }
        };
    }
    static get elementRef() { return "hostElement"; }
    static get watchers() {
        return [{
                "propName": "value",
                "methodName": "onValueChange"
            }];
    }
    static get attachInternalsMemberName() { return "formInternals"; }
}
__decorate([
    HookValidationLifecycle()
], NumberInput.prototype, "updateClassMappings", null);
