// @siemens/ix/dist/collection/components/empty-state/empty-state.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host } from "@stencil/core";
export class EmptyState {
    constructor() {
        /**
         * Optional empty state layout - one of 'large', 'compact' or 'compactBreak'
         */
        this.layout = 'large';
    }
    render() {
        return (h(Host, { key: '40044bc561b77f0ca121fec086c4e50959cd84b8', class: `emptyState emptyState--${this.layout}` }, this.icon && (h("div", { key: '4f7caad68b5bfdae9a3becc0d9cb86e6aa121b9a', class: "emptyState__icon" }, h("ix-icon", { key: '4f3540bc8589358f3a81136bf18e37a963ccf8c4', name: this.icon, size: this.layout === 'large' ? '32' : '32', color: "color-soft-text", "aria-label": this.ariaLabelEmptyStateIcon }))), h("div", { key: 'aa658e73609ff249ada8b4a11897bc6fd16f3e27', class: "emptyState__content" }, h("div", { key: 'f46a03dcf711bdfc3919d3f94e5b60e64f434173', class: "content__label" }, h("ix-typography", { key: 'dd56a7bc024921adedf792ed956614a13a7dd723', format: this.layout === 'large' ? 'h3' : 'body' }, this.header), this.subHeader && (h("div", { key: '7813b0fd9e490328b24566ae0abb805dbc2d1b3b', class: "label__subHeader" }, this.subHeader))), this.action && (h("div", { key: '4ffbf8f873bff95e04376dac344cd312d39a68d6', class: "content__action" }, h("ix-button", { key: '20739fd736c3176034dcc6d564c09879f745b9b3', onClick: () => this.actionClick.emit() }, this.action))))));
    }
    static get is() { return "ix-empty-state"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["empty-state.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["empty-state.css"]
        };
    }
    static get properties() {
        return {
            "layout": {
                "type": "string",
                "attribute": "layout",
                "mutable": false,
                "complexType": {
                    "original": "EmptyStateLayout",
                    "resolved": "\"compact\" | \"compactBreak\" | \"large\"",
                    "references": {
                        "EmptyStateLayout": {
                            "location": "import",
                            "path": "./empty-state.types",
                            "id": "src/components/empty-state/empty-state.types.ts::EmptyStateLayout"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Optional empty state layout - one of 'large', 'compact' or 'compactBreak'"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'large'"
            },
            "icon": {
                "type": "string",
                "attribute": "icon",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Optional empty state icon"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "header": {
                "type": "string",
                "attribute": "header",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": true,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Empty state header"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "subHeader": {
                "type": "string",
                "attribute": "sub-header",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Optional empty state sub header"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "action": {
                "type": "string",
                "attribute": "action",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Optional empty state action"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "ariaLabelEmptyStateIcon": {
                "type": "string",
                "attribute": "aria-label-empty-state-icon",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.2.0"
                        }],
                    "text": "ARIA label for the empty state icon"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            }
        };
    }
    static get events() {
        return [{
                "method": "actionClick",
                "name": "actionClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Empty state action click event"
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }];
    }
}
