// @siemens/ix/dist/collection/components/content-header/content-header.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { iconArrowLeft } from "@siemens/ix-icons/icons";
import { h, Host } from "@stencil/core";
/**
 * @slot header - Content to be placed in the header area next to the title
 * @slot - Default slot for action buttons or other content
 */
export class ContentHeader {
    constructor() {
        /**
         * Variant of content header
         */
        this.variant = 'primary';
        /**
         * Subtitle of Header
         */
        this.headerSubtitle = undefined;
        /**
         * Display a back button
         */
        this.hasBackButton = false;
    }
    render() {
        return (h(Host, { key: 'b422be3706aac399ed76b4c491d61e2da8561b24' }, this.hasBackButton ? (h("ix-icon-button", { class: 'backButton', variant: "tertiary", icon: iconArrowLeft, onClick: () => this.backButtonClick.emit() })) : null, h("div", { key: '7606ba97f834f591a78fc2cca74bdaf3b3a09feb', class: "titleGroup" }, h("div", { key: 'df174e297a1b2075bcab0182a80f626162d1c319', class: "headerTitleRow" }, h("ix-typography", { key: 'd8ab10dc801306dafe0f3e6e39ed93e0fa162ca1', format: this.variant === 'secondary' ? 'h4' : 'h3', class: {
                secondary: this.variant === 'secondary',
                titleOverflow: true,
            } }, this.headerTitle), h("div", { key: 'e24d6b450426931aaead3732fa20161bc5aab4e2', class: "headerSlot" }, h("slot", { key: 'e9a471492f2693fc6abe546d1febbc06834fef1c', name: "header" }))), !!this.headerSubtitle && (h("ix-typography", { key: '5fdb5ebce68a3aae09df3299093bca156468933c', format: 'h6', "text-color": 'soft', class: {
                subtitle: this.variant === 'secondary',
                titleOverflow: true,
            }, title: this.headerSubtitle }, this.headerSubtitle))), h("div", { key: '6648f93246d2e8d79cc2d3e7eae65009d578632c', class: "buttons" }, h("slot", { key: '1c3ee8084521a9a799e3b2444b1edc0c1cee5066' }))));
    }
    static get is() { return "ix-content-header"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["content-header.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["content-header.css"]
        };
    }
    static get properties() {
        return {
            "variant": {
                "type": "string",
                "attribute": "variant",
                "mutable": false,
                "complexType": {
                    "original": "ContentHeaderVariant",
                    "resolved": "\"primary\" | \"secondary\"",
                    "references": {
                        "ContentHeaderVariant": {
                            "location": "import",
                            "path": "./content-header.types",
                            "id": "src/components/content-header/content-header.types.ts::ContentHeaderVariant"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Variant of content header"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'primary'"
            },
            "headerTitle": {
                "type": "string",
                "attribute": "header-title",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Title of Header"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "headerSubtitle": {
                "type": "string",
                "attribute": "header-subtitle",
                "mutable": false,
                "complexType": {
                    "original": "string | undefined",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Subtitle of Header"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "undefined"
            },
            "hasBackButton": {
                "type": "boolean",
                "attribute": "has-back-button",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Display a back button"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
    static get events() {
        return [{
                "method": "backButtonClick",
                "name": "backButtonClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Triggered when back button is clicked"
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }];
    }
}
