// @siemens/ix/dist/collection/components/datetime-picker/datetime-picker.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host } from "@stencil/core";
export class DatetimePicker {
    constructor() {
        /**
         * If true disables date range selection (from/to).
         */
        this.singleSelection = false;
        /**
         * Date format string.
         * See {@link https://moment.github.io/luxon/#/formatting?id=table-of-tokens} for all available tokens.
         */
        this.dateFormat = 'yyyy/LL/dd';
        /**
         * Time format string.
         * See {@link https://moment.github.io/luxon/#/formatting?id=table-of-tokens} for all available tokens.
         */
        this.timeFormat = 'HH:mm:ss';
        /**
         * Show time reference input
         * Time reference is default aligned with @see {this.timeFormat}
         */
        this.showTimeReference = false;
        /**
         * Text of date select button
         */
        this.i18nDone = 'Done';
        /**
         * Top label of time picker
         *
         * @since 3.0.0
         */
        this.i18nTime = 'Time';
        /**
         * The index of which day to start the week on, based on the Locale#weekdays array.
         * E.g. if the locale is en-us, weekStartIndex = 1 results in starting the week on monday.
         */
        this.weekStartIndex = 0;
        /**
         * Shows week numbers displayed on the left side of the date picker
         *
         * @since 3.0.0
         */
        this.showWeekNumbers = false;
    }
    async onDone() {
        var _a, _b, _c, _d;
        const date = await ((_a = this.datePickerElement) === null || _a === void 0 ? void 0 : _a.getCurrentDate());
        const time = await ((_b = this.timePickerElement) === null || _b === void 0 ? void 0 : _b.getCurrentTime());
        this.dateSelect.emit({
            from: (_c = date === null || date === void 0 ? void 0 : date.from) !== null && _c !== void 0 ? _c : '',
            to: (_d = date === null || date === void 0 ? void 0 : date.to) !== null && _d !== void 0 ? _d : '',
            time: time !== null && time !== void 0 ? time : '',
        });
    }
    async onDateChange(event) {
        event.preventDefault();
        event.stopPropagation();
        const { detail: date } = event;
        this.dateChange.emit(date);
    }
    async onTimeChange(event) {
        event.preventDefault();
        event.stopPropagation();
        const { detail: time } = event;
        this.timeChange.emit(time);
    }
    render() {
        return (h(Host, { key: '2b4812deec9408836a8187c3a520cd69b846e935' }, h("ix-layout-grid", { key: '610fad916eb544d6687bcbfb3543b926f7521b9e', class: "no-padding" }, h("ix-row", { key: '63dcab9421e05163577db32fda3bd714a0c00de7', class: "row-separator" }, h("ix-col", { key: '8ac25a8147b15ac20b45c3f1e48bfda34919eae1', class: "col-separator" }, h("ix-date-picker", { key: '1831729d8dcb31e5361fbef4a6290111847b8679', ref: (ref) => (this.datePickerElement = ref), corners: "left", singleSelection: this.singleSelection, onDateChange: (event) => this.onDateChange(event), from: this.from, to: this.to, format: this.dateFormat, minDate: this.minDate, maxDate: this.maxDate, weekStartIndex: this.weekStartIndex, embedded: true, locale: this.locale, showWeekNumbers: this.showWeekNumbers, ariaLabelPreviousMonthButton: this.ariaLabelPreviousMonthButton, ariaLabelNextMonthButton: this.ariaLabelNextMonthButton })), h("ix-col", { key: '51b004e4362c9dc26d0e59909cfb431fd874cbac' }, h("ix-time-picker", { key: '40672098b7e7cc369afd8429b34d14b97bcd1380', class: "min-width", ref: (ref) => (this.timePickerElement = ref), embedded: true, dateTimePickerAppearance: true, onTimeChange: (event) => this.onTimeChange(event), format: this.timeFormat, time: this.time }))), h("ix-row", { key: '876f3e8defe5de50013e8ea7e1f3465bd6e339a1' }, h("ix-col", { key: '1515e3e7b9ac5b121c8b0c3284092698f85652bf' }, h("div", { key: '6aa46d00d240c20de2e92bd5fa62f6ca59f53274', class: "btn-select-date-container" }, h("ix-button", { key: 'ca4e27ecc31600f97ac2ab061cbbf29b2c717078', class: "btn-select-date", onClick: () => this.onDone() }, this.i18nDone)))))));
    }
    static get is() { return "ix-datetime-picker"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["datetime-picker.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["datetime-picker.css"]
        };
    }
    static get properties() {
        return {
            "singleSelection": {
                "type": "boolean",
                "attribute": "single-selection",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "If true disables date range selection (from/to)."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "minDate": {
                "type": "string",
                "attribute": "min-date",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The earliest date that can be selected by the date picker.\nIf not set there will be no restriction."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "maxDate": {
                "type": "string",
                "attribute": "max-date",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The latest date that can be selected by the date picker.\nIf not set there will be no restriction."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "dateFormat": {
                "type": "string",
                "attribute": "date-format",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Date format string.\nSee {@link https://moment.github.io/luxon/#/formatting?id=table-of-tokens} for all available tokens."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'yyyy/LL/dd'"
            },
            "timeFormat": {
                "type": "string",
                "attribute": "time-format",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Time format string.\nSee {@link https://moment.github.io/luxon/#/formatting?id=table-of-tokens} for all available tokens."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'HH:mm:ss'"
            },
            "from": {
                "type": "string",
                "attribute": "from",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The selected starting date. If the picker is not in range mode this is the selected date.\nFormat has to match the `format` property."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "to": {
                "type": "string",
                "attribute": "to",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "The selected end date. If the the picker is not in range mode this property has no impact.\nFormat has to match the `format` property."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "time": {
                "type": "string",
                "attribute": "time",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Select time with format string"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "showTimeReference": {
                "type": "boolean",
                "attribute": "show-time-reference",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "see",
                            "text": "{ this.timeFormat}"
                        }],
                    "text": "Show time reference input\nTime reference is default aligned with"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "timeReference": {
                "type": "string",
                "attribute": "time-reference",
                "mutable": false,
                "complexType": {
                    "original": "'AM' | 'PM'",
                    "resolved": "\"AM\" | \"PM\" | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Set time reference"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "i18nDone": {
                "type": "string",
                "attribute": "i18n-done",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Text of date select button"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'Done'"
            },
            "i18nTime": {
                "type": "string",
                "attribute": "i18n-time",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.0.0"
                        }],
                    "text": "Top label of time picker"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'Time'"
            },
            "ariaLabelPreviousMonthButton": {
                "type": "string",
                "attribute": "aria-label-previous-month-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "ARIA label for the previous month icon button\nWill be set as aria-label on the nested HTML button element"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "ariaLabelNextMonthButton": {
                "type": "string",
                "attribute": "aria-label-next-month-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "ARIA label for the next month icon button\nWill be set as aria-label on the nested HTML button element"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "weekStartIndex": {
                "type": "number",
                "attribute": "week-start-index",
                "mutable": false,
                "complexType": {
                    "original": "number",
                    "resolved": "number",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The index of which day to start the week on, based on the Locale#weekdays array.\nE.g. if the locale is en-us, weekStartIndex = 1 results in starting the week on monday."
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "0"
            },
            "locale": {
                "type": "string",
                "attribute": "locale",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Format of time string\nSee {@link https://moment.github.io/luxon/#/formatting?id=table-of-tokens} for all available tokens."
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "showWeekNumbers": {
                "type": "boolean",
                "attribute": "show-week-numbers",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.0.0"
                        }],
                    "text": "Shows week numbers displayed on the left side of the date picker"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
    static get events() {
        return [{
                "method": "timeChange",
                "name": "timeChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Time change"
                },
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                }
            }, {
                "method": "dateChange",
                "name": "dateChange",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Date change"
                },
                "complexType": {
                    "original": "DateTimeDateChangeEvent",
                    "resolved": "string | { from?: string | undefined; to?: string | undefined; }",
                    "references": {
                        "DateTimeDateChangeEvent": {
                            "location": "import",
                            "path": "./datetime-picker.types",
                            "id": "src/components/datetime-picker/datetime-picker.types.ts::DateTimeDateChangeEvent"
                        }
                    }
                }
            }, {
                "method": "dateSelect",
                "name": "dateSelect",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "Datetime selection event is fired after confirm button is pressed"
                },
                "complexType": {
                    "original": "DateTimeSelectEvent",
                    "resolved": "{ from?: string | undefined; to?: string | undefined; time: string; }",
                    "references": {
                        "DateTimeSelectEvent": {
                            "location": "import",
                            "path": "./datetime-picker.types",
                            "id": "src/components/datetime-picker/datetime-picker.types.ts::DateTimeSelectEvent"
                        }
                    }
                }
            }];
    }
}
