// @siemens/ix/dist/collection/components/card-accordion/card-accordion.js
import { iconChevronDownSmall } from "@siemens/ix-icons/icons";
import { h, Host, } from "@stencil/core";
let accordionControlId = 0;
const getAriaControlsId = (prefix = 'expand-content') => {
    return [prefix, accordionControlId++].join('-');
};
export class CardAccordion {
    constructor() {
        /**
         * Collapse the card
         */
        this.collapse = false;
        /**
         * Show accordion with different color variants
         *
         * @since 4.0.0
         */
        this.variant = 'outline';
        this.expandContent = false;
    }
    onInitialExpandChange() {
        this.expandContent = !this.collapse;
    }
    get expandedContent() {
        return this.hostElement.shadowRoot.querySelector('.expand-content');
    }
    onExpandActionClick(event) {
        event.preventDefault();
        event.stopPropagation();
        this.expandContent = !this.expandContent;
        this.accordionExpand.emit({
            expand: this.expandContent,
            nativeEvent: event,
        });
        if (this.expandContent) {
            this.scrollExpandedContentIntoView();
        }
    }
    scrollExpandedContentIntoView() {
        setTimeout(() => {
            const rect = this.expandedContent.getBoundingClientRect();
            if (rect.bottom > window.innerHeight) {
                this.hostElement
                    .shadowRoot.querySelector('.expand-content')
                    .scrollIntoView(false);
            }
        }, 150);
    }
    componentWillLoad() {
        this.onInitialExpandChange();
    }
    render() {
        return (h(Host, { key: 'dc1b41f8878a1aa87869d11829c9a3cc218ca12e', slot: "card-accordion", class: {
                [`variant-${this.variant}`]: true,
            } }, h("button", { key: '0fc447aadb0c86405700f00c18d17be887eef3fb', tabIndex: 0, class: { 'expand-action': true, show: this.expandContent }, onClick: (event) => this.onExpandActionClick(event), role: "button", type: "button", "aria-expanded": this.expandContent, "aria-controls": getAriaControlsId(), "aria-label": this.ariaLabelExpandButton }, h("ix-icon", { key: 'b044d6c5c9561355750483c29824bfe27af75db5', name: iconChevronDownSmall, class: {
                'expand-icon': true,
                show: this.expandContent,
            } })), h("div", { key: 'df7c48aaa25ad38136479fb972ffe0776efa0ae3', class: {
                'expand-content': true,
                show: this.expandContent,
            } }, h("div", { key: '9f9bd7ef6d674f0d6561ac9c237443062eec9693', class: "expand-content-inner" }, h("div", { key: '1576b901ba50a5ec7e765ae64e8cd6acba047973', class: "expand-content-body" }, h("slot", { key: 'ef118ded7d20ba6dd70c45a0d34f86a89959e69c' })), h("div", { key: '248e9c925bf36e9598386a78562abd8a3db77c06', class: "expand-content-footer" })))));
    }
    static get is() { return "ix-card-accordion"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["card-accordion.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["card-accordion.css"]
        };
    }
    static get properties() {
        return {
            "ariaLabelExpandButton": {
                "type": "string",
                "attribute": "aria-label-expand-button",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string | undefined",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "3.2.0"
                        }],
                    "text": "ARIA label for the card's expand button.\nWill be set as aria-label on the nested HTML button element"
                },
                "getter": false,
                "setter": false,
                "reflect": false
            },
            "collapse": {
                "type": "boolean",
                "attribute": "collapse",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Collapse the card"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            },
            "variant": {
                "type": "string",
                "attribute": "variant",
                "mutable": false,
                "complexType": {
                    "original": "CardAccordionVariant",
                    "resolved": "\"alarm\" | \"critical\" | \"filled\" | \"info\" | \"neutral\" | \"outline\" | \"primary\" | \"success\" | \"warning\"",
                    "references": {
                        "CardAccordionVariant": {
                            "location": "import",
                            "path": "./card-accordion.types",
                            "id": "src/components/card-accordion/card-accordion.types.ts::CardAccordionVariant"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [{
                            "name": "since",
                            "text": "4.0.0"
                        }],
                    "text": "Show accordion with different color variants"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'outline'"
            }
        };
    }
    static get states() {
        return {
            "expandContent": {}
        };
    }
    static get events() {
        return [{
                "method": "accordionExpand",
                "name": "accordionExpand",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }],
                    "text": ""
                },
                "complexType": {
                    "original": "CardAccordionExpandChangeEvent",
                    "resolved": "{ expand: boolean; nativeEvent: Event; }",
                    "references": {
                        "CardAccordionExpandChangeEvent": {
                            "location": "import",
                            "path": "./card-accordion.types",
                            "id": "src/components/card-accordion/card-accordion.types.ts::CardAccordionExpandChangeEvent"
                        }
                    }
                }
            }];
    }
    static get elementRef() { return "hostElement"; }
    static get watchers() {
        return [{
                "propName": "collapse",
                "methodName": "onInitialExpandChange"
            }];
    }
}
