// @siemens/ix/dist/collection/components/card/card.js
import { h, Host } from "@stencil/core";
export class Card {
    constructor() {
        /**
         * Card variant
         */
        this.variant = 'outline';
        /**
         * Show card in selected state
         */
        this.selected = false;
    }
    render() {
        return (h(Host, { key: '3b734eddf8516af24b52b1b2ba27766a913f943d', class: {
                selected: this.selected,
                [`card-${this.variant}`]: true,
            } }, h("div", { key: '7a2cce1797c1e89c8568e9e196020747f0f4150b', class: "card-content" }, h("slot", { key: 'ea837bde282a62637ed15b96b8ca407219e949ae' })), h("div", { key: 'acc07c72e4a63a671fb945e4c0332825f4781a1d', class: "card-footer" }, h("slot", { key: '870004ea22909180bac56c8b7c040d914e268f1d', name: "card-accordion" }))));
    }
    static get is() { return "ix-card"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["card.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["card.css"]
        };
    }
    static get properties() {
        return {
            "variant": {
                "type": "string",
                "attribute": "variant",
                "mutable": false,
                "complexType": {
                    "original": "CardVariant",
                    "resolved": "\"alarm\" | \"critical\" | \"filled\" | \"info\" | \"neutral\" | \"outline\" | \"primary\" | \"success\" | \"warning\"",
                    "references": {
                        "CardVariant": {
                            "location": "import",
                            "path": "./card.types",
                            "id": "src/components/card/card.types.ts::CardVariant"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Card variant"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "'outline'"
            },
            "selected": {
                "type": "boolean",
                "attribute": "selected",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Show card in selected state"
                },
                "getter": false,
                "setter": false,
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
    static get elementRef() { return "hostElement"; }
}
