// @siemens/ix/dist/collection/components/css-grid/css-grid.js
/*
 * SPDX-FileCopyrightText: 2023 Siemens AG
 *
 * SPDX-License-Identifier: MIT
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { h, Host } from "@stencil/core";
import { templateBuilder } from "./builder";
const smQuery = 'only screen and (max-width: 576px)';
const mdQuery = 'only screen and (max-width: 768px)';
const lgQuery = 'only screen and (max-width: 992px)';
const mediaQueryCollection = [
    { name: 'sm', query: smQuery },
    { name: 'md', query: mdQuery },
    { name: 'lg', query: lgQuery },
];
/**
 * @internal
 */
export class CssGrid {
    constructor() {
        this.disposeMediaQueryListener = [];
        this.mediaQueries = [];
    }
    componentWillRender() {
        this.mediaQueries = mediaQueryCollection.map((query) => {
            const mediaQuery = window.matchMedia(query.query);
            const callback = () => {
                this.applyTemplate();
            };
            this.disposeMediaQueryListener.push(callback);
            mediaQuery.addEventListener('change', callback);
            return Object.assign(Object.assign({}, query), { mediaQuery });
        });
        this.applyTemplate();
    }
    disconnectedCallback() {
        this.mediaQueries.forEach((mq, index) => {
            mq.mediaQuery.removeEventListener('change', this.disposeMediaQueryListener[index]);
        });
    }
    findNextTemplate(type) {
        if (!this.templates) {
            return [];
        }
        const typeIndex = this.mediaQueries.findIndex((mq) => mq.name === type);
        const nextTemplate = this.templates[this.mediaQueries[typeIndex + 1].name];
        if (!nextTemplate) {
            return this.findNextTemplate(this.mediaQueries[typeIndex + 1].name);
        }
        return nextTemplate;
    }
    applyTemplate() {
        if (!this.templates) {
            return;
        }
        let active = this.mediaQueries.find((mq) => mq.mediaQuery.matches);
        if (!active) {
            active = this.mediaQueries[this.mediaQueries.length - 1];
        }
        const template = this.templates[active.name];
        if (template) {
            this.currentTemplate = template;
        }
        else {
            this.currentTemplate = this.findNextTemplate(active.name);
        }
    }
    render() {
        const style = {};
        if (this.currentTemplate && this.currentTemplate.length !== 0) {
            style['grid-template-areas'] = templateBuilder(this.currentTemplate);
        }
        return (h(Host, { key: '470c3da8301e7d99e603b78cfd9c2883e4801c76', style: style }, h("slot", { key: '64f7e86981ba37aeb20223bd672bc1190dba4619' })));
    }
    static get is() { return "ix-css-grid"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["css-grid.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["css-grid.css"]
        };
    }
    static get properties() {
        return {
            "templates": {
                "type": "unknown",
                "attribute": "templates",
                "mutable": false,
                "complexType": {
                    "original": "Partial<Record<CssGridTemplateType, string[][]>>",
                    "resolved": "undefined | { sm?: string[][] | undefined; md?: string[][] | undefined; lg?: string[][] | undefined; }",
                    "references": {
                        "Partial": {
                            "location": "global",
                            "id": "global::Partial"
                        },
                        "Record": {
                            "location": "global",
                            "id": "global::Record"
                        },
                        "CssGridTemplateType": {
                            "location": "import",
                            "path": "./css-grid.types",
                            "id": "src/components/css-grid/css-grid.types.ts::CssGridTemplateType"
                        }
                    }
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "Define css grid template"
                },
                "getter": false,
                "setter": false
            }
        };
    }
    static get states() {
        return {
            "currentTemplate": {}
        };
    }
}
